{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build Excel stoich for GAMS (originally A01_build_excel_stoich_for_GAMS.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "import cobra,json,os,sys,openpyxl,re,pandas as pd\n",
    "from collections import OrderedDict\n",
    "from copy import deepcopy\n",
    "\n",
    "model_root_path = '../'\n",
    "sys.path.append(model_root_path)\n",
    "from pycore.gsm_custom_functions import *\n",
    "from pycore.cobrapy_custom_extras import *\n",
    "kribo_default = 22*3600 #(22 aa/ribo/s converted to aa/ribo/h)\n",
    "mito_compartments = ['m', 'mm'] # IDs for mitochondria compartments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reactions to ignore when checking if enzymes must be added:\n",
      "['BIOMASS_AERO_SC_hvd', 'BIOMASS_MFA', 'BIOMASS_MFA_NO_GAM']\n"
     ]
    }
   ],
   "source": [
    "# Metabolic model (COBRApy json)\n",
    "# If you have to use multiple stoichiometric models (SMs), \n",
    "#   it's best to keep 1 as the main copy and then use transaction files to apply differences to the others. Helps improve consistency in annotations, stoich. etc.\n",
    "chosen_model = '_rba' # c: carbon, n: nitrogen\n",
    "\n",
    "# List parameters for your model version(s), such as:\n",
    "\t# max_allowed_mito_proteome_allo_fraction: maximum fraction of proteome that can be allocated to mitochondria\n",
    "\t# nonmodeled_proteome_allocation: fraction of proteome not accounted for in the model\n",
    "\t# predicted_but_unused_proteins_path: path to file with predicted proteins that are not used in the model\n",
    "\t# sheet_name: name of sheet in relevant excel files (e.g., biomass, dummy protein) to use\n",
    "modelVersions = {\n",
    "    '_rba':{\n",
    "        'max_allowed_mito_proteome_allo_fraction':0.05,\n",
    "        # 'nonmodeled_proteome_allocation':0.3148, # disable since MPFS constraints now replace this\n",
    "        'dummy_sheet':''\n",
    "    }\n",
    "    }\n",
    "modelname = 'GSM_iSace1144_rba'\n",
    "model = load_cobra_model('./input/'+modelname+'.json')\n",
    "# if not choosing between nutrient-limited models, uncomment the following lines:\n",
    "# model = cobra.io.load_json_model('./input/iRhtoC.json')\n",
    "# chosen_model = ''\n",
    "\n",
    "# Update GAMS settings file to reflect model version, \n",
    "#   and add NGAM, mito capacity, etc.\n",
    "#   If not using multiple models, replace everything in the curly braces with the values you want\n",
    "with open('./model/model-version-settings.txt', 'w') as f:\n",
    "\tnpa = modelVersions[chosen_model].get('nonmodeled_proteome_allocation', 0)\n",
    "\tmampaf = modelVersions[chosen_model]['max_allowed_mito_proteome_allo_fraction'] if 'max_allowed_mito_proteome_allo_fraction' in modelVersions[chosen_model] else 1\n",
    "\tf.write(f\"* file with settings updated by build_model scripts, to help with managing multiple SMs\\n\"\n",
    "\t\tf\"$setGlobal max_allowed_mito_proteome_allo_fraction {mampaf}\\n\"\n",
    "\t\tf\"$setGlobal nonmodeled_proteome_allocation {npa}\\n\")\n",
    "\tif 'measured_but_unused_proteins_path' in modelVersions[chosen_model]:\n",
    "\t\tf.write('* unless specified otherwise (e.g., in test_kapp.gms), include prosyn constraints to reflect production of unused proteins; comment out if using dummy protein instead\\n')\n",
    "\t\tif modelVersions[chosen_model]['predicted_but_unused_proteins_path']:\n",
    "\t\t\tif os.path.exists(modelVersions[chosen_model]['predicted_but_unused_proteins_path']):\n",
    "\t\t\t\tf.write(f\"$if not %ignore_measured_unused_constraints%==1 $include {modelVersions[chosen_model]['predicted_but_unused_proteins_path']}\\n\")\n",
    "\n",
    "# Protein\n",
    "df_pro = read_spreadsheet('./input/PROTEIN_stoich_curation.xlsx')\n",
    "## OPTIONAL: add \"special cases only\" sheet to df_pro. \n",
    "##   Good for distinguishing proteins in the organism from ones added for testing purposes.\n",
    "include_special_case_proteins = False # affects protein and enzyme stoichiometry\n",
    "if include_special_case_proteins:\n",
    "\tdf_pro_special = read_spreadsheet('./input/PROTEIN_stoich_curation.xlsx',sheet_name='special cases only',header=1)\n",
    "\tdf_pro = pd.concat([df_pro,df_pro_special],ignore_index=True)\n",
    "\n",
    "# Amino acid info\n",
    "df_aamap = pd.read_csv('./input/PROTEIN_amino_acid_map.txt', sep='\\t')\n",
    "df_aamap.index = df_aamap.aa_abbv.to_list()\n",
    "\n",
    "# Dummy protein\n",
    "dummy_path = './input/PROTEIN_dummy_prot_calc.xlsx'\n",
    "# df_aa_dummy = read_spreadsheet(dummy_path)\n",
    "# df_aa_dummy2 = read_spreadsheet(dummy_path,sheet_name='unidentified')\n",
    "# 1st entry is default dummy protein; remove 2nd entry unless needed\n",
    "dummy_list = [{\"df\":read_spreadsheet(dummy_path),\"name\":\"PROSYN-PROTDUMMY\"}]\n",
    "\t\t\t#   {\"df\":read_spreadsheet(dummy_path,sheet_name='unidentified'),\"name\":\"PROSYN-PROTDUMMYUNIDENTIFIED\"}]\n",
    "for prot in dummy_list:\n",
    "\tprot[\"df\"].index = prot[\"df\"].aa_abbv.to_list()\n",
    "\tprot[\"medianL\"] = int(round(prot[\"df\"].loc['A', 'Unnamed: 5'], 0))\n",
    "\t# Add 1e-5 g/mmol to prevent number round lost of protein MW\n",
    "\t#dummy_MW\n",
    "\tprot[\"MW\"] = round(prot[\"df\"].loc['C', 'Unnamed: 5'], 5) + 1e-5\n",
    "\tprot[\"biosyn\"] = prot[\"name\"].replace('PROSYN-','BIOSYN-')\n",
    "\tprot[\"met\"] = prot[\"name\"].lower().replace('prosyn-','BIO-')\n",
    "\tprot[\"translation_loc\"] = 'unknown' if not 'translation_loc' in prot else prot['translation_loc']\n",
    "\n",
    "# Enzyme\n",
    "df_enz = read_spreadsheet('./input/ENZYME_stoich_curation.tsv')\n",
    "# add 'id' column to df_enz, which is 'RXN-' + rxn_src + '_' + dir + '-' + enz\n",
    "if include_special_case_proteins:\n",
    "\tdf_enz_special = read_spreadsheet('./input/ENZYME_stoich_curation_special_cases.tsv')\n",
    "\tdf_enz = pd.concat([df_enz,df_enz_special],ignore_index=True)\n",
    "\n",
    "# RNA\n",
    "df_rnas = read_spreadsheet('./input/RNA_stoich.xlsx')\n",
    "df_rnas.index = df_rnas.RNAid.to_list()\n",
    "\n",
    "# Ribosome ('_c' for nucleus or nucleoid, '_m' for mitochondria)\n",
    "ribo_dict = {'_c':read_spreadsheet('./input/RIBOSOME_nucleus.xlsx'), \n",
    "\t\t\t '_m':read_spreadsheet('./input/RIBOSOME_mitochondria.xlsx')}\n",
    "\n",
    "# Biomass\n",
    "# heading start at cell X4\n",
    "df_biom = read_spreadsheet('./input/BIOMASS_RBA.xlsx', sheet_name='RBABioRxns')\n",
    "# list of biomass rxns to turn off by default and only turn on when needed; any rxns not found will be excluded from list\n",
    "whole_biomass_rxns = ['BIOSYN-' + i for i in ['BIODILAERO','COFACTORANAEROBIC','compCERANAEROBIC','BIODILBATCHANAERO','BIODILCHEMOANAERO','BIODILSTARVE','BIODILNOGAM','BIODILAERO-NOGAM']]\n",
    "\n",
    "# checks which rxns to ignore when seeing if all necessary enzymes are modeled.\n",
    "rxns_not_needing_enzymes = [rxn.id for rxn in find_biomass_reactions(model)] + ['BIOMASS_MFA','BIOMASS_MFA_NO_GAM']\n",
    "# Gene-Protein Reaction associations (GPRs) representing spontaneous reactions (i.e., no enzyme needed in model)\n",
    "spont_GPRs = ['SPONT', 'UNKNOWN']\n",
    "# take IDs of all rxns\n",
    "print('Reactions to ignore when checking if enzymes must be added:')\n",
    "print(rxns_not_needing_enzymes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Expanding df_pro to include sublocations as separate proteins\n",
    "## remove the 'id' column\n",
    "# df_pro = df_pro.drop(columns=['id'])\n",
    "# if the 'gene_src' column is empty, fill it with the 'name' column\n",
    "df_pro['gene_src'] = df_pro['gene_src'].fillna(df_pro['name'])\n",
    "## for each protein, if there are commas in subloc_assigned, split them into separate entries with 'id' modified to reflect sublocation\n",
    "df_pro['subloc_assigned'] = df_pro['subloc_assigned'].apply(lambda x: x.split(',') if type(x) == str and ',' in x else x)\n",
    "# add the 'id' column back in, assigning it to \"gene_src\" plus \"subloc_assigned\" if the latter is not 'unknown'\n",
    "# modify IDs to reflect sublocation\n",
    "# add location only if there are multiple proteins with the same gene source\n",
    "df_pro = df_pro.explode('subloc_assigned').reset_index(drop=True)\n",
    "if 'id' not in df_pro.columns:\n",
    "\tdf_pro['id'] = df_pro.apply(\n",
    "\t\tlambda row: row['id_manual_override (clear when using new sheet)']\n",
    "\t\tif pd.notna(row['id_manual_override (clear when using new sheet)'])\n",
    "\t\telse (\n",
    "\t\t\tstr(row['gene_src'])\n",
    "\t\t\t+ (\n",
    "\t\t\t\t'_' + row['subloc_assigned']\n",
    "\t\t\t\tif row['subloc_assigned'] != 'unknown'\n",
    "\t\t\t\tand df_pro[df_pro['gene_src'] == row['gene_src']].shape[0] > 1\n",
    "\t\t\t\telse ''\n",
    "\t\t\t)\n",
    "\t\t),\n",
    "\t\taxis=1\n",
    "\t)\n",
    "df_pro.index = df_pro.id.to_list()\n",
    "# save df_pro to model folder\n",
    "df_pro[['id'] + [col for col in df_pro.columns if col != 'id']].to_csv('./model/PROTEIN_stoich_curation.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate MW of enzymes = sum of MW of all proteins in protein_stoich (MW from df_pro) * number of protein_stoich\n",
    "df_enz['id'] = df_enz.apply(lambda x: 'RXN-' + x['rxn_src'] + '_' + x['dir'] + '-' + x['enz'], axis=1)\n",
    "df_enz.index = df_enz.id.to_list()\n",
    "mw = 'MW (g/mmol)'\n",
    "for enz in df_enz.index:\n",
    "\tenz_mw = 0\n",
    "\tfor pro in df_enz.loc[enz, 'protein_stoich'].split(','):\n",
    "\t\t# find protein stoichiometry in df_pro\n",
    "\t\tdelim = ':'\n",
    "\t\tif delim in pro:\n",
    "\t\t\tprot_id = pro.split(delim)[0].strip()\n",
    "\t\t\tprot_stoich = float(pro.split(delim)[1].strip())\n",
    "\t\t\tpro_matches = df_pro.loc[df_pro.gene_src == prot_id]\n",
    "\t\t\t# if no matches, check if prot_id is in df_pro.index\n",
    "\t\t\tif len(pro_matches) == 0:\n",
    "\t\t\t\tif prot_id in df_pro.index:\n",
    "\t\t\t\t\tpro_matches = df_pro.loc[prot_id]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tprint(f'Warning: {prot_id} not found in protein stoichiometry file')\n",
    "\t\t\t\t\tcontinue\n",
    "\t\t\tpro_mw = pro_matches[mw]\n",
    "\t\t\tif isinstance(pro_matches[mw], pd.Series):\n",
    "\t\t\t\tif pro_matches[mw].nunique() > 1:\n",
    "\t\t\t\t\tprint(f'Warning: {prot_id} has multiple MWs in df_pro')\n",
    "\t\t\t\tpro_mw = pro_matches[mw].iloc[0]\n",
    "\t\t\tif float(pro_mw) == 0:\n",
    "\t\t\t\tprint(f'Warning: MW for {prot_id} is 0 in protein stoichiometry file')\n",
    "\t\t\t\tcontinue\n",
    "\t\t\tenz_mw += pro_mw * prot_stoich\n",
    "\t# set enz_mw to NaN if it is 0\n",
    "\tif enz_mw == 0:\n",
    "\t\tenz_mw = float('nan')\n",
    "\tdf_enz.loc[enz, mw] = enz_mw\n",
    "# move 'id' column to the front\n",
    "df_enz[['id'] + [col for col in df_enz.columns if col != 'id']].to_csv('./model/ENZYME_stoich_curation.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list all genes, to map them to translation (PROSYN) rxns and other rxns making their products where needed\n",
    "# combine all genes from df_pro and model\n",
    "# combine all genes from df_pro and model, ensuring all are strings and not NaN\n",
    "genes_model = [str(gene.id) for gene in model.genes if pd.notna(gene.id)]\n",
    "genes_pro = [str(g) for g in df_pro.gene_src.to_list() if pd.notna(g)]\n",
    "genes_rna = [str(r) for r in df_rnas.RNAid.to_list() if pd.notna(r)]\n",
    "genes = sorted(list(dict.fromkeys(genes_model + genes_pro + genes_rna)))\n",
    "gene_expression_dict = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "# testing only\n",
    "# # tag_c formula weight\n",
    "# # print(model.metabolites.get_by_id('tag_c').formula_weight-227.000000) # 227.000000 from Ac element accidentally listed in it\n",
    "# # TAG MW is 89.07 w/o Acyl3, 882.40 w/ Acyl3 according to https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6838544/. Thus, Acyl MW is (882.40-89.07)/3 = 264.4433333333\n",
    "# # calculate_molecular_weight('C28H43O')\n",
    "# # (12.010700*28)+(1.007940*43)+15.999400 = 395.64042\n",
    "# # print(model.metabolites.get_by_id('ergstest_c').formula_weight-227.000000+264.4433333333) # 227.000000 from Ac element accidentally listed in it\n",
    "# print(model.metabolites.get_by_id('o2_c').formula_weight)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPIA3_r\n",
      "gpail_r + pmtcoa_r <=> coa_r + gacpail_r\n",
      "coa_r\tC21H32N7O16P3S1\t-4\tC00010\tcoenzyme A [endoplasmic reticulum]\t      \n",
      "gacpail_r\t\t0\t      \t6-(alpha-D-glucosaminyl)-O-acyl-1-phosphatidyl-1D-myo-inositol [endoplasmic reticulum]\t      \n",
      "gpail_r\tC17H28NO17PR2\t0\tC04248\t6-(alpha-D-glucosaminyl)-1-phosphatidyl-1D-myo-inositol [endoplasmic reticulum]\t      \n",
      "pmtcoa_r\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA [endoplasmic reticulum]\t      \n",
      "C:-33, H:-58, N:-1, O:-18, P:-1, S:0, R:-2, charge:0\n",
      "\n",
      "GPIA8_r\n",
      "1em2gacpail_r + dolmanp_r <=> 1em3gacpail_r\n",
      "1em2gacpail_r\t\t0\t      \t6-O-{alpha-D-mannosyl-(1->6)-2-O-[(2-aminoethyl)phosphoryl]-alpha-D-mannosyl-(1->4)-alpha-D-glucosaminyl}-O-acyl-1-phosphatidyl-1D-myo-inositol [endoplasmic reticulum]\t      \n",
      "1em3gacpail_r\t\t0\t      \t6-O-{alpha-D-mannosyl-(1->2)-alpha-D-mannosyl-(1->6)-2-O-[(2-aminoethyl)phosphoryl]-alpha-D-mannosyl-(1->4)-alpha-D-glucosaminyl}-O-acyl-1-phosphatidyl-1D-myo-inositol [endoplasmic reticulum]\t      \n",
      "dolmanp_r\tC6H12O9PPolyprenyl\t-1\tC03862\tdolichyl D-mannosyl phosphate [endoplasmic reticulum]\t      \n",
      "C:-6, H:-12, O:-9, P:-1, Polyprenyl:-1, charge:1\n",
      "\n",
      "GPIA9_r\n",
      "1em3gacpail_r + pe_gen_r <=> 13em3gacpail_r + dag_r\n",
      "13em3gacpail_r\tC43H44N9O\t0\t      \t6-[6-(2-aminoethylphosphoryl)-alpha-mannosyl-(1->6)-2-(2-aminoethylphosphoryl)-alpha-mannosyl-(1->6)-alpha-2-(2-aminoethylphosphoryl)mannosyl-(1->4)-alpha-glucosaminyl]-O-acyl-1-phosphatidyl-1D-myo-inositol [endoplasmic reticulum]\t      \n",
      "1em3gacpail_r\t\t0\t      \t6-O-{alpha-D-mannosyl-(1->2)-alpha-D-mannosyl-(1->6)-2-O-[(2-aminoethyl)phosphoryl]-alpha-D-mannosyl-(1->4)-alpha-D-glucosaminyl}-O-acyl-1-phosphatidyl-1D-myo-inositol [endoplasmic reticulum]\t      \n",
      "dag_r\tC3H6O3Acyl2\t0\tC00165\tdiglyceride backbone [endoplasmic reticulum]\t      \n",
      "pe_gen_r\tC39H75NO8P\t0\tC00350\tphosphatidylethanolamine backbone [endoplasmic reticulum]\t      \n",
      "C:7, H:-25, N:8, O:-4, Acyl:2, P:-1, charge:0\n",
      "\n",
      "Q6MT2_m\n",
      "2hpmhmbq_m + amet_m + h_m --> ahcys_m + q6h2_m\n",
      "2hpmhmbq_m\tC38H56O4\t0\tC05805\t2-hexaprenyl-5-hydroxy-6-methoxy-3-methyl-1,4-benzoquinone [mitochondrion]\t      \n",
      "ahcys_m\tC14H20N6O5S1\t0\tC00021\tS-adenosyl-L-homocysteine [mitochondrion]\t      \n",
      "amet_m\tC15H23N6O5S\t1\tC00019\tS-adenosyl-L-methionine [mitochondrion]\t      \n",
      "h_m\tH1\t1\tC00080\tH+ [mitochondrion]\t      \n",
      "q6h2_m\tC39H60O4\t0\tC00390\tubiquinol-6 [mitochondrion]\t      \n",
      "C:0, H:0, O:0, N:0, S:0, charge:-2\n",
      "\n",
      "KYNIs_c\n",
      "kynate_c --> qunld_c\n",
      "kynate_c\tC10H5N1O3\t-2\tC01717\tkynurenic acid [cytoplasm]\t      \n",
      "qunld_c\tC10H7NO2\t0\tC06325\tquinaldic acid [cytoplasm]\t      \n",
      "C:0, H:2, N:0, O:-1, charge:2\n",
      "\n",
      "APOLIPT_m\n",
      "apop_m + lipACP_m <=> ACP_m + apoplip_m\n",
      "ACP_m\tAcp\t-1\tC00229\tACP1 [mitochondrion]\t      \n",
      "apop_m\tNH2R\t0\tC16240\tApoprotein [mitochondrion]\t      \n",
      "apoplip_m\tC8H14NOS2R\t0\tC16237\tProtein N6-(lipoyl)lysine [mitochondrion]\t      \n",
      "lipACP_m\tC8H13OS3R\t0\tC16239\tLipoyl-[acp] [mitochondrion]\t      \n",
      "Acp:1, N:0, H:-1, R:-1, C:0, O:0, S:-1, charge:-1\n",
      "\n",
      "r_4323_m\n",
      "2.0 amet_m + apopoct_m + 2.0 sdonor_m --> apoplip_m + 2.0 dad_5_m + 2.0 met__L_m\n",
      "amet_m\tC15H23N6O5S\t1\tC00019\tS-adenosyl-L-methionine [mitochondrion]\t      \n",
      "apoplip_m\tC8H14NOS2R\t0\tC16237\tProtein N6-(lipoyl)lysine [mitochondrion]\t      \n",
      "apopoct_m\tC8H16NOR\t0\tC16236\tProtein N6-(octanoyl)lysine [mitochondrion]\t      \n",
      "dad_5_m\tC10H13N5O3\t0\tC05198\t5'-Deoxyadenosine [mitochondrion]\t      \n",
      "met__L_m\tC5H11N1O2S1\t0\tC00073\tL-methionine [mitochondrion]\t      \n",
      "sdonor_m\t\t0\tC17023\tSulfur donor [mitochondrion]\t      \n",
      "C:0, H:0, N:0, O:0, S:2, R:0, charge:-2\n",
      "\n",
      "r_4324_m\n",
      "2.0 amet_m + ocACP_m + 2.0 sdonor_m --> 2.0 dad_5_m + lipACP_m + 2.0 met__L_m\n",
      "amet_m\tC15H23N6O5S\t1\tC00019\tS-adenosyl-L-methionine [mitochondrion]\t      \n",
      "dad_5_m\tC10H13N5O3\t0\tC05198\t5'-Deoxyadenosine [mitochondrion]\t      \n",
      "lipACP_m\tC8H13OS3R\t0\tC16239\tLipoyl-[acp] [mitochondrion]\t      \n",
      "met__L_m\tC5H11N1O2S1\t0\tC00073\tL-methionine [mitochondrion]\t      \n",
      "ocACP_m\tC8H14OAcp\t-1\tC05752\toctanoyl-ACP [mitochondrion]\t      \n",
      "sdonor_m\t\t0\tC17023\tSulfur donor [mitochondrion]\t      \n",
      "C:0, H:1, N:0, O:0, S:3, R:1, Acp:-1, charge:-1\n",
      "\n",
      "compCER_m\n",
      "0.1 cer1_24_m + 0.1 cer1_26_m + 0.1 cer2A_24_m + 0.1 cer2A_26_m + 0.1 cer2B_24_m + 0.1 cer2B_26_m + 0.1 cer3_24_m + 0.1 cer3_26_m + 0.1 cer4_24_m + 0.1 cer4_26_m <=> cer_m\n",
      "cer1_24_m\tC42H85N1O3\t0\t      \tceramide-1 (C24) [mitochondrion]\t      \n",
      "cer1_26_m\tC44H89N1O3\t0\t      \tceramide-1 (C26) [mitochondrion]\t      \n",
      "cer2A_24_m\tC42H85N1O4\t0\t      \tceramide-2 (C24) [mitochondrion]\t      \n",
      "cer2A_26_m\tC44H89N1O4\t0\t      \tceramide-2 (C26) [mitochondrion]\t      \n",
      "cer2B_24_m\tC42H85NO4\t0\t      \tceramide-2' (C24) [mitochondrion]\t      \n",
      "cer2B_26_m\tC44H89NO4\t0\t      \tceramide-2' (C26) [mitochondrion]\t      \n",
      "cer3_24_m\tC42H85NO5\t0\t      \tceramide-3 (C24) [mitochondrion]\t      \n",
      "cer3_26_m\tC44H89NO5\t0\t      \tceramide-3 (C26) [mitochondrion]\t      \n",
      "cer4_24_m\tC42H85NO6\t0\t      \tceramide-4 (C24) [mitochondrion]\t      \n",
      "cer4_26_m\tC44H89NO6\t0\t      \tceramide-4 (C26) [mitochondrion]\t      \n",
      "cer_m\tCer\t0\tC00195\tCeramide\t      \n",
      "C:-43.00000000000001, H:-87, N:-0.9999999999999999, O:-4.4, Cer:1, charge:0\n",
      "\n",
      "compCER_r\n",
      "0.1 cer1_24_r + 0.1 cer1_26_r + 0.1 cer2A_24_r + 0.1 cer2A_26_r + 0.1 cer2B_24_r + 0.1 cer2B_26_r + 0.1 cer3_24_r + 0.1 cer3_26_r + 0.1 cer4_24_r + 0.1 cer4_26_r <=> cer_r\n",
      "cer1_24_r\tC42H85N1O3\t0\t      \tceramide-1 (C24) [endoplasmic reticulum]\t      \n",
      "cer1_26_r\tC44H89N1O3\t0\t      \tceramide-1 (C26) [endoplasmic reticulum]\t      \n",
      "cer2A_24_r\tC42H85N1O4\t0\t      \tceramide-2 (C24) [endoplasmic reticulum]\t      \n",
      "cer2A_26_r\tC44H89N1O4\t0\t      \tceramide-2 (C26) [endoplasmic reticulum]\t      \n",
      "cer2B_24_r\tC42H85N1O4\t0\t      \tceramide-2' (C24) [endoplasmic reticulum]\t      \n",
      "cer2B_26_r\tC44H89N1O4\t0\t      \tceramide-2' (C26) [endoplasmic reticulum]\t      \n",
      "cer3_24_r\tC42H85N1O5\t0\t      \tceramide-3 (C24) [endoplasmic reticulum]\t      \n",
      "cer3_26_r\tC44H89N1O5\t0\t      \tceramide-3 (C26) [endoplasmic reticulum]\t      \n",
      "cer4_24_r\tC42H85N1O6\t0\t      \tceramide-4 (C24) [endoplasmic reticulum]\t      \n",
      "cer4_26_r\tC44H89N1O6\t0\t      \tceramide-4 (C26) [endoplasmic reticulum]\t      \n",
      "cer_r\tCer\t0\tC00195\tCeramide\t      \n",
      "C:-43.00000000000001, H:-87, N:-0.9999999999999999, O:-4.4, Cer:1, charge:0\n",
      "\n",
      "compCER_g\n",
      "0.1 cer1_24_g + 0.1 cer1_26_g + 0.1 cer2A_24_g + 0.1 cer2A_26_g + 0.1 cer2B_24_g + 0.1 cer2B_26_g + 0.1 cer3_24_g + 0.1 cer3_26_g + 0.1 cer4_24_g + 0.1 cer4_26_g <=> cer_g\n",
      "cer1_24_g\tC42H85N1O3\t0\t      \tceramide-1 (C24) [Golgi]\t      \n",
      "cer1_26_g\tC44H89N1O3\t0\t      \tceramide-1 (C26) [Golgi]\t      \n",
      "cer2A_24_g\tC42H85N1O4\t0\t      \tceramide-2 (C24) [Golgi]\t      \n",
      "cer2A_26_g\tC44H89N1O4\t0\t      \tceramide-2 (C26) [Golgi]\t      \n",
      "cer2B_24_g\tC42H85NO4\t0\t      \tceramide-2' (C24) [Golgi]\t      \n",
      "cer2B_26_g\tC44H89NO4\t0\t      \tceramide-2' (C26) [Golgi]\t      \n",
      "cer3_24_g\tC42H85NO5\t0\t      \tceramide-3 (C24) [Golgi]\t      \n",
      "cer3_26_g\tC44H89NO5\t0\t      \tceramide-3 (C26) [Golgi]\t      \n",
      "cer4_24_g\tC42H85NO6\t0\t      \tceramide-4 (C24) [Golgi]\t      \n",
      "cer4_26_g\tC44H89NO6\t0\t      \tceramide-4 (C26) [Golgi]\t      \n",
      "cer_g\tCer\t0\tC00195\tCeramide\t      \n",
      "C:-43.00000000000001, H:-87, N:-0.9999999999999999, O:-4.4, Cer:1, charge:0\n",
      "\n",
      "compACYLCOA_c\n",
      "0.554586 hdcoa_c + 0.203479 odecoa_c + 0.18897 pmtcoa_c + 0.052965 stcoa_c --> acylcoa_c\n",
      "acylcoa_c\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\t      \n",
      "hdcoa_c\tC37H60N7O17P3S1\t-4\tC21072\tpalmitoleoyl-CoA(4-) [cytoplasm]\t      \n",
      "odecoa_c\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA [cytoplasm]\t      \n",
      "pmtcoa_c\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA [cytoplasm]\t      \n",
      "stcoa_c\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA [cytoplasm]\t      \n",
      "C:-16.512888, H:-30.509646, N:-5.551115123125783e-17, O:-1.0000000000000009, P:-1.1102230246251565e-16, S:-1.3877787807814457e-17, Acyl:1, charge:5.551115123125783e-17\n",
      "\n",
      "compACYLCOA_l\n",
      "0.554586 hdcoa_l + 0.203479 odecoa_l + 0.18897 pmtcoa_l + 0.052965 stcoa_l --> acylcoa_l\n",
      "acylcoa_l\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\t      \n",
      "hdcoa_l\tC37H60N7O17P3S1\t-4\tC21072\tpalmitoleoyl-CoA(4-) [lipid particle]\t      \n",
      "odecoa_l\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA [lipid particle]\t      \n",
      "pmtcoa_l\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA [lipid particle]\t      \n",
      "stcoa_l\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA [lipid particle]\t      \n",
      "C:-16.512888, H:-30.509646, N:-5.551115123125783e-17, O:-1.0000000000000009, P:-1.1102230246251565e-16, S:-1.3877787807814457e-17, Acyl:1, charge:5.551115123125783e-17\n",
      "\n",
      "compACYLCOA_m\n",
      "0.554586 hdcoa_m + 0.203479 odecoa_m + 0.18897 pmtcoa_m + 0.052965 stcoa_m --> acylcoa_m\n",
      "acylcoa_m\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\t      \n",
      "hdcoa_m\tC37H60N7O17P3S\t-4\tC21072\tpalmitoleoyl-CoA(4-) [cytoplasm]\t      \n",
      "odecoa_m\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA [cytoplasm]\t      \n",
      "pmtcoa_m\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA [cytoplasm]\t      \n",
      "stcoa_m\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA [cytoplasm]\t      \n",
      "C:-16.512888, H:-30.509646, N:-5.551115123125783e-17, O:-1.0000000000000009, P:-1.1102230246251565e-16, S:-1.3877787807814457e-17, Acyl:1, charge:5.551115123125783e-17\n",
      "\n",
      "compACYLCOA_r\n",
      "0.554586 hdcoa_r + 0.203479 odecoa_r + 0.18897 pmtcoa_r + 0.052965 stcoa_r --> acylcoa_r\n",
      "acylcoa_r\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\t      \n",
      "hdcoa_r\tC37H60N7O17P3S\t-4\tC21072\tpalmitoleoyl-CoA(4-) [cytoplasm]\t      \n",
      "odecoa_r\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA [cytoplasm]\t      \n",
      "pmtcoa_r\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA [endoplasmic reticulum]\t      \n",
      "stcoa_r\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA [cytoplasm]\t      \n",
      "C:-16.512888, H:-30.509646, N:-5.551115123125783e-17, O:-1.0000000000000009, P:-1.1102230246251565e-16, S:-1.3877787807814457e-17, Acyl:1, charge:5.551115123125783e-17\n",
      "\n",
      "compACYLCOA_rm\n",
      "0.554586 hdcoa_rm + 0.203479 odecoa_rm + 0.18897 pmtcoa_rm + 0.052965 stcoa_rm --> acylcoa_rm\n",
      "acylcoa_rm\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\t      \n",
      "hdcoa_rm\tC37H60N7O17P3S1\t-4\tC21072\tpalmitoleoyl-CoA(4-) [endoplasmic reticulum membrane]\t      \n",
      "odecoa_rm\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA [endoplasmic reticulum membrane]\t      \n",
      "pmtcoa_rm\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA [endoplasmic reticulum membrane]\t      \n",
      "stcoa_rm\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA [endoplasmic reticulum membrane]\t      \n",
      "C:-16.512888, H:-30.509646, N:-5.551115123125783e-17, O:-1.0000000000000009, P:-1.1102230246251565e-16, S:-1.3877787807814457e-17, Acyl:1, charge:5.551115123125783e-17\n",
      "\n",
      "compACYLCOA_x\n",
      "0.554586 hdcoa_x + 0.203479 odecoa_x + 0.18897 pmtcoa_x + 0.052965 stcoa_x --> acylcoa_x\n",
      "acylcoa_x\tC21H31N7O16P3SAcyl\t-4\tC00040\tAcyl-CoA\t      \n",
      "hdcoa_x\tC37H60N7O17P3S1\t-4\tC21072\tpalmitoleoyl-CoA(4-) [peroxisome]\t      \n",
      "odecoa_x\tC39H64N7O17P3S1\t-4\tC00510\toleoyl-CoA [peroxisome]\t      \n",
      "pmtcoa_x\tC37H62N7O17P3S1\t-4\tC00154\tpalmitoyl-CoA [peroxisome]\t      \n",
      "stcoa_x\tC39H66N7O17P3S1\t-4\tC00412\tstearoyl-CoA [peroxisome]\t      \n",
      "C:-16.512888, H:-30.509646, N:-5.551115123125783e-17, O:-1.0000000000000009, P:-1.1102230246251565e-16, S:-1.3877787807814457e-17, Acyl:1, charge:5.551115123125783e-17\n",
      "\n",
      "compFALPD_mm\n",
      "0.18897 hdca_mm + 0.554586 hdcea_mm + 0.052965 ocdca_mm + 0.203479 ocdcea_mm <=> falpd_mm\n",
      "falpd_mm\tAcylO\t-1\tC00162\tFatty acid (from lipid)\t      \n",
      "hdca_mm\tC16H31O2\t-1\tC00249\tpalmitate [mitochondrial membrane]\t      \n",
      "hdcea_mm\tC16H29O2\t-1\tC08362\tpalmitoleate [cytoplasm]\t      \n",
      "ocdca_mm\tC18H35O2\t-1\tC01530\tstearate [mitochondrial membrane]\t      \n",
      "ocdcea_mm\tC18H33O2\t-1\tC00712\toleate [cytoplasm]\t      \n",
      "Acyl:1, O:-1, C:-16.512888, H:-30.509645999999996, charge:-2.7755575615628914e-17\n",
      "\n",
      "compFALPD_l\n",
      "0.18897 hdca_l + 0.554586 hdcea_l + 0.052965 ocdca_l + 0.203479 ocdcea_l <=> falpd_l\n",
      "falpd_l\tAcylO\t-1\tC00162\tFatty acid (from lipid)\t      \n",
      "hdca_l\tC16H31O2\t-1\tC00249\tpalmitate [lipid particle]\t      \n",
      "hdcea_l\tC16H29O2\t-1\tC08362\tpalmitoleate [lipid particle]\t      \n",
      "ocdca_l\tC18H35O2\t-1\tC01530\tstearate [lipid particle]\t      \n",
      "ocdcea_l\tC18H33O2\t-1\tC00712\toleate [lipid particle]\t      \n",
      "Acyl:1, O:-1, C:-16.512888, H:-30.509645999999996, charge:-2.7755575615628914e-17\n",
      "\n",
      "compFALPD_m\n",
      "0.18897 hdca_m + 0.554586 hdcea_m + 0.052965 ocdca_m + 0.203479 ocdcea_m <=> falpd_m\n",
      "falpd_m\tAcylO\t-1\tC00162\tFatty acid (from lipid)\t      \n",
      "hdca_m\tC16H31O2\t-1\tC00249\tpalmitate [cytoplasm]\t      \n",
      "hdcea_m\tC16H29O2\t-1\tC08362\tpalmitoleate [cytoplasm]\t      \n",
      "ocdca_m\tC18H35O2\t-1\tC01530\tstearate [cytoplasm]\t      \n",
      "ocdcea_m\tC18H33O2\t-1\tC00712\toleate [mitochondrion]\t      \n",
      "Acyl:1, O:-1, C:-16.512888, H:-30.509645999999996, charge:-2.7755575615628914e-17\n",
      "\n",
      "compFALPD_rm\n",
      "0.18897 hdca_rm + 0.554586 hdcea_rm + 0.052965 ocdca_rm + 0.203479 ocdcea_rm <=> falpd_rm\n",
      "falpd_rm\tAcylO\t-1\tC00162\tFatty acid (from lipid)\t      \n",
      "hdca_rm\tC16H31O2\t-1\tC00249\tpalmitate [endoplasmic reticulum membrane]\t      \n",
      "hdcea_rm\tC16H29O2\t-1\tC08362\tpalmitoleate [endoplasmic reticulum membrane]\t      \n",
      "ocdca_rm\tC18H35O2\t-1\tC01530\tstearate [endoplasmic reticulum membrane]\t      \n",
      "ocdcea_rm\tC18H33O2\t-1\tC00712\toleate [endoplasmic reticulum membrane]\t      \n",
      "Acyl:1, O:-1, C:-16.512888, H:-30.509645999999996, charge:-2.7755575615628914e-17\n",
      "\n",
      "compFALPD_en\n",
      "0.18897 hdca_en + 0.554586 hdcea_en + 0.052965 ocdca_en + 0.203479 ocdcea_en <=> falpd_en\n",
      "falpd_en\tAcylO\t-1\tC00162\tFatty acid (from lipid)\t      \n",
      "hdca_en\tC16H31O2\t-1\tC00249\tpalmitate [cell envelope]\t      \n",
      "hdcea_en\tC16H29O2\t-1\tC08362\tpalmitoleate [cell envelope]\t      \n",
      "ocdca_en\tC18H35O2\t-1\tC01530\tstearate [cell envelope]\t      \n",
      "ocdcea_en\tC18H33O2\t-1\tC00712\toleate [cell envelope]\t      \n",
      "Acyl:1, O:-1, C:-16.512888, H:-30.509645999999996, charge:-2.7755575615628914e-17\n",
      "\n",
      "compCERANAERO_r\n",
      "0.5 cer1_24_r + 0.5 cer1_26_r --> cer_r\n",
      "cer1_24_r\tC42H85N1O3\t0\t      \tceramide-1 (C24) [endoplasmic reticulum]\t      \n",
      "cer1_26_r\tC44H89N1O3\t0\t      \tceramide-1 (C26) [endoplasmic reticulum]\t      \n",
      "cer_r\tCer\t0\tC00195\tCeramide\t      \n",
      "C:-43, H:-87, N:-1, O:-3, Cer:1, charge:0\n",
      "\n",
      "DICAT1_c\n",
      "mg2_c --> dicat_c\n",
      "dicat_c\tNone\t2.0\t      \tGeneric divalent cation\t      \n",
      "mg2_c\tMg\t2\tC00305\tMg(2+) [cytoplasm]\t      \n",
      "Mg:-1, charge:0\n",
      "\n",
      "DICAT2_c\n",
      "zn2_c --> dicat_c\n",
      "dicat_c\tNone\t2.0\t      \tGeneric divalent cation\t      \n",
      "zn2_c\tZn\t2\tC00038\tZn(2+) [cytoplasm]\t      \n",
      "Zn:-1, charge:0\n",
      "\n",
      "DICAT3_c\n",
      "ca2_c --> dicat_c\n",
      "ca2_c\tCa\t2\tC00076\tCa(2+) [cytoplasm]\t      \n",
      "dicat_c\tNone\t2.0\t      \tGeneric divalent cation\t      \n",
      "Ca:-1, charge:0\n",
      "\n",
      "DICAT4_c\n",
      "cu2_c --> dicat_c\n",
      "cu2_c\tCu\t2\tC00070\tCu2(+) [cytoplasm]\t      \n",
      "dicat_c\tNone\t2.0\t      \tGeneric divalent cation\t      \n",
      "Cu:-1, charge:0\n",
      "\n",
      "DICAT5_c\n",
      "fe2_c --> dicat_c\n",
      "dicat_c\tNone\t2.0\t      \tGeneric divalent cation\t      \n",
      "fe2_c\tFe1\t2\tC14818\tiron(2+) [cytoplasm]\t      \n",
      "Fe:-1, charge:0\n",
      "\n",
      "DICAT6_c\n",
      "ni2_c --> dicat_c\n",
      "dicat_c\tNone\t2.0\t      \tGeneric divalent cation\t      \n",
      "ni2_c\tNi\t2.0\tC19609\tNickel\t      \n",
      "Ni:-1, charge:0\n",
      "\n",
      "DICAT7_c\n",
      "cobalt2_c --> dicat_c\n",
      "cobalt2_c\tCo\t2.0\tC00175\tCo2+\t      \n",
      "dicat_c\tNone\t2.0\t      \tGeneric divalent cation\t      \n",
      "Co:-1, charge:0\n",
      "\n",
      "DICAT8_c\n",
      "mn2_c --> dicat_c\n",
      "dicat_c\tNone\t2.0\t      \tGeneric divalent cation\t      \n",
      "mn2_c\tMn\t2\tC19610\tMn(2+) [cytoplasm]\t      \n",
      "Mn:-1, charge:0\n",
      "\n",
      "HEME1_c\n",
      "hemeA_c --> hemegen_c\n",
      "hemeA_c\tC49H54FeN4O6\t-2\tC15670\theme a [cytoplasm]\t      \n",
      "hemegen_c\tNone\tNone\t      \tGeneric heme\t      \n",
      "C:-49, H:-54, Fe:-1, N:-4, O:-6, charge:2\n",
      "\n",
      "HEME2_c\n",
      "hemeO_m --> hemegen_c\n",
      "hemeO_m\tC49H56FeN4O5\t-2\tC15672\theme o [mitochondrion]\t      \n",
      "hemegen_c\tNone\tNone\t      \tGeneric heme\t      \n",
      "C:-49, H:-56, Fe:-1, N:-4, O:-5, charge:2\n",
      "\n",
      "HEME3_c\n",
      "pheme_m --> hemegen_c\n",
      "hemegen_c\tNone\tNone\t      \tGeneric heme\t      \n",
      "pheme_m\tC34H30N4O4Fe1\t-2\tC00032\tferroheme b [mitochondrion]\t      \n",
      "C:-34, H:-30, N:-4, O:-4, Fe:-1, charge:2\n",
      "\n",
      "HEME4_c\n",
      "sheme_c --> hemegen_c\n",
      "hemegen_c\tNone\tNone\t      \tGeneric heme\t      \n",
      "sheme_c\tC42H36N4O16Fe1\t-8\tC00748\tsiroheme [cytoplasm]\t      \n",
      "C:-42, H:-36, N:-4, O:-16, Fe:-1, charge:8\n",
      "\n",
      "LIPOATE_m\n",
      "lipACP_m --> ACP_m + lipoate_m\n",
      "ACP_m\tAcp\t-1\tC00229\tACP1 [mitochondrion]\t      \n",
      "lipACP_m\tC8H13OS3R\t0\tC16239\tLipoyl-[acp] [mitochondrion]\t      \n",
      "lipoate_m\tC8H13O2S2\t-1.0\tC16241\tLipoate\t      \n",
      "Acp:1, C:0, H:0, O:1, S:-1, R:-1, charge:-2\n",
      "\n",
      "MONCAT1_c\n",
      "na1_c --> moncat_c\n",
      "moncat_c\tNone\t1.0\t      \tGeneric monovalent cation\t      \n",
      "na1_c\tNa\t1\tC01330\tsodium [cytoplasm]\t      \n",
      "Na:-1, charge:0\n",
      "\n",
      "MONCAT2_c\n",
      "k_c --> moncat_c\n",
      "k_c\tK1\t1\tC00238\tpotassium [cytoplasm]\t      \n",
      "moncat_c\tNone\t1.0\t      \tGeneric monovalent cation\t      \n",
      "K:-1, charge:0\n",
      "\n",
      "MONCAT3_c\n",
      "nh4_c --> moncat_c\n",
      "moncat_c\tNone\t1.0\t      \tGeneric monovalent cation\t      \n",
      "nh4_c\tH4N1\t1\tC01342\tammonium [cytoplasm]\t      \n",
      "H:-4, N:-1, charge:0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Make version of model with no generic cofactors, for publication and for MEMOTE testing\n",
    "model_no_generics = deepcopy(model)\n",
    "# remove rxns with names beginning with \"GENERIC_\"\n",
    "for rxn in model.reactions:\n",
    "\tif rxn.id.startswith(\"GENERIC_\"):\n",
    "\t\t# print(rxn.id)\n",
    "\t\tmodel_no_generics.reactions.get_by_id(rxn.id).remove_from_model()\n",
    "report_mass_balance(model_no_generics)\n",
    "# remove metabolites with names beginning with \"pseudometabolite \"\n",
    "for met in model.metabolites:\n",
    "\tif met.name.startswith(\"pseudometabolite \"):\n",
    "\t\t# print(met.name)\n",
    "\t\tmodel_no_generics.metabolites.get_by_id(met.id).remove_from_model()\n",
    "# write JSON version of the model w/o these additions (for future use)\n",
    "cobra.io.save_json_model(model_no_generics, \"./input/\"+modelname+\"-no-generic-cofactors.json\")\n",
    "# convert to xml file\n",
    "cobra.io.write_sbml_model(model_no_generics, \"./input/\"+modelname+\"-no-generic-cofactors.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assemble reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'rt04': True, 'rt01': False, 'rt012': False, 'rt05': True, 'rt0123': False}, {'rt04': False, 'rt01': True, 'rt012': True, 'rt05': True, 'rt0123': False}, {'rt04': False, 'rt01': True, 'rt012': False, 'rt05': True, 'rt0123': True}]\n"
     ]
    }
   ],
   "source": [
    "import itertools\n",
    "import cobra.manipulation\n",
    "\n",
    "def find_true_combinations(expression, use_knockout=False, variables:list=[]):\n",
    "    \"\"\"\n",
    "    Finds all combinations of boolean values that make the given expression true.\n",
    "\n",
    "    Args:\n",
    "        expression: The boolean expression as a string.\n",
    "        use_knockout: If True, the function will test combinations where only one variable is False, and use those to eliminate redundant combinations, saving time. Will return an error if the expression contains 'not' operators unless set to 'ignore warning'.\n",
    "        variables: A list of strings (variable names in expression) to consider. If none provided, the function will extract variable names from the expression. Added in case consistent RegEx parsing proves too unreliable.\n",
    "    Returns:\n",
    "        A list of dictionaries, where each dictionary represents a true combination.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Extract variables from the expression\n",
    "    if variables == []:\n",
    "        variables = set(re.findall(r'\\b\\w+\\b', expression))\n",
    "        if use_knockout and any([var in ['not','!'] for var in variables]):\n",
    "            raise ValueError(\"The expression contains 'not' operators, which may cause issues with the 'use_knockout' option. If you still want to use this option, set it to 'ignore warning'.\")\n",
    "        # remove 'and', 'or', 'not' from variables\n",
    "        variables = [var for var in variables if var not in ['and', 'or', 'not']]\n",
    "    # print(variables)\n",
    "\n",
    "    # Generate all possible combinations of True/False values for variables\n",
    "    combinations = itertools.product([True, False], repeat=len(variables))\n",
    "    # sort combinations from fewest to most True values\n",
    "    combinations = sorted(combinations, key=lambda x: sum(x))\n",
    "    combos_to_check = combinations.copy()\n",
    "    # print(combinations)\n",
    "    # find all combinations where only one variable is False\n",
    "    knockout_combos = []\n",
    "    essential_vars = [] # variables that must be True\n",
    "    # test if combo where all variables are True makes the expression True\n",
    "    all_true = dict(zip(variables, [True]*len(variables)))\n",
    "    expr = expression\n",
    "    for var in sorted(variables, key=len, reverse=True):\n",
    "        expr = expr.replace(var, str(all_true[var]))\n",
    "    if use_knockout != False:\n",
    "        if eval(expr):\n",
    "            # for each variable, find the combination where only that variable is False\n",
    "            for var in variables:\n",
    "                combo = [True]*len(variables)\n",
    "                combo[variables.index(var)] = False\n",
    "                knockout_combos.append(combo)\n",
    "                expr = expression\n",
    "                for var2 in sorted(variables, key=len, reverse=True):\n",
    "                    expr = expr.replace(var2, str(combo[variables.index(var2)]))\n",
    "                # add variable to essential_vars if it must be True\n",
    "                if not eval(expr):\n",
    "                    essential_vars.append(var)\n",
    "                    # remove all combinations where this variable is False from combos_to_check\n",
    "                    combos_to_check = [combo for combo in combos_to_check if combo[variables.index(var)]]\n",
    "    # if len(combinations) != len(combos_to_check):\n",
    "    #     print(f'Reduced combinations to check from {len(combinations)} to {len(combos_to_check)} by eliminating redundant combinations.')\n",
    "\n",
    "    true_combinations = []\n",
    "    for combo in combos_to_check:\n",
    "        # Create a dictionary mapping variables to their assigned values\n",
    "        values = dict(zip(variables, combo))\n",
    "        # print(values)\n",
    "        # print(combo)\n",
    "        # replace variables with their values in 'values'\n",
    "        # cobra.manipulation.knock_out_model_genes\n",
    "\n",
    "        expr = expression\n",
    "        # Replace variables with their values in the expression, starting with the longest variable names to avoid the risk of replacing substrings\n",
    "        for var in sorted(variables, key=len, reverse=True):\n",
    "            expr = expr.replace(var, str(values[var]))\n",
    "\n",
    "        # print(expression)\n",
    "        # print(expr,type(expr))\n",
    "        # Evaluate the expression\n",
    "        if eval(expr):\n",
    "            # check if the combination's True values are a subset of any previous ones; if so, don't add it\n",
    "            # print('\\t',values)\n",
    "            redundant_combos = False\n",
    "            true_vals = [var for var in values if values[var]]\n",
    "            for prev in true_combinations:\n",
    "                prev_vals = [var for var in prev if prev[var]]\n",
    "                if set(true_vals).issubset(set(prev_vals)) or set(prev_vals).issubset(set(true_vals)):\n",
    "                    redundant_combos = True\n",
    "                    # print('redundant:',true_vals)\n",
    "                    break\n",
    "            if not redundant_combos:\n",
    "                true_combinations.append(values)\n",
    "\n",
    "    return true_combinations\n",
    "\n",
    "# Example usage:\n",
    "expression = \"rt05 and ((rt01 and (rt012 or rt0123)) or rt04)\"\n",
    "true_combos = find_true_combinations(expression, use_knockout=True)\n",
    "print(true_combos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/fx/q_x2rm651s91fj2m6_tb19pxvpxbmb/T/ipykernel_5621/1784060244.py:98: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_enz_rxns = pd.concat([df_enz_rxns, pd.DataFrame(new_rows)], sort=False, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# Automatically create placeholders for enzymes based on GPRs in model.reactions\n",
    "rxns_to_skip = set(rxns_not_needing_enzymes)\n",
    "enz_rxn_dict = {}\n",
    "spont_rxn_dict = {}\n",
    "rxn_updates = {rxn.id: {} for rxn in model.reactions}\n",
    "GPR_enz_dict = {}\n",
    "try:\n",
    "\tdf_enz_rxns = read_spreadsheet('./input/ENZYME_stoich.tsv')\n",
    "\t# make set of rxns to skip, from combining the 'rxns' column cells in df_enz_rxns (to avoid reviewing existing rxns) and separating by commas. However, relying on this will prevent updates to their GPRs from being added to the model.\n",
    "\t# rxns_to_skip.update(set(df_enz_rxns.rxns.str.split(',').sum()))\n",
    "\t# make enz_rxn_dict from df_enz_rxns, with 'enz' as keys and 'rxns' and 'protein_stoich' as values\n",
    "\tfor index, row in df_enz_rxns.iterrows():\n",
    "\t\tenz_rxn_dict[row['enz']] = {'rxns': sorted(list(dict.fromkeys(row['rxns'].split(',')))),\n",
    "\t\t\t\t\t\t\t\t\t\t'protein_stoich': {prot.split(':')[0]: float(prot.split(':')[1]) for prot in row['protein_stoich'].split(',')}}\n",
    "\t\tif not pd.isna(row[mw]):\n",
    "\t\t\tenz_rxn_dict[row['enz']][mw] = row[mw]\n",
    "except:\n",
    "\tdf_enz_rxns = pd.DataFrame(columns=['enz','rxns','protein_stoich',mw])\n",
    "# add dict of protein_stoich to enz_rxn_dict, with each protein as a key and its stoichiometry as the value\n",
    "\n",
    "# find all unique GPRs in model.reactions, and their associated reactions\n",
    "for rxn in model.reactions:\n",
    "\t# print(rxn.id, rxn.gene_reaction_rule)\n",
    "\tif rxn.id in rxns_to_skip:\n",
    "\t\tcontinue\n",
    "\tif rxn.gene_reaction_rule not in spont_GPRs + ['']:\n",
    "\t\tif rxn.gene_reaction_rule in GPR_enz_dict.keys():\n",
    "\t\t\t# add rxn to corresponding enz in enz_rxn_dict\n",
    "\t\t\tenz_ids = GPR_enz_dict[rxn.gene_reaction_rule]\n",
    "\t\t\tfor enz_id in enz_ids:\n",
    "\t\t\t\tenz_rxn_dict[enz_id]['rxns'].append(rxn.id)\n",
    "\t\t\t\trxn_updates[rxn.id].setdefault('enz',enz_id)\n",
    "\t\telse:\n",
    "\t\t\tGPR_enz_dict[rxn.gene_reaction_rule] = set()\n",
    "\t\t\tenzs = sorted(find_true_combinations(rxn.gene_reaction_rule, use_knockout=True, variables=[g.id for g in rxn.genes]), key=lambda x: sorted(x.keys()))\n",
    "\t\t\tfor enz in enzs:\n",
    "\t\t\t\t# print('\\t',enz)\n",
    "\t\t\t\tenz_found = None\n",
    "\t\t\t\t# default: 1 copy of each protein is used\n",
    "\t\t\t\t# prot_stoich = ','.join([gene+':1' for gene in sorted(enz.keys()) if enz[gene]])\n",
    "\t\t\t\tprot_stoich = {gene:1 for gene in sorted(enz.keys()) if enz[gene]}\n",
    "\t\t\t\t# check if any enzymes with the same stoichiometry are already in enz_rxn_dict\n",
    "\t\t\t\tenz_id = ''.join(sorted(prot_stoich.keys()))\n",
    "\n",
    "\t\t\t\tfor old_enz_id in enz_rxn_dict:\n",
    "\t\t\t\t\t# print(enz_id,enz_rxn_dict[enz_id])\n",
    "\t\t\t\t\t# automatically assumes enzymes are the same if they have the same proteins, regardless of order/count\n",
    "\t\t\t\t\tif sorted(enz_rxn_dict[old_enz_id]['protein_stoich'].keys()) == sorted(prot_stoich.keys()):\n",
    "\t\t\t\t\t\tenz_rxn_dict[old_enz_id]['rxns'].append(rxn.id)\n",
    "\t\t\t\t\t\tenz_found = old_enz_id\n",
    "\t\t\t\t\t\tbreak\n",
    "\t\t\t\tif not enz_found:\n",
    "\t\t\t\t\t# enz_rxn_dict[rxn.id] = enz\n",
    "\t\t\t\t\tenz_rxn_dict[enz_id] = {'rxns':[rxn.id],\n",
    "\t\t\t\t\t\t\t\t\t\t\t'protein_stoich':prot_stoich}\n",
    "\t\t\t\t\trxn_updates[rxn.id].setdefault('enz',enz_id)\n",
    "\t\t\t\t\tGPR_enz_dict[rxn.gene_reaction_rule].add(enz_id)\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tGPR_enz_dict[rxn.gene_reaction_rule].add(enz_found)\n",
    "\telse:\n",
    "\t\t# add to spont_rxn_dict under its GPR as a list of rxn IDs, if it is not already there\n",
    "\t\tspont_rxn_dict[rxn.gene_reaction_rule] = spont_rxn_dict.get(rxn.gene_reaction_rule, {'rxns':[],'protein_stoich':{'zeroCost':1}})\n",
    "\t\tspont_rxn_dict[rxn.gene_reaction_rule]['rxns'].append(rxn.id)\n",
    "\t\trxn_updates[rxn.id].setdefault('gene_reaction_rule',rxn.gene_reaction_rule)\n",
    "\t\t# spont_rxn_dict[rxn.gene_reaction_rule] = spont_rxn_dict.get(rxn.gene_reaction_rule, [rxn.id])\n",
    "\t\t# spont_rxn_dict[rxn.gene_reaction_rule] = rxn.id\n",
    "\t\t\t\t\t\n",
    "# give each enz in enz_rxn_dict a row in df_enz_rxns\n",
    "new_rows = []\n",
    "for enz in enz_rxn_dict:\n",
    "\t# only add enzymes that are not already in df_enz_rxns\n",
    "\tif df_enz_rxns[df_enz_rxns.enz == enz].empty:\n",
    "\t\t# calculate MW of each enzyme based on its protein stoichiometry\n",
    "\t\tif mw in enz_rxn_dict[enz] and enz_rxn_dict[enz][mw]:\n",
    "\t\t\tenz_mw = enz_rxn_dict[enz][mw]\n",
    "\t\telse:\n",
    "\t\t\tenz_mw = 0\n",
    "\t\t\tfor prot in enz_rxn_dict[enz]['protein_stoich']:\n",
    "\t\t\t\tpro_mw = df_pro.loc[prot,mw] if prot in df_pro.index else 0\n",
    "\t\t\t\t# check in gene_src column\n",
    "\t\t\t\tif pro_mw == 0:\n",
    "\t\t\t\t\tif prot in df_pro.gene_src.to_list():\n",
    "\t\t\t\t\t\t# print(prot,df_pro.loc[df_pro.gene_src == prot,mw])\n",
    "\t\t\t\t\t\t# if there are multiple proteins with the same gene_src, check if their MW is the same\n",
    "\t\t\t\t\t\t# if not, print a warning\n",
    "\t\t\t\t\t\t# pick the first one\n",
    "\t\t\t\t\t\tpro_match = df_pro.loc[df_pro.gene_src == prot,mw].to_list()\n",
    "\t\t\t\t\t\tif df_pro.loc[df_pro.gene_src == prot,mw].nunique() > 1:\n",
    "\t\t\t\t\t\t\tprint(f'Warning: {prot} has multiple MWs in df_pro')\n",
    "\t\t\t\t\t\tpro_mw = pro_match[0]\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tprint(f'Warning: {prot} not found in df_pro')\n",
    "\t\t\t\tenz_mw += pro_mw*enz_rxn_dict[enz]['protein_stoich'][prot]\n",
    "\t\tnew_rows.append({'enz':enz,'rxns':','.join(enz_rxn_dict[enz]['rxns']),\n",
    "\t\t\t\t\t\t'protein_stoich':','.join([prot+':'+str(enz_rxn_dict[enz]['protein_stoich'][prot]) for prot in enz_rxn_dict[enz]['protein_stoich']]),\n",
    "\t\t\t\t\t\tmw:enz_mw,\n",
    "\t\t\t\t\t\t'status':'automatically added'})\n",
    "df_enz_rxns = pd.concat([df_enz_rxns, pd.DataFrame(new_rows)], sort=False, ignore_index=True)\n",
    "# sort by id, MW, and then by rxns\n",
    "df_enz_rxns = df_enz_rxns.sort_values(by=['rxns',mw,'enz'])\n",
    "# save automatically updated df_enz_rxns as separate file\n",
    "df_enz_rxns.to_csv('./model/ENZYME_stoich.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating RLFC2O_c with {'enz': 'YDL174C'}\n",
      "Updating RLFC2O_m with {'enz': 'YDL178W'}\n",
      "Updating BTDDH_c with {'enz': 'YAL060W'}\n",
      "Updating LLFC2O_c with {'enz': 'YEL039CYML054C'}\n",
      "Updating 13GS_c with {'enz': 'YGR032WYPR165W'}\n",
      "Updating 16GS_c with {'enz': 'YGR143W'}\n",
      "Updating PRMICI_c with {'enz': 'YIL020C'}\n",
      "Updating P5CD_m with {'enz': 'YHR037W'}\n",
      "Updating DKMPPD_c with {'enz': 'YEL038WYMR009W'}\n",
      "Updating DHPPDA2_c with {'enz': 'YOL066C'}\n",
      "Updating DRAPPRy_c with {'enz': 'YBR153W'}\n",
      "Updating ACHBS_m with {'enz': 'YMR108W'}\n",
      "Updating HPPK2_m with {'enz': 'YNL256W'}\n",
      "Updating AATA_c with {'enz': 'YER152C'}\n",
      "Updating DPR_c with {'enz': 'YHR063C'}\n",
      "Updating Q6MT1_m with {'enz': 'YDR204WYGL119WYGR255CYLR201CYML110CYOL096CYOR125C'}\n",
      "Updating Q6MO_m with {'enz': 'YDR204WYGL119WYGR255CYLR201CYML110CYOL096CYOR125C'}\n",
      "Updating IPPMIb_c with {'enz': 'YGL009C'}\n",
      "Updating IPPS_c with {'enz': 'YOR108W'}\n",
      "Updating UNK3_c with {'enz': 'YHR137W'}\n",
      "Updating HCITR_m with {'enz': 'YDR234W'}\n",
      "Updating MCITS_m with {'enz': 'YPR001W'}\n",
      "Updating BPNT_c with {'enz': 'YOL064C'}\n",
      "Updating PDE1_c with {'enz': 'YOR360C'}\n",
      "Updating PDE3_c with {'enz': 'YOR360C'}\n",
      "Updating PDE4_c with {'enz': 'YOR360C'}\n",
      "Updating PDE5_c with {'enz': 'YOR360C'}\n",
      "Updating DB4PS_c with {'enz': 'YDR487C'}\n",
      "Updating DHQTi_c with {'enz': 'YDR127W'}\n",
      "Updating DHQS_c with {'enz': 'YDR127W'}\n",
      "Updating 3DSPHR_r with {'enz': 'YBR265W'}\n",
      "Updating DDPA_c with {'enz': 'YDR035W'}\n",
      "Updating 3HXPHBH_c with {'enz': 'YDR376W'}\n",
      "Updating 3HPH5MBDC_m with {'enz': 'YDR539W'}\n",
      "Updating HKYNH_c with {'enz': 'YLR231C'}\n",
      "Updating HACD140i_x with {'enz': 'YKR009C'}\n",
      "Updating 3HAO_c with {'enz': 'YJR025C'}\n",
      "Updating 3IPM3MT_c with {'enz': 'YER175C'}\n",
      "Updating IPPMIa_c with {'enz': 'YGL009C'}\n",
      "Updating IPMD_c with {'enz': 'YCL018W'}\n",
      "Updating 3MOBDC_c with {'enz': 'YGR087C'}\n",
      "Updating MTMOHT_m with {'enz': 'YBR176W'}\n",
      "Updating PSCIT_c with {'enz': 'YDR127W'}\n",
      "Updating ADCS_c with {'enz': 'YNR033W'}\n",
      "Updating ADCL_c with {'enz': 'YMR289W'}\n",
      "Updating ABTA_c with {'enz': 'YGR019W'}\n",
      "Updating 4HTHRS_c with {'enz': 'YCR053W'}\n",
      "Updating 4HBZF_m with {'gene_reaction_rule': ''}\n",
      "Updating 4DPMIPPH_c with {'enz': 'YOR163W'}\n",
      "Updating IPPYP1_c with {'enz': 'YDR017C'}\n",
      "Updating MTAP_c with {'enz': 'YLR017W'}\n",
      "Updating NTD4pp_c with {'enz': 'YGL224C'}\n",
      "Updating NTD11_c with {'enz': 'YOR155C'}\n",
      "Updating NTD2pp_c with {'enz': 'YER037W'}\n",
      "Updating PRFGS_c with {'enz': 'YGR061C'}\n",
      "Updating MTHFR3_c with {'enz': 'YPL023C'}\n",
      "Updating ALAS_m with {'enz': 'YDR232W'}\n",
      "Updating PMI12346PS_c with {'enz': 'YDR017C'}\n",
      "Updating MHPGLUT_c with {'enz': 'YER091C'}\n",
      "Updating MDRPD_c with {'enz': 'YJR024C'}\n",
      "Updating MTRI_c with {'enz': 'YPR118W'}\n",
      "Updating IPPYP3_c with {'enz': 'YLR410W'}\n",
      "Updating PFK_4_c with {'enz': 'YOL136C'}\n",
      "Updating PGL_c with {'enz': 'YHR163W'}\n",
      "Updating 6DPMIPPH_c with {'enz': 'YOR163W'}\n",
      "Updating IPPYP4_c with {'enz': 'YDR017C'}\n",
      "Updating AOXS_c with {'enz': 'YAR069W-A'}\n",
      "Updating ACALDCD_c with {'enz': 'YGR087C'}\n",
      "Updating KARA1i_m with {'enz': 'YLR355C'}\n",
      "Updating ACLS_m with {'enz': 'YMR108W'}\n",
      "Updating ACOATA_m with {'enz': 'YKL192C'}\n",
      "Updating ACACT180i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT260i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT160i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT40ir_c with {'enz': 'YPL028W'}\n",
      "Updating ACACT40ir_m with {'enz': 'YPL028W'}\n",
      "Updating ACACT140i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT100i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT120i_x with {'enz': 'YIL160C'}\n",
      "Updating ACCOAC_m with {'enz': 'YMR207C'}\n",
      "Updating ACCOAC_c with {'enz': 'YDL141WYNR016C'}\n",
      "Updating ACOAHi_m with {'enz': 'YBL015W'}\n",
      "Updating ACS_c with {'enz': 'YLR153C'}\n",
      "Updating ACGK_m with {'enz': 'YER069W'}\n",
      "Updating ACP1_e with {'enz': 'YHR215W'}\n",
      "Updating ACONT2_m with {'enz': 'YPR002W'}\n",
      "Updating ACOTAi_m with {'enz': 'YOL140W'}\n",
      "Updating COAHL_m with {'enz': 'YPL148C'}\n",
      "Updating ACOAO100_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO120_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO260_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO160_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO180_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO140_x with {'enz': 'YGL205W'}\n",
      "Updating STATe181_rm with {'enz': 'YCR048W'}\n",
      "Updating STATg181_rm with {'enz': 'YNR019W'}\n",
      "Updating STATf181_rm with {'enz': 'YCR048W'}\n",
      "Updating STATl181_rm with {'enz': 'YCR048W'}\n",
      "Updating STATz181_rm with {'enz': 'YCR048W'}\n",
      "Updating STATe161_rm with {'enz': 'YCR048W'}\n",
      "Updating STATg161_rm with {'enz': 'YNR019W'}\n",
      "Updating STATf161_rm with {'enz': 'YCR048W'}\n",
      "Updating STATl161_rm with {'enz': 'YCR048W'}\n",
      "Updating STATz161_rm with {'enz': 'YCR048W'}\n",
      "Updating GPIA3_r with {'enz': 'YJL091C'}\n",
      "Updating ADD_c with {'enz': 'YNL141W'}\n",
      "Updating ADPT_c with {'enz': 'YML022W'}\n",
      "Updating ADA_c with {'enz': 'YNL141W'}\n",
      "Updating ADNK1_c with {'enz': 'YJR105W'}\n",
      "Updating AMPDA_c with {'enz': 'YML035C'}\n",
      "Updating AHCi_c with {'enz': 'YER043C'}\n",
      "Updating ADMDC_c with {'enz': 'YOL052C'}\n",
      "Updating AMAOTr_c with {'enz': 'YNR058W'}\n",
      "Updating ADNCYC_c with {'enz': 'YJL005W'}\n",
      "Updating ADK1_c with {'enz': 'YDL166C'}\n",
      "Updating ADK1_m with {'enz': 'YER170W'}\n",
      "Updating ADK3_m with {'enz': 'YER170W'}\n",
      "Updating ADSL2i_c with {'enz': 'YLR359W'}\n",
      "Updating ADSL1r_c with {'enz': 'YLR359W'}\n",
      "Updating ADSS_c with {'enz': 'YNL220W'}\n",
      "Updating ADSK_c with {'enz': 'YKL001C'}\n",
      "Updating 23CAPPD_c with {'enz': 'YGR247W'}\n",
      "Updating AGTi_c with {'enz': 'YFL030W'}\n",
      "Updating ALATRS_c with {'enz': 'YOR335C'}\n",
      "Updating OHACT3_c with {'enz': 'YOR377W'}\n",
      "Updating OHACT1_c with {'enz': 'YGR177C'}\n",
      "Updating OHACT2_c with {'enz': 'YGR177C'}\n",
      "Updating OHACT4_c with {'enz': 'YGR177C'}\n",
      "Updating OHACT5_c with {'enz': 'YGR177C'}\n",
      "Updating ALCD2i1_c with {'enz': 'YMR303C'}\n",
      "Updating ALCD19y_c with {'enz': 'YDR368W'}\n",
      "Updating ALCD2i2_m with {'enz': 'YGL256W'}\n",
      "Updating ALCD22xi_c with {'enz': 'YBR145W'}\n",
      "Updating ALCD22xi_m with {'enz': 'YMR083W'}\n",
      "Updating ALCD22yi_c with {'enz': 'YCR105W'}\n",
      "Updating ALCD25xi_c with {'enz': 'YBR145W'}\n",
      "Updating ALCD25xi_m with {'enz': 'YMR083W'}\n",
      "Updating ALDD22x_c with {'enz': 'YMR110C'}\n",
      "Updating ALDD2y_c with {'enz': 'YPL061W'}\n",
      "Updating ALDD2x_m with {'enz': 'YOR374W'}\n",
      "Updating ALDD2y_m with {'enz': 'YER073W'}\n",
      "Updating ALDD20x_m with {'enz': 'YOR374W'}\n",
      "Updating ALDD20y_c with {'enz': 'YPL061W'}\n",
      "Updating ALCD24xi_c with {'enz': 'YBR145W'}\n",
      "Updating ALCD24xi_m with {'enz': 'YMR083W'}\n",
      "Updating ALCD24yi_c with {'enz': 'YCR105W'}\n",
      "Updating ALCD23xi_c with {'enz': 'YBR145W'}\n",
      "Updating ALCD23xi_m with {'enz': 'YMR083W'}\n",
      "Updating ALCD23yi_c with {'enz': 'YCR105W'}\n",
      "Updating ALCD26xi_c with {'enz': 'YBR145W'}\n",
      "Updating ALCD26xi_m with {'enz': 'YMR083W'}\n",
      "Updating AKP1_c with {'enz': 'YDR481C'}\n",
      "Updating ATAH_c with {'enz': 'YIR029W'}\n",
      "Updating ALLTN_c with {'enz': 'YIR027C'}\n",
      "Updating ALPHNH_c with {'enz': 'YBR208C'}\n",
      "Updating 12AMANTF_g with {'enz': 'YBR205W'}\n",
      "Updating TREH_v with {'enz': 'YPR026W'}\n",
      "Updating TREH_c with {'enz': 'YDR001C'}\n",
      "Updating TRE6PS_c with {'enz': 'YBR126CYDR074WYML100W'}\n",
      "Updating MALT_c with {'enz': 'YBR299W'}\n",
      "Updating AMID_c with {'enz': 'YDR242W'}\n",
      "Updating AACTOOR_c with {'gene_reaction_rule': ''}\n",
      "Updating ABUTD_m with {'enz': 'YOR374W'}\n",
      "Updating ANPRT_c with {'enz': 'YDR354W'}\n",
      "Updating ANS_c with {'enz': 'YER090WYKL211C'}\n",
      "Updating AP4AH_c with {'enz': 'YDR305C'}\n",
      "Updating ARGN_c with {'enz': 'YPL111W'}\n",
      "Updating ARGSL_c with {'enz': 'YHR018C'}\n",
      "Updating ARGSS_c with {'enz': 'YOL058W'}\n",
      "Updating ARGTRS_c with {'enz': 'YDR341C'}\n",
      "Updating ARGTRS_m with {'enz': 'YHR091C'}\n",
      "Updating ASNS1_c with {'enz': 'YGR124W'}\n",
      "Updating ASNTRS_c with {'enz': 'YHR019C'}\n",
      "Updating ASNTRS_m with {'enz': 'YCR024C'}\n",
      "Updating ASPCT_c with {'enz': 'YJL130C'}\n",
      "Updating ASPK_c with {'enz': 'YER052C'}\n",
      "Updating ASPTA_c with {'enz': 'YLR027C'}\n",
      "Updating ASPTAi_m with {'enz': 'YKL106W'}\n",
      "Updating ASPTA_x with {'enz': 'YLR027C'}\n",
      "Updating ASAD_c with {'enz': 'YDR158W'}\n",
      "Updating ASPTRS_c with {'enz': 'YLL018C'}\n",
      "Updating ASPTRS_m with {'enz': 'YPL104W'}\n",
      "Updating ATPATF1_c with {'enz': 'YDR530C'}\n",
      "Updating ATPATF2_c with {'enz': 'YCL050C'}\n",
      "Updating ATPATF3_c with {'enz': 'YCL050C'}\n",
      "Updating ATPPRT_c with {'enz': 'YER055C'}\n",
      "Updating ATPS_m with {'enz': 'Q0080Q0085Q0130YBL099WYBR039WYDL004WYDR298CYDR322C-AYDR377WYJR121WYKL016CYLR295CYML081C-AYPL078CYPL271W'}\n",
      "Updating ATPASEP2e_c with {'enz': 'YCR024C-AYEL017C-AYPL036W'}\n",
      "Updating 14BMANTF_c with {'enz': 'YBR110W'}\n",
      "Updating BTS1_c with {'enz': 'YGR286C'}\n",
      "Updating C14STR_c with {'enz': 'YNL280C'}\n",
      "Updating C22STDSy_c with {'enz': 'YMR015C'}\n",
      "Updating C3STDH2_c with {'enz': 'YGL001C'}\n",
      "Updating C3STDH1_c with {'enz': 'YGL001C'}\n",
      "Updating C3STKR1_c with {'enz': 'YLR100W'}\n",
      "Updating C3STKR2_c with {'enz': 'YLR100W'}\n",
      "Updating C4STMO1_c with {'enz': 'YGR060W'}\n",
      "Updating C4STMO2_c with {'enz': 'YGR060W'}\n",
      "Updating C4STMO3_c with {'enz': 'YGR060W'}\n",
      "Updating C4STMO4_c with {'enz': 'YGR060W'}\n",
      "Updating C5STDS_c with {'enz': 'YLR056W'}\n",
      "Updating C8STI_c with {'enz': 'YMR202W'}\n",
      "Updating C24STR_r with {'enz': 'YGL012W'}\n",
      "Updating FRTT4HBZ_c with {'enz': 'YDL090CYKL019W'}\n",
      "Updating CBPS_c with {'enz': 'YJL130CYJR109CYOR303W'}\n",
      "Updating CRNOAT_c with {'enz': 'YER024W'}\n",
      "Updating CRNOAT_x with {'enz': 'YML042W'}\n",
      "Updating CRNOAT_m with {'enz': 'YML042W'}\n",
      "Updating CAT_c with {'enz': 'YGR088W'}\n",
      "Updating CAT_x with {'enz': 'YDR256C'}\n",
      "Updating CERH124A_r with {'enz': 'YDR297W'}\n",
      "Updating CERH124B_r with {'enz': 'YMR272C'}\n",
      "Updating CERH126A_r with {'enz': 'YDR297W'}\n",
      "Updating CERH126B_r with {'enz': 'YMR272C'}\n",
      "Updating CERS124_r with {'enz': 'YHL003CYMR298W'}\n",
      "Updating CERS126_r with {'enz': 'YKL008CYMR298W'}\n",
      "Updating CERS2A24_r with {'enz': 'YKL008CYMR298W'}\n",
      "Updating CERS2A26_r with {'enz': 'YKL008CYMR298W'}\n",
      "Updating CERH2A24_r with {'enz': 'YMR272C'}\n",
      "Updating CERH2A26_r with {'enz': 'YMR272C'}\n",
      "Updating CERH324_r with {'enz': 'YMR272C'}\n",
      "Updating CERH326_r with {'enz': 'YMR272C'}\n",
      "Updating CHTNDA_c with {'enz': 'YLR307W'}\n",
      "Updating CHTNS_c with {'enz': 'YNL192W'}\n",
      "Updating CHOLK_c with {'enz': 'YLR133W'}\n",
      "Updating CHLPCTD_c with {'enz': 'YGR202C'}\n",
      "Updating CHORM_c with {'enz': 'YPR060C'}\n",
      "Updating CHORS_c with {'enz': 'YGL148W'}\n",
      "Updating ACONTb_m with {'enz': 'YLR304C'}\n",
      "Updating FRTT_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating GGTT_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PPTT_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT6_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT7_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT8_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT9_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT10_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT11_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT12_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT13_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT14_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT15_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT16_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT17_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT18_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT19_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT20_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating PRENT21_l with {'enz': 'YDL193WYMR101C'}\n",
      "Updating CS_m with {'enz': 'YNR001C'}\n",
      "Updating CS_x with {'enz': 'YCR005C'}\n",
      "Updating ACONTa_m with {'enz': 'YLR304C'}\n",
      "Updating ACONTa_c with {'enz': 'YLR304C'}\n",
      "Updating CPPPGO_c with {'enz': 'YDR044W'}\n",
      "Updating CTPS2_c with {'enz': 'YJR103W'}\n",
      "Updating CBL_c with {'enz': 'YFR055W'}\n",
      "Updating CYSTS_c with {'enz': 'YGR155W'}\n",
      "Updating CYSTGL_c with {'enz': 'YAL012W'}\n",
      "Updating METB1_c with {'enz': 'YJR130C'}\n",
      "Updating CYSS_c with {'enz': 'YLR303W'}\n",
      "Updating CYSTRS_c with {'enz': 'YNL247W'}\n",
      "Updating CYTD_c with {'enz': 'YLR245C'}\n",
      "Updating CYTDK2_c with {'enz': 'YNR012W'}\n",
      "Updating LNS14DMy_c with {'enz': 'YHR007CYHR042W'}\n",
      "Updating CSND_c with {'enz': 'YPR062W'}\n",
      "Updating ARAB14LO_c with {'enz': 'YML086C'}\n",
      "Updating ARAB1Dx_c with {'enz': 'YMR041C'}\n",
      "Updating ARAB1D2_c with {'enz': 'YBR149W'}\n",
      "Updating FBA2_c with {'enz': 'YKL060C'}\n",
      "Updating SBTD_D2_c with {'enz': 'YDL246C'}\n",
      "Updating DCMPDA_c with {'enz': 'YHR144C'}\n",
      "Updating DCTPD_c with {'enz': 'YHR144C'}\n",
      "Updating DADA_c with {'enz': 'YNL141W'}\n",
      "Updating DCYTD_c with {'enz': 'YLR245C'}\n",
      "Updating DGK1_c with {'enz': 'YDR454C'}\n",
      "Updating DXHPS_c with {'enz': 'YHR068W'}\n",
      "Updating DRBK_c with {'enz': 'YCR036W'}\n",
      "Updating DPCOAK_m with {'enz': 'YDR196C'}\n",
      "Updating DBTS_c with {'enz': 'YNR057C'}\n",
      "Updating DHCRD124_r with {'enz': 'YPL087W'}\n",
      "Updating DHCRD126_r with {'enz': 'YPL087W'}\n",
      "Updating DHCRD2A24_r with {'enz': 'YPL087W'}\n",
      "Updating DHCRD2A26_r with {'enz': 'YPL087W'}\n",
      "Updating DHFRi_c with {'enz': 'YOR236W'}\n",
      "Updating DHFRi_m with {'enz': 'YOR236W'}\n",
      "Updating DHFS_1_c with {'enz': 'YMR113W'}\n",
      "Updating DHNPA2i_m with {'enz': 'YNL256W'}\n",
      "Updating DNMPPA_c with {'enz': 'YDL100C'}\n",
      "Updating DHORTS_c with {'enz': 'YLR420W'}\n",
      "Updating DHPTS_m with {'enz': 'YNL256W'}\n",
      "Updating FOLD3_m with {'enz': 'YNL256W'}\n",
      "Updating DHAD1_m with {'enz': 'YJR016C'}\n",
      "Updating DHAD2_m with {'enz': 'YJR016C'}\n",
      "Updating DHAK_c with {'enz': 'YML070W'}\n",
      "Updating DMATT_c with {'enz': 'YJL167W'}\n",
      "Updating DPGM_c with {'enz': 'YKL152C'}\n",
      "Updating PMI1346PH_c with {'enz': 'YOR163W'}\n",
      "Updating PMI1346PS_c with {'enz': 'YDR017C'}\n",
      "Updating DIPS_c with {'enz': 'YLR172C'}\n",
      "Updating DOLK_c with {'enz': 'YMR013C'}\n",
      "Updating DOLPMT_c with {'enz': 'YPR183W'}\n",
      "Updating DOLPMMT_r with {'enz': 'YJR143C'}\n",
      "Updating DTMPK_c with {'enz': 'YJR057W'}\n",
      "Updating DUTPDP_c with {'enz': 'YBR252W'}\n",
      "Updating EPGALURS_e with {'enz': 'YJR153W'}\n",
      "Updating ENO_c with {'enz': 'YMR323W'}\n",
      "Updating ETHAK_c with {'enz': 'YDR147W'}\n",
      "Updating ACHLE3_c with {'enz': 'YOR126C'}\n",
      "Updating 13BGH_e with {'enz': 'YGR282C'}\n",
      "Updating FRTT_c with {'enz': 'YPL069C'}\n",
      "Updating FACOAL100_x with {'enz': 'YER015W'}\n",
      "Updating FACOAL120_x with {'enz': 'YER015W'}\n",
      "Updating FACOAL160_x with {'enz': 'YER015W'}\n",
      "Updating FACOAL80_x with {'enz': 'YER015W'}\n",
      "Updating FACOAL140_x with {'enz': 'YER015W'}\n",
      "Updating FCLT_m with {'enz': 'YOR176W'}\n",
      "Updating FECPOR_m with {'enz': 'YEL039CYKR066C'}\n",
      "Updating FECOOR_m with {'enz': 'Q0045Q0250Q0275YDL067CYDR231CYGR062CYHR116WYJL003WYLL018C-AYPL132W'}\n",
      "Updating FECRq6_m with {'enz': 'Q0105YBL045CYDR529CYEL024WYFR033CYGR183CYHR001W-AYJL166WYJR048WYOR065WYPR191W'}\n",
      "Updating AFAT_c with {'enz': 'YDL045C'}\n",
      "Updating FMNRx_c with {'enz': 'YLR011W'}\n",
      "Updating FMNRx2_c with {'enz': 'YLR011W'}\n",
      "Updating FDH_c with {'enz': 'YOR388C'}\n",
      "Updating FTHFL_c with {'enz': 'YGR204W'}\n",
      "Updating FTHFL_m with {'enz': 'YBR084W'}\n",
      "Updating FBP2_c with {'enz': 'YJL155C'}\n",
      "Updating FBP_c with {'enz': 'YLR377C'}\n",
      "Updating FBA_c with {'enz': 'YKL060C'}\n",
      "Updating FUM_m with {'enz': 'YPL262W'}\n",
      "Updating FUM_c with {'enz': 'YPL262W'}\n",
      "Updating DHORDfum_c with {'enz': 'YKL216W'}\n",
      "Updating FRD_m with {'enz': 'YJR051W'}\n",
      "Updating FRD_c with {'enz': 'YEL047C'}\n",
      "Updating GTMLT_c with {'enz': 'YLR299W'}\n",
      "Updating GALKi_c with {'enz': 'YBR020W'}\n",
      "Updating GLUCYS_c with {'enz': 'YJL101C'}\n",
      "Updating GGTT_c with {'gene_reaction_rule': ''}\n",
      "Updating GRTT_c with {'enz': 'YJL167W'}\n",
      "Updating GLCGSD_c with {'enz': 'YPR184W'}\n",
      "Updating GLCGSD_v with {'enz': 'YIL099W'}\n",
      "Updating G6PDA_c with {'gene_reaction_rule': ''}\n",
      "Updating G6PDH2i_c with {'enz': 'YNL241C'}\n",
      "Updating PGI_c with {'enz': 'YBR196C'}\n",
      "Updating GLU5K_c with {'enz': 'YDR300C'}\n",
      "Updating GLUDC_c with {'enz': 'YMR250W'}\n",
      "Updating GLUDxi_c with {'enz': 'YDL215C'}\n",
      "Updating GLUDy_c with {'enz': 'YAL062W'}\n",
      "Updating GLUSx_c with {'enz': 'YDL171C'}\n",
      "Updating G5SDy_c with {'enz': 'YOR323C'}\n",
      "Updating GLUN_c with {'enz': 'YFL060C'}\n",
      "Updating GLNS_c with {'enz': 'YPR035W'}\n",
      "Updating GF6PTA_c with {'enz': 'YMR084W'}\n",
      "Updating GLNTRS_c with {'enz': 'YOR168W'}\n",
      "Updating GLUTRS_c with {'enz': 'YGL105WYGL245W'}\n",
      "Updating GLUTRS_m with {'enz': 'YOL033W'}\n",
      "Updating GTHO_c with {'enz': 'YCL035CYPL091W'}\n",
      "Updating GTHO_m with {'enz': 'YPL059WYPL091W'}\n",
      "Updating GTHP_c with {'enz': 'YPL091W'}\n",
      "Updating GTHP_m with {'enz': 'YBL064C'}\n",
      "Updating GTHS_c with {'enz': 'YOL049W'}\n",
      "Updating GAPD_c with {'enz': 'YJR009C'}\n",
      "Updating GLYCDy_c with {'enz': 'YOR120W'}\n",
      "Updating GLYK_c with {'enz': 'YHL032C'}\n",
      "Updating G3PT_c with {'enz': 'YER062C'}\n",
      "Updating G3PDf_m with {'enz': 'YIL155C'}\n",
      "Updating G3PD1r_c with {'enz': 'YDL022W'}\n",
      "Updating GPDDA1pp_c with {'enz': 'YPL110C'}\n",
      "Updating GARFT_c with {'enz': 'YDR408C'}\n",
      "Updating GLYAT_c with {'gene_reaction_rule': ''}\n",
      "Updating GLYCL_m with {'enz': 'YAL044CYDR019CYFL018CYMR189W'}\n",
      "Updating GHMT2r_c with {'enz': 'YLR058C'}\n",
      "Updating GHMT2r_m with {'enz': 'YBR263W'}\n",
      "Updating GLYGS_c with {'enz': 'YFR015CYJL137C'}\n",
      "Updating GLCP_c with {'enz': 'YPR160W'}\n",
      "Updating GLYTRS_c with {'enz': 'YPR081C'}\n",
      "Updating GMPS2_c with {'enz': 'YMR217W'}\n",
      "Updating GPIA2_r with {'enz': 'YMR281W'}\n",
      "Updating GPIA5_r with {'enz': 'YCL052CYJR013W'}\n",
      "Updating GPIA6_r with {'enz': 'YKL165C'}\n",
      "Updating GPIA7_r with {'enz': 'YBR004CYNL158W'}\n",
      "Updating GPIA8_r with {'enz': 'YGL142C'}\n",
      "Updating GPIA9_r with {'enz': 'YLL031C'}\n",
      "Updating GTPCI_c with {'enz': 'YGR267C'}\n",
      "Updating GTPCII2_c with {'enz': 'YBL033C'}\n",
      "Updating GUAD_1_c with {'enz': 'YDL238C'}\n",
      "Updating GUAPRT_c with {'enz': 'YDR399W'}\n",
      "Updating GK1_c with {'enz': 'YDR454C'}\n",
      "Updating GK2_c with {'enz': 'YDR454C'}\n",
      "Updating HOMOX_m with {'enz': 'YDR376WYER141WYPL252C'}\n",
      "Updating HEMEOS_m with {'enz': 'YPL172C'}\n",
      "Updating Q6MT2_m with {'enz': 'YDR204WYGL119WYGR255CYLR201CYML110CYOL096CYOR125C'}\n",
      "Updating HEX7_c with {'enz': 'YGL253W'}\n",
      "Updating HEX1_c with {'enz': 'YCL040W'}\n",
      "Updating HEX4_c with {'enz': 'YGL253W'}\n",
      "Updating HISTD_c with {'enz': 'YCL030C'}\n",
      "Updating HISTP_c with {'enz': 'YFR025C'}\n",
      "Updating HSTPT_c with {'enz': 'YIL116W'}\n",
      "Updating HISTRS_c with {'enz': 'YPR033C'}\n",
      "Updating HISTRS_m with {'enz': 'YPR033C'}\n",
      "Updating ARMT_c with {'enz': 'YBR034C'}\n",
      "Updating HACNH_m with {'enz': 'YJL200C'}\n",
      "Updating HCYSMT_c with {'enz': 'YLL062C'}\n",
      "Updating HICITD_m with {'enz': 'YIL094C'}\n",
      "Updating HSDx_c with {'enz': 'YJR139C'}\n",
      "Updating HSDy_c with {'enz': 'YJR139C'}\n",
      "Updating HSK_c with {'enz': 'YHR025W'}\n",
      "Updating HSERTA_c with {'enz': 'YNL277W'}\n",
      "Updating THIORDXi_c with {'enz': 'YDR453CYLR043C'}\n",
      "Updating THIORDXi_x with {'enz': 'YLR043CYLR109W'}\n",
      "Updating GLYOX_c with {'enz': 'YDR272W'}\n",
      "Updating GLYOX_m with {'enz': 'YOR040W'}\n",
      "Updating HBZOPT6_m with {'enz': 'YNR041C'}\n",
      "Updating HETZK_c with {'enz': 'YPL214C'}\n",
      "Updating PPBNGD_c with {'enz': 'YDL205C'}\n",
      "Updating HMGCOAR_c with {'enz': 'YLR450W'}\n",
      "Updating HMGCOAS_c with {'enz': 'YML126C'}\n",
      "Updating HMGCOAS_m with {'enz': 'YML126C'}\n",
      "Updating HMPK1_c with {'enz': 'YOL055C'}\n",
      "Updating HXPRT_c with {'enz': 'YDR399W'}\n",
      "Updating IG3PS_c with {'enz': 'YBR248C'}\n",
      "Updating IGPDH_c with {'enz': 'YOR202W'}\n",
      "Updating IMPD_c with {'enz': 'YLR432W'}\n",
      "Updating IGPS_c with {'enz': 'YKL211C'}\n",
      "Updating INDPYRD_c with {'enz': 'YGR087C'}\n",
      "Updating PPA_c with {'enz': 'YBR011C'}\n",
      "Updating PPA_m with {'enz': 'YMR267W'}\n",
      "Updating IMPC_c with {'enz': 'YMR120C'}\n",
      "Updating IPK1_n with {'enz': 'YDR315C'}\n",
      "Updating IPK2_n with {'enz': 'YDR173C'}\n",
      "Updating IPK3_n with {'enz': 'YDR173C'}\n",
      "Updating IPK4_n with {'enz': 'YDR173C'}\n",
      "Updating IPK5_n with {'enz': 'YDR173C'}\n",
      "Updating ACHLE1_c with {'enz': 'YOR126C'}\n",
      "Updating ACHLE2_c with {'enz': 'YOR126C'}\n",
      "Updating ICDHx_m with {'enz': 'YNL037CYOR136W'}\n",
      "Updating ICDHy_c with {'enz': 'YLR174W'}\n",
      "Updating ICDHy_x with {'enz': 'YNL009W'}\n",
      "Updating ICL_1_c with {'enz': 'YER065C'}\n",
      "Updating ILETA_c with {'enz': 'YJR148W'}\n",
      "Updating ILETA_m with {'enz': 'YHR208W'}\n",
      "Updating ILETRS_c with {'enz': 'YBL076C'}\n",
      "Updating ILETRS_m with {'enz': 'YPL040C'}\n",
      "Updating IPDDI_c with {'enz': 'YPL117C'}\n",
      "Updating ITCOAL_m with {'enz': 'YGR244CYOR142W'}\n",
      "Updating KARA2i_m with {'enz': 'YLR355C'}\n",
      "Updating KYN_c with {'enz': 'YLR231C'}\n",
      "Updating KYN3OX_c with {'enz': 'YBL098W'}\n",
      "Updating PHCD_m with {'enz': 'YHR037W'}\n",
      "Updating 4HGLSD_m with {'enz': 'YHR037W'}\n",
      "Updating ALATA_L_m with {'enz': 'YLR089C'}\n",
      "Updating THRA2_c with {'enz': 'YEL046C'}\n",
      "Updating ATHRDHr_c with {'enz': 'YMR226C'}\n",
      "Updating AASADy_c with {'enz': 'YBR115CYGL154C'}\n",
      "Updating ASNN_c with {'enz': 'YDR321W'}\n",
      "Updating ASNN_e with {'enz': 'YLR158C'}\n",
      "Updating EGHLAT_c with {'enz': 'YLR027C'}\n",
      "Updating EGHLAT_m with {'enz': 'YKL106W'}\n",
      "Updating EGHLAT_x with {'enz': 'YLR027C'}\n",
      "Updating HPRORy_c with {'enz': 'YER023W'}\n",
      "Updating LALDO3_c with {'enz': 'YOL151W'}\n",
      "Updating SERD_L_c with {'enz': 'YCL064C'}\n",
      "Updating LSERDHi_c with {'enz': 'YMR226C'}\n",
      "Updating THRD_L_c with {'enz': 'YCL064C'}\n",
      "Updating THRD_L_m with {'enz': 'YCL064C'}\n",
      "Updating TRPO2_c with {'enz': 'YJR078W'}\n",
      "Updating TYRNFT_c with {'enz': 'YDR403W'}\n",
      "Updating LGTHL_c with {'enz': 'YML004C'}\n",
      "Updating LNSTLS_c with {'enz': 'YHR072W'}\n",
      "Updating LEUTA_c with {'enz': 'YJR148W'}\n",
      "Updating LEUTA_m with {'enz': 'YHR208W'}\n",
      "Updating LEUTRS_c with {'enz': 'YPL160W'}\n",
      "Updating LEUTRS_m with {'enz': 'YLR382C'}\n",
      "Updating LTA4H_c with {'enz': 'YNL045W'}\n",
      "Updating LTA4H_n with {'enz': 'YNL045W'}\n",
      "Updating LTA4H2_c with {'enz': 'YNL045W'}\n",
      "Updating LTA4H2_n with {'enz': 'YNL045W'}\n",
      "Updating LTA4H3_c with {'enz': 'YNL045W'}\n",
      "Updating LTA4H3_n with {'enz': 'YNL045W'}\n",
      "Updating LYSTRS_c with {'enz': 'YDR037W'}\n",
      "Updating LYSTRS_m with {'enz': 'YNL073W'}\n",
      "Updating MDH_m with {'enz': 'YKL085W'}\n",
      "Updating MDH_c with {'enz': 'YOL126C'}\n",
      "Updating MDH_x with {'enz': 'YDL078C'}\n",
      "Updating MALS_c with {'enz': 'YNL117W'}\n",
      "Updating ME1_m with {'enz': 'YKL029C'}\n",
      "Updating ME2_m with {'enz': 'YKL029C'}\n",
      "Updating MCOATA_m with {'enz': 'YKL192CYOR221C'}\n",
      "Updating MAN1PT_c with {'enz': 'YDL055C'}\n",
      "Updating MAN6PI_c with {'enz': 'YER003C'}\n",
      "Updating MTHFC_m with {'enz': 'YBR084W'}\n",
      "Updating MTHFC_c with {'enz': 'YGR204W'}\n",
      "Updating METAT_c with {'enz': 'YDR502C'}\n",
      "Updating METS_c with {'enz': 'YER091C'}\n",
      "Updating FMETTRS_m with {'enz': 'YBL013W'}\n",
      "Updating METTRS_c with {'enz': 'YGL105WYGR264C'}\n",
      "Updating METTRS_m with {'enz': 'YGR171C'}\n",
      "Updating MTHFD2i_c with {'enz': 'YKR080W'}\n",
      "Updating MTHFD_c with {'enz': 'YGR204W'}\n",
      "Updating MTHFD_m with {'enz': 'YBR084W'}\n",
      "Updating MCITL2_m with {'enz': 'YPR006C'}\n",
      "Updating MEVK1_c with {'enz': 'YMR208W'}\n",
      "Updating MEVK2_c with {'enz': 'YMR208W'}\n",
      "Updating MEVK3_c with {'enz': 'YMR208W'}\n",
      "Updating MEVK4_c with {'enz': 'YMR208W'}\n",
      "Updating DPMVD_c with {'enz': 'YNR043W'}\n",
      "Updating MI1PP_c with {'enz': 'YHR046C'}\n",
      "Updating M1PS_c with {'enz': 'YJL153C'}\n",
      "Updating AGPRi_m with {'enz': 'YER069W'}\n",
      "Updating ACGAM6PS_c with {'enz': 'YFL017C'}\n",
      "Updating ACGS_m with {'enz': 'YJL071W'}\n",
      "Updating FKYNH_c with {'enz': 'YDR428C'}\n",
      "Updating NFTYROX_c with {'enz': 'YDR402CYDR403W'}\n",
      "Updating NADDP_x with {'enz': 'YGL067W'}\n",
      "Updating NADK_c with {'enz': 'YEL041W'}\n",
      "Updating NADK_m with {'enz': 'YPL188W'}\n",
      "Updating NADN_n with {'enz': 'YPL015C'}\n",
      "Updating NADS2_c with {'enz': 'YHR074W'}\n",
      "Updating NADS2_n with {'enz': 'YHR074W'}\n",
      "Updating NADHq6_c with {'enz': 'YMR145C'}\n",
      "Updating NADHK1_c with {'enz': 'YEL041W'}\n",
      "Updating NADHK1_m with {'enz': 'YJR049C'}\n",
      "Updating NADHq6_m with {'enz': 'YML120C'}\n",
      "Updating NAMNPP_c with {'enz': 'YOR209C'}\n",
      "Updating NAMNPP_m with {'enz': 'YOR209C'}\n",
      "Updating NNAM_c with {'enz': 'YGL037C'}\n",
      "Updating NNMT_c with {'enz': 'YLR285W'}\n",
      "Updating ANNAT_c with {'enz': 'YCL047C'}\n",
      "Updating ANNAT_n with {'enz': 'YGR010W'}\n",
      "Updating NNATi_c with {'enz': 'YLR328W'}\n",
      "Updating NNDPR_c with {'enz': 'YFR047C'}\n",
      "Updating NNDPR_m with {'enz': 'YFR047C'}\n",
      "Updating NDP1_c with {'enz': 'YER005W'}\n",
      "Updating NDP3_c with {'enz': 'YER005W'}\n",
      "Updating NDP7_c with {'enz': 'YER005W'}\n",
      "Updating NDP7_g with {'enz': 'YEL042W'}\n",
      "Updating NDP5_c with {'enz': 'YER005W'}\n",
      "Updating NDP10_c with {'enz': 'YER005W'}\n",
      "Updating NDPK3_c with {'enz': 'YKL067W'}\n",
      "Updating NDPK8_c with {'enz': 'YKL067W'}\n",
      "Updating NDPK7_c with {'enz': 'YKL067W'}\n",
      "Updating NDPK5_c with {'enz': 'YKL067W'}\n",
      "Updating NDPK4_c with {'enz': 'YKL067W'}\n",
      "Updating NDPK1_c with {'enz': 'YKL067W'}\n",
      "Updating NDPK9_c with {'enz': 'YKL067W'}\n",
      "Updating NDPK10_c with {'enz': 'YKL067W'}\n",
      "Updating NDPK6_c with {'enz': 'YKL067W'}\n",
      "Updating NTP3_c with {'enz': 'YAL035W'}\n",
      "Updating NTP7_c with {'enz': 'YER005W'}\n",
      "Updating NTP5_c with {'enz': 'YER005W'}\n",
      "Updating NTP10_c with {'enz': 'YER005W'}\n",
      "Updating NDP3_g with {'enz': 'YEL042W'}\n",
      "Updating NDPK2_c with {'enz': 'YKL067W'}\n",
      "Updating AHSERL_c with {'enz': 'YLR303W'}\n",
      "Updating AHSERL2_c with {'enz': 'YLR303W'}\n",
      "Updating SHSL1_c with {'enz': 'YLL058W'}\n",
      "Updating OCBT_c with {'enz': 'YJL088W'}\n",
      "Updating ORNDC_c with {'enz': 'YKL184W'}\n",
      "Updating ORNTA_c with {'enz': 'YLR438W'}\n",
      "Updating ORPT_c with {'enz': 'YMR271C'}\n",
      "Updating OMPDC_c with {'enz': 'YEL021W'}\n",
      "Updating PTPAT_c with {'enz': 'YGR277C'}\n",
      "Updating PNTK_c with {'enz': 'YDR531W'}\n",
      "Updating PBAL_c with {'enz': 'YIL145C'}\n",
      "Updating FACOAE100_x with {'enz': 'YJR019C'}\n",
      "Updating FACOAE120_x with {'enz': 'YJR019C'}\n",
      "Updating FACOAE160_x with {'enz': 'YJR019C'}\n",
      "Updating FACOAE140_x with {'enz': 'YJR019C'}\n",
      "Updating FACOAE180_x with {'enz': 'YJR019C'}\n",
      "Updating FACOAE80_x with {'enz': 'YJR019C'}\n",
      "Updating PHETRS_c with {'enz': 'YFL022CYLR060W'}\n",
      "Updating PHETRS_m with {'enz': 'YPR047W'}\n",
      "Updating PPYRDC_c with {'enz': 'YDR380W'}\n",
      "Updating PRAIS_c with {'enz': 'YGL234W'}\n",
      "Updating ACGAMPM_c with {'enz': 'YEL058W'}\n",
      "Updating PAPSR_c with {'enz': 'YLR043CYPR167C'}\n",
      "Updating PPCK_c with {'enz': 'YKR097W'}\n",
      "Updating PETHCT_c with {'enz': 'YGR007W'}\n",
      "Updating PFK_c with {'enz': 'YMR205C'}\n",
      "Updating PGMT_c with {'enz': 'YMR105C'}\n",
      "Updating GND_c with {'enz': 'YHR183W'}\n",
      "Updating PGAMT_c with {'enz': 'YEL058W'}\n",
      "Updating PGCD_c with {'enz': 'YER081W'}\n",
      "Updating PGK_c with {'enz': 'YCR012W'}\n",
      "Updating PGM_c with {'enz': 'YOR283W'}\n",
      "Updating PMANM_c with {'enz': 'YFL045C'}\n",
      "Updating PMPK_c with {'enz': 'YPR121W'}\n",
      "Updating PMEVK_c with {'enz': 'YMR220W'}\n",
      "Updating PPNCL2_c with {'enz': 'YIL083C'}\n",
      "Updating PPCDC_c with {'enz': 'YKL088WYOR054C'}\n",
      "Updating PPM_c with {'enz': 'YMR278W'}\n",
      "Updating PRASCSi_c with {'enz': 'YAR015W'}\n",
      "Updating PRAMPC_c with {'enz': 'YCL030C'}\n",
      "Updating PRATPP_c with {'enz': 'YCL030C'}\n",
      "Updating AIRC1_c with {'enz': 'YOR128C'}\n",
      "Updating AICART_c with {'enz': 'YLR028C'}\n",
      "Updating PRAIi_c with {'enz': 'YDR007W'}\n",
      "Updating PRAGSi_c with {'enz': 'YGL234W'}\n",
      "Updating GLUPRT_c with {'enz': 'YMR300C'}\n",
      "Updating PRPPS_c with {'enz': 'YBL068WYKL181W'}\n",
      "Updating PSP_L_c with {'enz': 'YGR208W'}\n",
      "Updating PSERT_c with {'enz': 'YOR184W'}\n",
      "Updating CERSFA2A24_r with {'enz': 'YBR183W'}\n",
      "Updating PSPHPL_r with {'enz': 'YDR294C'}\n",
      "Updating PSPHS_r with {'enz': 'YDR297W'}\n",
      "Updating POLYAO1_c with {'enz': 'YMR020W'}\n",
      "Updating PPBNGS_1_c with {'enz': 'YGL040C'}\n",
      "Updating POLYAO2_c with {'enz': 'YMR020W'}\n",
      "Updating POLYAO3_c with {'enz': 'YMR020W'}\n",
      "Updating PPNDH_c with {'enz': 'YNL316C'}\n",
      "Updating PPND2_c with {'enz': 'YBR166C'}\n",
      "Updating PROD2_m with {'enz': 'YLR142W'}\n",
      "Updating PROTRS_c with {'enz': 'YHR020W'}\n",
      "Updating PPPGO_m with {'enz': 'YER014W'}\n",
      "Updating PNP_c with {'enz': 'YLR209C'}\n",
      "Updating PUNP3_c with {'enz': 'YLR209C'}\n",
      "Updating PUNP3_m with {'enz': 'YLR209C'}\n",
      "Updating PUNP5_c with {'enz': 'YLR209C'}\n",
      "Updating PYDXO_c with {'enz': 'YBR035C'}\n",
      "Updating PYAM5PO_c with {'enz': 'YBR035C'}\n",
      "Updating PDX5POi_c with {'enz': 'YBR035C'}\n",
      "Updating PYDXNO_c with {'enz': 'YBR035C'}\n",
      "Updating P5CR_c with {'enz': 'YER023W'}\n",
      "Updating PC_c with {'enz': 'YBR218C'}\n",
      "Updating PYRDC_c with {'enz': 'YGR087C'}\n",
      "Updating PYRDC2_c with {'enz': 'YGR087C'}\n",
      "Updating PDH_m with {'enz': 'YBR221CYER178WYFL018CYGR193CYNL071W'}\n",
      "Updating PYK_c with {'enz': 'YAL038W'}\n",
      "Updating Q6OR_m with {'enz': 'YDR204WYGL119WYGR255CYLR201CYML110CYOL096CYOR125C'}\n",
      "Updating RBFK_c with {'enz': 'YDR236C'}\n",
      "Updating RBFK_m with {'enz': 'YDR236C'}\n",
      "Updating RBFSa_c with {'enz': 'YOL143C'}\n",
      "Updating RBFSb_c with {'enz': 'YBR256C'}\n",
      "Updating RBK_c with {'enz': 'YCR036W'}\n",
      "Updating RNDR1_c with {'enz': 'YER070WYGR180CYJL026W'}\n",
      "Updating RNDR1_n with {'enz': 'YER070WYGR180CYJL026W'}\n",
      "Updating RNDR3_c with {'enz': 'YER070WYGR180CYJL026W'}\n",
      "Updating RNDR3_n with {'enz': 'YER070WYGR180CYJL026W'}\n",
      "Updating RNDR4_c with {'enz': 'YER070WYGR180CYJL026W'}\n",
      "Updating RNDR4_n with {'enz': 'YER070WYGR180CYJL026W'}\n",
      "Updating RNDR2_c with {'enz': 'YER070WYGR180CYJL026W'}\n",
      "Updating RNDR2_n with {'enz': 'YER070WYGR180CYJL026W'}\n",
      "Updating RPI_c with {'enz': 'YOR095C'}\n",
      "Updating RNMK_c with {'enz': 'YNL129W'}\n",
      "Updating RPE_c with {'enz': 'YJL121C'}\n",
      "Updating 3DH5HPBMT_m with {'enz': 'YDR204WYGL119WYGR255CYLR201CYML110CYOL096CYOR125C'}\n",
      "Updating SAM24MT_c with {'enz': 'YML008C'}\n",
      "Updating SFGTHi_c with {'enz': 'YJL068C'}\n",
      "Updating SACCD2_c with {'enz': 'YIR034C'}\n",
      "Updating SACCD1_c with {'enz': 'YNR050C'}\n",
      "Updating FBA3_c with {'enz': 'YKL060C'}\n",
      "Updating SERAT_c with {'enz': 'YDL040CYHR013CYOR253W'}\n",
      "Updating SERPT_r with {'enz': 'YDR062WYMR296C'}\n",
      "Updating SERTRS_c with {'enz': 'YHR011W'}\n",
      "Updating SHK3Di_c with {'enz': 'YDR127W'}\n",
      "Updating SHKK_c with {'enz': 'YDR127W'}\n",
      "Updating SHCHD2_c with {'enz': 'YBR213W'}\n",
      "Updating SHCHF_c with {'enz': 'YBR213W'}\n",
      "Updating SPMS_c with {'enz': 'YPR069C'}\n",
      "Updating SPRMS_c with {'enz': 'YLR146C'}\n",
      "Updating SPHPL_r with {'enz': 'YDR294C'}\n",
      "Updating SBPP2_r with {'enz': 'YJL134W'}\n",
      "Updating SBPP1_r with {'enz': 'YJL134W'}\n",
      "Updating SPHGLpK_r with {'enz': 'YLR260W'}\n",
      "Updating SLCBK1_r with {'enz': 'YOR171C'}\n",
      "Updating ACONIs_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ACONIs_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating SQLEx_r with {'enz': 'YGR175C'}\n",
      "Updating SQLEy_r with {'enz': 'YGR175C'}\n",
      "Updating SQLS_c with {'enz': 'YHR190W'}\n",
      "Updating SUCDq6_m with {'enz': 'YDR178WYJL045WYKL141WYLL041C'}\n",
      "Updating SUCOAS_m with {'enz': 'YGR244CYOR142W'}\n",
      "Updating SSALy_c with {'enz': 'YBR006W'}\n",
      "Updating SUCR_e with {'enz': 'YIL162W'}\n",
      "Updating SADT_c with {'enz': 'YJR010W'}\n",
      "Updating SLFAT_c with {'enz': 'YCL050C'}\n",
      "Updating SO3R_c with {'enz': 'YFR030WYJR137C'}\n",
      "Updating TAUDO_c with {'enz': 'YLL057C'}\n",
      "Updating THFGLUS_c with {'enz': 'YOR241W'}\n",
      "Updating THMP_e with {'enz': 'YBR092C'}\n",
      "Updating TMDPK_c with {'enz': 'YOR143C'}\n",
      "Updating TMDPPK_c with {'enz': 'YOR143C'}\n",
      "Updating TMPPP_c with {'enz': 'YPL214C'}\n",
      "Updating THIORDXi_n with {'enz': 'YIL010WYLR043C'}\n",
      "Updating TRDR_c with {'enz': 'YDR353W'}\n",
      "Updating TRDR_m with {'enz': 'YCR083WYHR106W'}\n",
      "Updating THRA_c with {'enz': 'YEL046C'}\n",
      "Updating THRS_c with {'enz': 'YCR053W'}\n",
      "Updating THRTRS_c with {'enz': 'YIL078W'}\n",
      "Updating THRTRS_m with {'enz': 'YKL194C'}\n",
      "Updating TMDS_c with {'enz': 'YOR074C'}\n",
      "Updating ACONMT_c with {'enz': 'YER175C'}\n",
      "Updating PPTT_m with {'enz': 'YBR003W'}\n",
      "Updating TALA_c with {'enz': 'YGR043C'}\n",
      "Updating TKT1_c with {'enz': 'YPR074C'}\n",
      "Updating TKT2_c with {'enz': 'YPR074C'}\n",
      "Updating TRE6PP_c with {'enz': 'YBR126CYDR074WYML100W'}\n",
      "Updating TPI_c with {'enz': 'YDR050C'}\n",
      "Updating TRPS1_c with {'enz': 'YGL026C'}\n",
      "Updating TRPTRS_c with {'enz': 'YOL097C'}\n",
      "Updating TRPTRS_m with {'enz': 'YDR268W'}\n",
      "Updating TYRTA_c with {'enz': 'YGL202W'}\n",
      "Updating TYRTAi_x with {'enz': 'YLR027C'}\n",
      "Updating TYRTRS_c with {'enz': 'YGR185C'}\n",
      "Updating TYRTRS_m with {'enz': 'YPL097W'}\n",
      "Updating ERGSTGLCT_c with {'enz': 'YLR189C'}\n",
      "Updating UAGDP_c with {'enz': 'YDL103C'}\n",
      "Updating UDPG4E_c with {'enz': 'YBR019C'}\n",
      "Updating UGLT_c with {'enz': 'YBR018C'}\n",
      "Updating UMPK_c with {'enz': 'YKL024C'}\n",
      "Updating UMPK_n with {'enz': 'YKL024C'}\n",
      "Updating UPPRT_c with {'enz': 'YHR128W'}\n",
      "Updating UREASE_c with {'enz': 'YBR208C'}\n",
      "Updating ULGYCH_c with {'enz': 'YIR032C'}\n",
      "Updating URIK1_c with {'enz': 'YDR020C'}\n",
      "Updating URIK2_c with {'enz': 'YNR012W'}\n",
      "Updating URIDK2i_c with {'enz': 'YKL024C'}\n",
      "Updating URIDK2_n with {'enz': 'YKL024C'}\n",
      "Updating UPPDC1_c with {'enz': 'YDR047W'}\n",
      "Updating UPP3MT_2_c with {'enz': 'YKR069W'}\n",
      "Updating UPP3S_c with {'enz': 'YOR278W'}\n",
      "Updating GALUi_c with {'enz': 'YKL035W'}\n",
      "Updating VALTA_c with {'enz': 'YJR148W'}\n",
      "Updating VALTA_m with {'enz': 'YHR208W'}\n",
      "Updating VALTRS_c with {'enz': 'YGR094W'}\n",
      "Updating VALTRS_m with {'enz': 'YGR094W'}\n",
      "Updating XPPT_c with {'enz': 'YJR133W'}\n",
      "Updating XYLTD_D_c with {'enz': 'YLR070C'}\n",
      "Updating XYLR_c with {'enz': 'YHR104W'}\n",
      "Updating XYLK_c with {'enz': 'YGR194C'}\n",
      "Updating YUMPS_c with {'enz': 'YNL292W'}\n",
      "Updating HCITS_m with {'enz': 'YDL131W'}\n",
      "Updating PYDXNK_c with {'enz': 'YEL029C'}\n",
      "Updating KYNIs_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating QUILSYN_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating SBP_c with {'enz': 'YKR043C'}\n",
      "Updating ICDHyi_m with {'enz': 'YDL066W'}\n",
      "Updating lumpFACS160_c with {'enz': 'YKL182WYPL231W'}\n",
      "Updating lumpFACS180_c with {'enz': 'YKL182WYPL231W'}\n",
      "Updating KAS14_m with {'enz': 'YER061CYKL192C'}\n",
      "Updating 3OAS60_m with {'enz': 'YER061CYKL192C'}\n",
      "Updating 3OAS80_m with {'enz': 'YER061CYKL192C'}\n",
      "Updating 3OAR40_m with {'enz': 'YKL055C'}\n",
      "Updating 3OAR60_m with {'enz': 'YKL055C'}\n",
      "Updating 3OAR80_m with {'enz': 'YKL055C'}\n",
      "Updating 3HAD40_m with {'enz': 'YHR067W'}\n",
      "Updating 3HAD60_m with {'enz': 'YHR067W'}\n",
      "Updating 3HAD80_m with {'enz': 'YHR067W'}\n",
      "Updating EAR40y_m with {'enz': 'YBR026C'}\n",
      "Updating EAR60y_m with {'enz': 'YBR026C'}\n",
      "Updating EAR80y_m with {'enz': 'YBR026C'}\n",
      "Updating 3OACE200_rm with {'enz': 'YCR034W'}\n",
      "Updating 3OACE220_rm with {'enz': 'YCR034W'}\n",
      "Updating 3OACE240_rm with {'enz': 'YCR034W'}\n",
      "Updating 3OACR200_rm with {'enz': 'YBR159W'}\n",
      "Updating 3OACR220_rm with {'enz': 'YBR159W'}\n",
      "Updating 3OACR240_rm with {'enz': 'YBR159W'}\n",
      "Updating 3OACR260_rm with {'enz': 'YBR159W'}\n",
      "Updating 3HACD200_rm with {'enz': 'YJL097W'}\n",
      "Updating 3HACD220_rm with {'enz': 'YJL097W'}\n",
      "Updating 3HACD240_rm with {'enz': 'YJL097W'}\n",
      "Updating 3HACD260_rm with {'enz': 'YJL097W'}\n",
      "Updating ECOAR200_rm with {'enz': 'YDL015C'}\n",
      "Updating ECOAR220_rm with {'enz': 'YDL015C'}\n",
      "Updating ECOAR240_rm with {'enz': 'YDL015C'}\n",
      "Updating ECOAR260_rm with {'enz': 'YDL015C'}\n",
      "Updating ACOADS160_rm with {'enz': 'YGL055W'}\n",
      "Updating ACOADS180_rm with {'enz': 'YGL055W'}\n",
      "Updating FACOAL120_rm with {'enz': 'YOR317W'}\n",
      "Updating FACOAL140_rm with {'enz': 'YMR246W'}\n",
      "Updating FACOAL160_rm with {'enz': 'YOR317W'}\n",
      "Updating FACOAL161_rm with {'enz': 'YOR317W'}\n",
      "Updating FACOAL180_rm with {'enz': 'YOR317W'}\n",
      "Updating FACOAL181_rm with {'enz': 'YOR317W'}\n",
      "Updating FACOAL120_l with {'enz': 'YOR317W'}\n",
      "Updating FACOAL140_l with {'enz': 'YOR317W'}\n",
      "Updating FACOAL160_l with {'enz': 'YOR317W'}\n",
      "Updating FACOAL161_l with {'enz': 'YOR317W'}\n",
      "Updating FACOAL180_l with {'enz': 'YOR317W'}\n",
      "Updating FACOAL181_l with {'enz': 'YOR317W'}\n",
      "Updating FACOAL161_x with {'enz': 'YER015W'}\n",
      "Updating FACOAL180_x with {'enz': 'YER015W'}\n",
      "Updating FACOAL181_x with {'enz': 'YER015W'}\n",
      "Updating FACOAL200_en with {'enz': 'YIL009W'}\n",
      "Updating FACOAL220_en with {'enz': 'YIL009W'}\n",
      "Updating FACOAL240_en with {'enz': 'YIL009W'}\n",
      "Updating FACOAL260_en with {'enz': 'YIL009W'}\n",
      "Updating FACOAL220_l with {'enz': 'YBR041W'}\n",
      "Updating FACOAL240_l with {'enz': 'YBR041W'}\n",
      "Updating FACOAL260_l with {'enz': 'YBR041W'}\n",
      "Updating FACOAE40_x with {'enz': 'YJR019C'}\n",
      "Updating FACOAE60_x with {'enz': 'YJR019C'}\n",
      "Updating FACOAE161_x with {'enz': 'YJR019C'}\n",
      "Updating FACOAE181_x with {'enz': 'YJR019C'}\n",
      "Updating ACOAO40_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO60_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO80_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO200_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO220_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO240_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO161b_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO141b_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO121a_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO181a_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO161a_x with {'enz': 'YGL205W'}\n",
      "Updating ACOAO141a_x with {'enz': 'YGL205W'}\n",
      "Updating ECOAH100_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH120_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH140_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH160_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH180_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH260_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH40_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH60_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH80_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH200_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH220_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH240_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH161b_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH141b_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH121a_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH181a_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH161c_x with {'enz': 'YKR009C'}\n",
      "Updating ECOAH141c_x with {'enz': 'YKR009C'}\n",
      "Updating HACD100i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD120i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD160i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD180i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD260i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD40i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD60i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD80i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD200i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD220i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD240i_x with {'enz': 'YKR009C'}\n",
      "Updating HACD161bi_x with {'enz': 'YKR009C'}\n",
      "Updating HACD141bi_x with {'enz': 'YKR009C'}\n",
      "Updating HACD121ai_x with {'enz': 'YKR009C'}\n",
      "Updating HACD181ai_x with {'enz': 'YKR009C'}\n",
      "Updating HACD161ai_x with {'enz': 'YKR009C'}\n",
      "Updating HACD141ai_x with {'enz': 'YKR009C'}\n",
      "Updating ACACT40i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT60i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT80i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT200i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT220i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT240i_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT161bi_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT141bi_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT121ai_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT181ai_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT161ai_x with {'enz': 'YIL160C'}\n",
      "Updating ACACT141ai_x with {'enz': 'YIL160C'}\n",
      "Updating ECOAI101a_x with {'enz': 'YLR284C'}\n",
      "Updating ECOAI122c_x with {'enz': 'YLR284C'}\n",
      "Updating ECOAI121b_x with {'enz': 'YLR284C'}\n",
      "Updating ECOAI121a_x with {'enz': 'YLR284C'}\n",
      "Updating ECOAI142c_x with {'enz': 'YLR284C'}\n",
      "Updating ECOAI141a_x with {'enz': 'YLR284C'}\n",
      "Updating ECOAI5d_x with {'enz': 'YOR180C'}\n",
      "Updating ECOAI6d_x with {'enz': 'YOR180C'}\n",
      "Updating ECOAR122_x with {'enz': 'YNL202W'}\n",
      "Updating ECOAR142_x with {'enz': 'YNL202W'}\n",
      "Updating ACONTb_c with {'enz': 'YLR304C'}\n",
      "Updating STEHg161_l with {'enz': 'YKL140W'}\n",
      "Updating STEHg181_l with {'enz': 'YKL140W'}\n",
      "Updating STEHe181_l with {'enz': 'YLL012W'}\n",
      "Updating STEHf161_l with {'enz': 'YLL012W'}\n",
      "Updating STEHf181_l with {'enz': 'YLL012W'}\n",
      "Updating STEHl161_l with {'enz': 'YLL012W'}\n",
      "Updating STEHl181_l with {'enz': 'YLL012W'}\n",
      "Updating STEHz161_l with {'enz': 'YLL012W'}\n",
      "Updating STEHz181_l with {'enz': 'YLL012W'}\n",
      "Updating STEHz161_en with {'enz': 'YLR020C'}\n",
      "Updating STEHz181_en with {'enz': 'YLR020C'}\n",
      "Updating SUCACCT_m with {'enz': 'YBL015W'}\n",
      "Updating RAFH_e with {'enz': 'YIL020C'}\n",
      "Updating URIH_c with {'enz': 'YDR400W'}\n",
      "Updating GTHRDtabc_v with {'enz': 'YDR135C'}\n",
      "Updating TCHOLAtabc_v with {'enz': 'YLL048C'}\n",
      "Updating ATPASEVg_c with {'enz': 'YBR127CYDL185WYEL027WYEL051WYGR020CYHR026WYHR039C-AYKL080WYLR447CYMR054WYOR332WYPL234CYPR036W'}\n",
      "Updating ATPASEVv_c with {'enz': 'YBR127CYDL185WYEL027WYEL051WYGR020CYHR026WYHR039C-AYKL080WYLR447CYOR270CYOR332WYPL234CYPR036W'}\n",
      "Updating MEVt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating BTDt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 2MBACt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 2OXOADPt_c_m with {'enz': 'YPL134C'}\n",
      "Updating 3DH5HPBt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 4ABUTtps_e with {'enz': 'YDL210W'}\n",
      "Updating 5FTHFt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 8AONNtps_e with {'enz': 'YNR056C'}\n",
      "Updating ACt_c_e with {'enz': 'YCR010C'}\n",
      "Updating ETACt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ADEtps_e with {'enz': 'YER060W'}\n",
      "Updating ADNtps_e with {'enz': 'YAL022C'}\n",
      "Updating ADPATPt_c_m with {'enz': 'YMR056C'}\n",
      "Updating ADPATPt_c_x with {'enz': 'YPR128C'}\n",
      "Updating AKGCITta_m with {'enz': 'YMR241W'}\n",
      "Updating ALLTTt_c_e with {'enz': 'YJR152W'}\n",
      "Updating ALLTNt_c_e with {'enz': 'YIR028W'}\n",
      "Updating NH4t_c_e with {'enz': 'YGR121C'}\n",
      "Updating AMPATPt_c_x with {'enz': 'YPR128C'}\n",
      "Updating ASPGLUt_c_m with {'enz': 'YPR021C'}\n",
      "Updating BTNtps_e with {'enz': 'YGR065C'}\n",
      "Updating CRNACRNt_c_m with {'enz': 'YOR100C'}\n",
      "Updating CER2B24_g_r with {'gene_reaction_rule': ''}\n",
      "Updating CER2B26_g_r with {'gene_reaction_rule': ''}\n",
      "Updating CHOLtps_e with {'enz': 'YGL077C'}\n",
      "Updating CITMALta_m with {'enz': 'YBR291C'}\n",
      "Updating CITPEPt_c_m with {'enz': 'YBR291C'}\n",
      "Updating CITICITt_c_m with {'enz': 'YBR291C'}\n",
      "Updating COAt_c_m with {'enz': 'YHR002W'}\n",
      "Updating CTPt2_m with {'enz': 'YBR192W'}\n",
      "Updating CTPCMPt_c_m with {'enz': 'YBR192W'}\n",
      "Updating CYTDtps_e with {'enz': 'YAL022C'}\n",
      "Updating CSNtps_e with {'enz': 'YER060W-A'}\n",
      "Updating FRUtps_e with {'enz': 'YMR011W'}\n",
      "Updating GALtps_e with {'enz': 'YLR081W'}\n",
      "Updating LACDtps_e with {'enz': 'YKL217W'}\n",
      "Updating LACtps_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating LACPYRt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating MANtps_e with {'enz': 'YJL214W'}\n",
      "Updating EPISTt_c_e with {'enz': 'YIL013C'}\n",
      "Updating ERGSTt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ETHAt_c_e with {'enz': 'YGL077C'}\n",
      "Updating FADH2t_c_m with {'enz': 'YIL134W'}\n",
      "Updating FECOSTt_e with {'enz': 'YOR011W'}\n",
      "Updating FECOSTt_en with {'enz': 'YOR011W'}\n",
      "Updating FOLt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GDPMANt_c_g with {'enz': 'YGL225W'}\n",
      "Updating GLCt_c_e with {'enz': 'YDR387C'}\n",
      "Updating GTHRDt_c_e with {'enz': 'YJL212C'}\n",
      "Updating GTHRDt_c_v with {'enz': 'YLL015W'}\n",
      "Updating G3PIt_c_e with {'enz': 'YCR098C'}\n",
      "Updating G3PCt_c_e with {'enz': 'YCR098C'}\n",
      "Updating GLYCtps_e with {'enz': 'YGL084C'}\n",
      "Updating GLYCt_e with {'enz': 'YFL054C'}\n",
      "Updating GLYtps_e with {'enz': 'YPL265W'}\n",
      "Updating GLYCOGENt_c_v with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GTPGDPtp_m with {'enz': 'YDL198C'}\n",
      "Updating GUAtps_e with {'enz': 'YER060W-A'}\n",
      "Updating INOSTtps_e with {'enz': 'YOL103W'}\n",
      "Updating FE2t_c_e with {'enz': 'YMR058W'}\n",
      "Updating FE2t_c_m with {'enz': 'YJL133W'}\n",
      "Updating IAMACt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating IBUTACt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating IPDPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ALAtps_e with {'enz': 'YDR046C'}\n",
      "Updating ARGtps_e with {'enz': 'YNL270C'}\n",
      "Updating ARGtpa_v with {'enz': 'YBR293W'}\n",
      "Updating ASNtps_e with {'enz': 'YDR508C'}\n",
      "Updating ASNtpa_v with {'enz': 'YJR001W'}\n",
      "Updating ASNtps_v with {'enz': 'YNL101W'}\n",
      "Updating ASPtps_v with {'enz': 'YER119C'}\n",
      "Updating ASPtps_e with {'enz': 'YFL055W'}\n",
      "Updating CRNt_e with {'enz': 'YBR132C'}\n",
      "Updating CYStps_e with {'enz': 'YBR069C'}\n",
      "Updating CYSItps_v with {'enz': 'YCR075C'}\n",
      "Updating GLUt_c_m with {'enz': 'YBR104W'}\n",
      "Updating GLUtps_v with {'enz': 'YER119C'}\n",
      "Updating GLNtpa_v with {'enz': 'YJR001W'}\n",
      "Updating GLNtps_v with {'enz': 'YNL101W'}\n",
      "Updating GLNtps_e with {'enz': 'YKR039W'}\n",
      "Updating HIStpa_v with {'enz': 'YMR088C'}\n",
      "Updating HIStps_e with {'enz': 'YGR191W'}\n",
      "Updating ILEt_c_v with {'enz': 'YCL038C'}\n",
      "Updating ILEtpa_v with {'enz': 'YJR001W'}\n",
      "Updating ILEtps_v with {'enz': 'YNL101W'}\n",
      "Updating LACt_c_m with {'enz': 'YKL217W'}\n",
      "Updating LACLtps_e with {'enz': 'YKL217W'}\n",
      "Updating LEUt_c_v with {'enz': 'YCL038C'}\n",
      "Updating LEUtpa_v with {'enz': 'YJR001W'}\n",
      "Updating LEUtps_v with {'enz': 'YNL101W'}\n",
      "Updating LYStpa_v with {'enz': 'YMR088C'}\n",
      "Updating LYStps_e with {'enz': 'YNL268W'}\n",
      "Updating METtps_e with {'enz': 'YGR055W'}\n",
      "Updating SERtps_e with {'enz': 'YDR105C'}\n",
      "Updating TYRt_c_v with {'enz': 'YCL038C'}\n",
      "Updating TYRtps_v with {'enz': 'YNL101W'}\n",
      "Updating VALtps_e with {'enz': 'YBR069C'}\n",
      "Updating LANOSTt_c_e with {'enz': 'YOR011W'}\n",
      "Updating MALt2_m with {'enz': 'YLR348C'}\n",
      "Updating MALTtps_e with {'enz': 'YGR289C'}\n",
      "Updating NFORTYR2t_c_e with {'enz': 'YBR180W'}\n",
      "Updating NADt_c_m with {'enz': 'YEL006W'}\n",
      "Updating NADt2_m with {'enz': 'YIL006W'}\n",
      "Updating NADt3_m with {'enz': 'YIL006W'}\n",
      "Updating NACt_c_e with {'enz': 'YGR260W'}\n",
      "Updating ORNtpa_m with {'enz': 'YOR130C'}\n",
      "Updating OAAtps_m with {'enz': 'YKL120W'}\n",
      "Updating PNTOtps_e with {'enz': 'YCR028C'}\n",
      "Updating PENDPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PHEACt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PItps_e with {'enz': 'YJL198W'}\n",
      "Updating PItps_m with {'enz': 'YER053C'}\n",
      "Updating Ktps_e with {'enz': 'YDR456W'}\n",
      "Updating PTRCtpa_e with {'enz': 'YOR273C'}\n",
      "Updating PTRCtpa_v with {'enz': 'YOR273C'}\n",
      "Updating PYRtps_e with {'enz': 'YKL217W'}\n",
      "Updating AMETtps_e with {'enz': 'YPL274W'}\n",
      "Updating AMETt_c_m with {'enz': 'YNL003C'}\n",
      "Updating MMETtps_e with {'enz': 'YLL061W'}\n",
      "Updating NAtpa_e with {'enz': 'YLR138W'}\n",
      "Updating SPMDtpa_e with {'enz': 'YHL016C'}\n",
      "Updating SPMDtpa_v with {'enz': 'YOR273C'}\n",
      "Updating SPRMtpa_e with {'enz': 'YPR156C'}\n",
      "Updating SPRMtpa_v with {'enz': 'YGR138C'}\n",
      "Updating SUCCt2_m with {'enz': 'YLR348C'}\n",
      "Updating SUCFUMt_c_m with {'enz': 'YJR095W'}\n",
      "Updating SO4t_c_e with {'enz': 'YLR092W'}\n",
      "Updating SO3t_c_e with {'enz': 'YPL092W'}\n",
      "Updating TAURt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating THMPPt_m with {'enz': 'YGR096W'}\n",
      "Updating THMtps_e with {'enz': 'YOR192C'}\n",
      "Updating URAtps_e with {'enz': 'YBR021W'}\n",
      "Updating URItps_e with {'enz': 'YBL042C'}\n",
      "Updating UTPt_c_m with {'enz': 'YBR192W'}\n",
      "Updating UTPUMPtp_m with {'enz': 'YBR192W'}\n",
      "Updating H2Ot_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ZYMSTt_e_en with {'enz': 'YOR011W'}\n",
      "Updating ARGtps_m with {'enz': 'YBR147W'}\n",
      "Updating PYRtps_m with {'enz': 'YGL080WYGR243W'}\n",
      "Updating AKGMALta_m with {'enz': 'YPL134C'}\n",
      "Updating FACOA120tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating FACOA140tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating FACOA160tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating FACOA161tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating FACOA180tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating FACOA181tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating FACOA200tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating FACOA220tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating FACOA240tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating FACOA260tabc_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating EX_13BDglucan_e with {'gene_reaction_rule': ''}\n",
      "Updating 13BDGLUCANt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_crn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_lac__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mev__R_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pnto__R_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_btd_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_3mop_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_lac__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mal__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_g3pi_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_1agpc_gen_e with {'gene_reaction_rule': ''}\n",
      "Updating MI13456Pt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating MI1Pt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_dad_2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_dgsn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_din_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_duri_e with {'gene_reaction_rule': ''}\n",
      "Updating 2DDA7Pt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 2DHPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_3c3hmp_e with {'gene_reaction_rule': ''}\n",
      "Updating 3C3HMPt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 3C3HMPt_c_m with {'enz': 'YKL120W'}\n",
      "Updating 2MBTOHt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 2MBTOHt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_2mbald_e with {'gene_reaction_rule': ''}\n",
      "Updating 2MBALDt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 2MBALDt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_2mbtoh_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_2mbac_e with {'gene_reaction_rule': ''}\n",
      "Updating IBTALDt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating IBTALDt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 2OBUTt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_akg_e with {'gene_reaction_rule': ''}\n",
      "Updating AKGt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating AKGt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_2phetoh_e with {'gene_reaction_rule': ''}\n",
      "Updating 2PHETOHt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 2PHETOHt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 3C4MOPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 3MOPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 3MOPt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_3mbald_e with {'gene_reaction_rule': ''}\n",
      "Updating 3MBALDt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 3MBALDt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 3HXPHBt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating AHMMPS4_c with {'enz': 'YFL058W'}\n",
      "Updating EX_4abz_e with {'gene_reaction_rule': ''}\n",
      "Updating 4ABZt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 4ABZt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 4ABUTNt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 4ABUTt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 4H2OGLTt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 4H2OGLTt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 4HBZt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NTD7pp_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD6pp_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD3pp_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD8_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD5_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD1pp_c with {'gene_reaction_rule': ''}\n",
      "Updating NTD10pp_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_5aop_e with {'gene_reaction_rule': ''}\n",
      "Updating 5AOPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating FTHFCL_m with {'enz': 'YER183C'}\n",
      "Updating EX_5fthf_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_dann_e with {'gene_reaction_rule': ''}\n",
      "Updating DANNt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_8aonn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_xan_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_acald_e with {'gene_reaction_rule': ''}\n",
      "Updating ACALDt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ACALDt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ac_e with {'gene_reaction_rule': ''}\n",
      "Updating ACt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ACCOAt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ACRNt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ADEt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_pap_e with {'gene_reaction_rule': ''}\n",
      "Updating PAPt_c_m with {'enz': 'YPR011C'}\n",
      "Updating ADPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ADPt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating AKGt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_alltt_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_tre_e with {'gene_reaction_rule': ''}\n",
      "Updating AKGMALt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_nh4_e with {'gene_reaction_rule': ''}\n",
      "Updating AMPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ASPGLUt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ATPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ATPt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HCO3E_m with {'gene_reaction_rule': ''}\n",
      "Updating HCO3E_n with {'gene_reaction_rule': ''}\n",
      "Updating HCO3E_c with {'enz': 'YNL036W'}\n",
      "Updating HCO3E_e with {'gene_reaction_rule': ''}\n",
      "Updating HCO3t_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_btn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_co2_e with {'gene_reaction_rule': ''}\n",
      "Updating CRNt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CRNCARt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CDPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CHLSTI_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_chol_e with {'gene_reaction_rule': ''}\n",
      "Updating CHOLt_c_en with {'enz': 'YOR161C'}\n",
      "Updating CHRPL_c with {'gene_reaction_rule': ''}\n",
      "Updating CITt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_cit_e with {'gene_reaction_rule': ''}\n",
      "Updating CITICTta_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CITMALta_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CMPN_c with {'gene_reaction_rule': ''}\n",
      "Updating CO2t_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CO2t_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CO2t_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CO2t_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CO2t_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating COAt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating COAt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating COAt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CYTK1_c with {'gene_reaction_rule': ''}\n",
      "Updating CYTK2_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_arab__D_e with {'gene_reaction_rule': ''}\n",
      "Updating E4Pt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_fru_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_gal_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_galur_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_sbt__D_e with {'gene_reaction_rule': ''}\n",
      "Updating GAM6Pt_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_glc__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_man_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_rib__D_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_xyl__D_e with {'gene_reaction_rule': ''}\n",
      "Updating DADPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DCDPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ATPHs_c with {'gene_reaction_rule': ''}\n",
      "Updating NTPDA2_c with {'gene_reaction_rule': ''}\n",
      "Updating NDPDA1_c with {'gene_reaction_rule': ''}\n",
      "Updating NDPDA2_c with {'gene_reaction_rule': ''}\n",
      "Updating DNADt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DADt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_dcyt_e with {'gene_reaction_rule': ''}\n",
      "Updating DCYTt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DGSNt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DINt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DURIK1_c with {'gene_reaction_rule': ''}\n",
      "Updating DURIt_c_e with {'enz': 'YBL042C'}\n",
      "Updating DPCOAK_c with {'gene_reaction_rule': ''}\n",
      "Updating DGDPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DHNPTt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating SPRMAT_c with {'enz': 'YDR071C'}\n",
      "Updating DHFt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DNTPPA_c with {'gene_reaction_rule': ''}\n",
      "Updating DHPTt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DHAPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PPIt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DOLPt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_dttp_e with {'gene_reaction_rule': ''}\n",
      "Updating DTTPt_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DUDPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DUMPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_epist_e with {'gene_reaction_rule': ''}\n",
      "Updating ERGTETROLt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ergst_e with {'gene_reaction_rule': ''}\n",
      "Updating ERGSTt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ERGSTt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ERGSTt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_etoh_e with {'gene_reaction_rule': ''}\n",
      "Updating ETOHt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ETOHt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_etha_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_etac_e with {'gene_reaction_rule': ''}\n",
      "Updating FRDPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DCAt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DDCAt_c_x with {'enz': 'YKL188CYPL147W'}\n",
      "Updating HDCEAt_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCTAt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TTDCAt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_fecost_e with {'gene_reaction_rule': ''}\n",
      "Updating FMNAT_m with {'enz': 'YDL045C'}\n",
      "Updating EX_fmn_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_fol_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_for_e with {'gene_reaction_rule': ''}\n",
      "Updating FORt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating FORt_c_e with {'enz': 'YNL065W'}\n",
      "Updating FUMt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating FRUK_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_fum_e with {'gene_reaction_rule': ''}\n",
      "Updating GDPt_c_g with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GDPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GDPMANNt_c_g with {'enz': 'YGL225W'}\n",
      "Updating GLCt_c_v with {'enz': 'YBR241C'}\n",
      "Updating EX_gthox_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_glyc_e with {'gene_reaction_rule': ''}\n",
      "Updating GLYC3Pt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_gly_e with {'gene_reaction_rule': ''}\n",
      "Updating GLYt_c_m with {'enz': 'YPR058W'}\n",
      "Updating GCALDt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GCALDt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_gcald_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_glx_e with {'gene_reaction_rule': ''}\n",
      "Updating GLXt_c_e with {'enz': 'YNL065W'}\n",
      "Updating GLXt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GUAt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GSNK_c with {'gene_reaction_rule': ''}\n",
      "Updating GSNt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GSNt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_g with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_v with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_h_e with {'gene_reaction_rule': ''}\n",
      "Updating HXCCOAt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_hxdcal_e with {'gene_reaction_rule': ''}\n",
      "Updating HDCAt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCEAt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HISt_c_m with {'enz': 'YBR147W'}\n",
      "Updating H2O2t_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HMGCOAt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_hxan_e with {'gene_reaction_rule': ''}\n",
      "Updating HXANt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_id3acald_e with {'gene_reaction_rule': ''}\n",
      "Updating ID3ACALDt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ID3ACALDt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating IND3ACt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating INSt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating INOSHPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating INSK_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_fe2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_iamac_e with {'gene_reaction_rule': ''}\n",
      "Updating IAMOHt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating IAMOHt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_iamoh_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ibutoh_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ibutac_e with {'gene_reaction_rule': ''}\n",
      "Updating IBUTOHt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating IBUTOHt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ibtald_e with {'gene_reaction_rule': ''}\n",
      "Updating PHCHGS_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating AOBUTDs_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_ala__L_e with {'gene_reaction_rule': ''}\n",
      "Updating ALAt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_abt__L_e with {'gene_reaction_rule': ''}\n",
      "Updating ABTt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ARABLt_c with {'enz': 'YHR092C'}\n",
      "Updating EX_arab__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_arg__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_asn__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_asp__L_e with {'gene_reaction_rule': ''}\n",
      "Updating CRNt_c_m with {'enz': 'YOR100C'}\n",
      "Updating EX_cys__L_e with {'gene_reaction_rule': ''}\n",
      "Updating E4HGLUt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating E4HGLUt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_sbt__L_e with {'gene_reaction_rule': ''}\n",
      "Updating G5SADr_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_glu__L_e with {'gene_reaction_rule': ''}\n",
      "Updating GLUt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_gln__L_e with {'gene_reaction_rule': ''}\n",
      "Updating GLNt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_his__L_e with {'gene_reaction_rule': ''}\n",
      "Updating HOMt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ile__L_e with {'gene_reaction_rule': ''}\n",
      "Updating ILEt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_leu__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_lys__L_e with {'gene_reaction_rule': ''}\n",
      "Updating MALt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_met__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_phe__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pro__L_e with {'gene_reaction_rule': ''}\n",
      "Updating PROt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ser__L_e with {'gene_reaction_rule': ''}\n",
      "Updating SBTLt_c_e with {'enz': 'YDL245C'}\n",
      "Updating EX_srb__L_e with {'gene_reaction_rule': ''}\n",
      "Updating SRBt_c_e with {'enz': 'YJL214W'}\n",
      "Updating EX_thr__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_trp__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_tyr__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_val__L_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_lanost_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ddca_e with {'gene_reaction_rule': ''}\n",
      "Updating LYSt_c_m with {'enz': 'YBR147W'}\n",
      "Updating MALOAAta_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_malt_e with {'gene_reaction_rule': ''}\n",
      "Updating MANNANt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating METt_c_m with {'enz': 'YDR508C'}\n",
      "Updating MGSA_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_inost_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_Nfortyr2_e with {'gene_reaction_rule': ''}\n",
      "Updating NADPt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NADPHt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NH4t_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NCAMt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_nac_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_nmn_e with {'gene_reaction_rule': ''}\n",
      "Updating NMNt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NMNt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NMNt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NDP4_c with {'gene_reaction_rule': ''}\n",
      "Updating NTP4_c with {'gene_reaction_rule': ''}\n",
      "Updating ACRNt_c_m with {'enz': 'YOR100C'}\n",
      "Updating O2t_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating O2t_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating O2t_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating O2t_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCDCAt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_octa_e with {'gene_reaction_rule': ''}\n",
      "Updating OAAt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_oaa_e with {'gene_reaction_rule': ''}\n",
      "Updating GTHOXt_c_e with {'enz': 'YJL212C'}\n",
      "Updating TRDOXt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_o2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hdca_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hdcea_e with {'gene_reaction_rule': ''}\n",
      "Updating PMTCOAt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAN4Pt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PTPATi_m with {'gene_reaction_rule': ''}\n",
      "Updating PAPt_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_pectin_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pheac_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_pacald_e with {'gene_reaction_rule': ''}\n",
      "Updating PACALDt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PACALDt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PHEt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_pi_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_k_e with {'gene_reaction_rule': ''}\n",
      "Updating PPPG9t_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PRPPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ptrc_e with {'gene_reaction_rule': ''}\n",
      "Updating PYDXK_c with {'enz': 'YNR027W'}\n",
      "Updating PYDAMK_c with {'gene_reaction_rule': ''}\n",
      "Updating HYPOE_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_pydxn_e with {'gene_reaction_rule': ''}\n",
      "Updating PMDPHT_c with {'gene_reaction_rule': ''}\n",
      "Updating PYNP2r_c with {'gene_reaction_rule': ''}\n",
      "Updating PPIt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_pyr_e with {'gene_reaction_rule': ''}\n",
      "Updating QULNt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TRDRDt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ribflv_e with {'gene_reaction_rule': ''}\n",
      "Updating RIBFLVt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating RIBFLVt_c_e with {'enz': 'YOR306C'}\n",
      "Updating RIBt_c_e with {'enz': 'YHR092C'}\n",
      "Updating AHCYSt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_amet_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mmet_e with {'gene_reaction_rule': ''}\n",
      "Updating SERt_c_m with {'enz': 'YDR508C'}\n",
      "Updating EX_g3pc_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_na1_e with {'gene_reaction_rule': ''}\n",
      "Updating SPMDAT_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_spmd_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_sprm_e with {'gene_reaction_rule': ''}\n",
      "Updating SQLt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating SSQ23EPXt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ocdca_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_succ_e with {'gene_reaction_rule': ''}\n",
      "Updating SUCCt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_sucr_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_so4_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_so3_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_taur_e with {'gene_reaction_rule': ''}\n",
      "Updating TTCCOAt_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating THMP_c with {'enz': 'YDL024C'}\n",
      "Updating TMN_c with {'gene_reaction_rule': ''}\n",
      "Updating EX_thmpp_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_thm_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_thmmp_e with {'gene_reaction_rule': ''}\n",
      "Updating TMPK_c with {'gene_reaction_rule': ''}\n",
      "Updating THZPSN2_c with {'gene_reaction_rule': ''}\n",
      "Updating THZPSN1_c with {'gene_reaction_rule': ''}\n",
      "Updating THRt_c_m with {'enz': 'YDR508C'}\n",
      "Updating EX_thymd_e with {'gene_reaction_rule': ''}\n",
      "Updating TMDK1_c with {'gene_reaction_rule': ''}\n",
      "Updating THYMDt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TREt_c_v with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TRPt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ind3eth_e with {'gene_reaction_rule': ''}\n",
      "Updating IND3ETHt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating IND3ETHt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TYRt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TYRt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating UMPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ura_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_urea_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_uri_e with {'gene_reaction_rule': ''}\n",
      "Updating VALt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_g with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_v with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_h2o_e with {'gene_reaction_rule': ''}\n",
      "Updating XANt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_xtsn_e with {'gene_reaction_rule': ''}\n",
      "Updating XTSNt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_xylt_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_zymst_e with {'gene_reaction_rule': ''}\n",
      "Updating ZYMSTt_c_en with {'enz': 'YOR011W'}\n",
      "Updating COAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_14dmlanost_e with {'gene_reaction_rule': ''}\n",
      "Updating 14DMLANOSTt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ergtetrol_e with {'gene_reaction_rule': ''}\n",
      "Updating ERGTETROLt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCTAt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DCAt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DDCAt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_but_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_hxa_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ocdcea_e with {'gene_reaction_rule': ''}\n",
      "Updating BUTt_c_e with {'enz': 'YNL065W'}\n",
      "Updating HXAt_c_e with {'enz': 'YNL065W'}\n",
      "Updating OCDCEAt_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ttdca_e with {'gene_reaction_rule': ''}\n",
      "Updating BUTt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HXAt_c_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DDCAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TTDCAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCEAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCDCAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCDCEAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating MALCOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DDCACOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TDCOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PMTCOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating STCOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ODECOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ICOSCOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DOCOSCOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TTCCOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HXCCOAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CO2t_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating COAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NADPHt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NADPt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating O2t_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NADHt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NADt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GLYC3Pt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DHAPt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PPIt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PIt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CTPt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CDPt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CMPt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CDPEAt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CDPCHOLt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ATPt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating AMPt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating SERt_c_rm with {'enz': 'YKR039W'}\n",
      "Updating INOSTt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating AMETt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating AHCYSt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EPISTt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating FECOSTt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating LANOSTt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ERGSTt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ZYMSTt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating G3PCt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCEAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCDCAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCDCEAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DDCACOAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TDCOAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PMTCOAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCOAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating STCOAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ODECOAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GLYC3Pt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DHAPt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NADPHt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NADPt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ATPt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating AMPt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PPIt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GLYCt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DOCOSCOAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TTCCOAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HXCCOAt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ATPt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ADPt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating G3PCt_c_en with {'enz': 'YCR098C'}\n",
      "Updating G3PIt_c_en with {'enz': 'YCR098C'}\n",
      "Updating HDCAt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCEAt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCDCAt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCDCEAt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ICOSCOAt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DOCOSCOAt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TTCCOAt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HXCCOAt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PIt_c_vm with {'enz': 'YNR013C'}\n",
      "Updating Ht_c_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ATPt_c_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ADPt_c_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CO2t_c_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_gm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PIt_c_gm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_c_gm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CO2t_c_gm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ATPt_c_gm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ADPt_c_gm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating H2Ot_c_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CO2t_c_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCOAt_c_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ODECOAt_c_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HDCAt_c_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating OCDCAt_c_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating COAt_c_mm with {'enz': 'YHR002W'}\n",
      "Updating DDCAt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TTDCAt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TTCt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HXCAt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EPISTt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating FECOSTt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating LANOSTt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ZYMSTt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ERGST161t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ERGST181t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EPIST161t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EPIST181t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating FECOST161t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating FECOST181t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating LANOST161t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating LANOST181t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ZYM161t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ZYM181t_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TTCt_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HXCAt_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ZYM161t_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ZYM181t_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TTCt_r_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HXCAt_r_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating ERGSTt_r_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PIt_r_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating Ht_m_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating GLYC3Pt_m_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CMPt_m_mm with {'enz': 'YBR192W'}\n",
      "Updating CTPt_m_mm with {'enz': 'YBR192W'}\n",
      "Updating PPIt_m_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HEMEAt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_melib_e with {'gene_reaction_rule': ''}\n",
      "Updating ATPM_c with {'gene_reaction_rule': ''}\n",
      "Updating CYCCL_m with {'enz': 'YKL087C'}\n",
      "Updating ACTNOR_c with {'enz': 'YAL061W'}\n",
      "Updating ACTNOR_n with {'enz': 'YAL061W'}\n",
      "Updating PADPRGH_c with {'enz': 'YBR022W'}\n",
      "Updating ADPRIB1P_c with {'enz': 'YBR046C'}\n",
      "Updating ALG14RXN_r with {'enz': 'YBR070CYGL047W'}\n",
      "Updating ARGt_c_v with {'enz': 'YDR352W'}\n",
      "Updating HISt_c_v with {'enz': 'YBR147W'}\n",
      "Updating LYSt_c_v with {'enz': 'YBR147W'}\n",
      "Updating NA1CLKt_c_v with {'enz': 'YBR235W'}\n",
      "Updating GTHH_c with {'enz': 'YBR281CYNL191W'}\n",
      "Updating MN2tabc_c_r with {'enz': 'YEL031W'}\n",
      "Updating OAADC_c with {'enz': 'YER010C'}\n",
      "Updating GTHCT_c with {'enz': 'YER163C'}\n",
      "Updating ZN2t_c_r with {'enz': 'YIL023C'}\n",
      "Updating 5OPROAH_c with {'enz': 'YKL215C'}\n",
      "Updating CYSGLYP_c with {'enz': 'YFR044C'}\n",
      "Updating FMNH_c with {'enz': 'YDL024C'}\n",
      "Updating NA1tabc_c_e with {'enz': 'YDR040C'}\n",
      "Updating GLCNK_c with {'enz': 'YDR248C'}\n",
      "Updating DNTPPY2_c with {'enz': 'YJR069C'}\n",
      "Updating DNTPPY3_c with {'enz': 'YJR069C'}\n",
      "Updating APOOCT_m with {'enz': 'YLR239C'}\n",
      "Updating APOLIPT_m with {'enz': 'YLR239C'}\n",
      "Updating 4ABTORx_c with {'enz': 'YMR110C'}\n",
      "Updating r_4323_m with {'enz': 'YOR196C'}\n",
      "Updating r_4324_m with {'enz': 'YOR196C'}\n",
      "Updating MG2t_c_m with {'enz': 'YPL060W'}\n",
      "Updating PYDXR_c with {'enz': 'YPR127W'}\n",
      "Updating ENPP_c with {'enz': 'YDR452W'}\n",
      "Updating 2HGOR_c with {'gene_reaction_rule': ''}\n",
      "Updating CA2tps_c_e with {'enz': 'YOL122C'}\n",
      "Updating CUtps_c_e with {'enz': 'YOL122C'}\n",
      "Updating ZN2tps_c_e with {'enz': 'YGL255W'}\n",
      "Updating MG2tps_c_e with {'enz': 'YOL130W'}\n",
      "Updating EX_cl_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_cu2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mn2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_zn2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_mg2_e with {'gene_reaction_rule': ''}\n",
      "Updating EX_ca2_e with {'gene_reaction_rule': ''}\n",
      "Updating PIt_c_g with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PPCOAt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating UDPt_c_g with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating UDPt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating UMPt_c_g with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DHAPAT_rm with {'enz': 'YBL011W'}\n",
      "Updating G3PAT_rm with {'enz': 'YKR067W'}\n",
      "Updating G3PAT_l with {'enz': 'YKR067W'}\n",
      "Updating AGNPR_rm with {'enz': 'YIL124W'}\n",
      "Updating AGNPR_l with {'enz': 'YIL124W'}\n",
      "Updating AGPAT_rm with {'enz': 'YOR175C'}\n",
      "Updating AGPAT_l with {'enz': 'YOR081C'}\n",
      "Updating ETHAPT_rm with {'enz': 'YHR123W'}\n",
      "Updating CHOLPT_rm with {'enz': 'YNL130C'}\n",
      "Updating CDPDAGS_rm with {'enz': 'YBR029C'}\n",
      "Updating PSSA_rm with {'enz': 'YER026C'}\n",
      "Updating PAILS_rm with {'enz': 'YPR113W'}\n",
      "Updating PGPS_mm with {'enz': 'YCL004W'}\n",
      "Updating PGPP_mm with {'enz': 'YHR100C'}\n",
      "Updating CLPNS_mm with {'enz': 'YDL142C'}\n",
      "Updating 1MLCLAT_mm with {'enz': 'YPR140W'}\n",
      "Updating 3MLCLAT_mm with {'enz': 'YPR140W'}\n",
      "Updating PAP_rm with {'enz': 'YMR165C'}\n",
      "Updating PAP_vm with {'enz': 'YDR284C'}\n",
      "Updating PAP_gm with {'enz': 'YDR503C'}\n",
      "Updating DGAT_rm with {'enz': 'YOR245C'}\n",
      "Updating DGAT_l with {'enz': 'YOR245C'}\n",
      "Updating TAGL_l with {'enz': 'YCR068W'}\n",
      "Updating DAGL_m with {'enz': 'YDR058C'}\n",
      "Updating MAGL_l with {'enz': 'YKL094W'}\n",
      "Updating DAGK_rm with {'enz': 'YOR311C'}\n",
      "Updating PAK_gm with {'gene_reaction_rule': ''}\n",
      "Updating PAK_vm with {'gene_reaction_rule': ''}\n",
      "Updating DGPP_vm with {'enz': 'YDR284C'}\n",
      "Updating DGPP_gm with {'enz': 'YDR503C'}\n",
      "Updating LPAP_vm with {'enz': 'YDR284C'}\n",
      "Updating LPAP_c with {'enz': 'YER037W'}\n",
      "Updating LPAP_gm with {'enz': 'YDR503C'}\n",
      "Updating PDAGATpe_rm with {'enz': 'YNR008W'}\n",
      "Updating PDAGATpc_rm with {'enz': 'YNR008W'}\n",
      "Updating PSD_mm with {'enz': 'YNL169C'}\n",
      "Updating PSD_gm with {'enz': 'YGR170W'}\n",
      "Updating PSD_vm with {'enz': 'YGR170W'}\n",
      "Updating PEMT_rm with {'enz': 'YGR157W'}\n",
      "Updating PMEMT_rm with {'enz': 'YJR073C'}\n",
      "Updating PDMEMT_rm with {'enz': 'YJR073C'}\n",
      "Updating LPCAT_mm with {'gene_reaction_rule': ''}\n",
      "Updating LPEAT_l with {'enz': 'YMR313C'}\n",
      "Updating PLAA2pe_l with {'enz': 'YKR089C'}\n",
      "Updating PLAclpn1_mm with {'enz': 'YGR110W'}\n",
      "Updating PLAclpn3_mm with {'enz': 'YGR110W'}\n",
      "Updating PLABlpc_en with {'enz': 'YMR008C'}\n",
      "Updating PLABlpc_rm with {'enz': 'YML059C'}\n",
      "Updating PLABlpe_en with {'enz': 'YMR006C'}\n",
      "Updating PLABlpi_en with {'enz': 'YOL011W'}\n",
      "Updating PLABpc_en with {'enz': 'YMR008C'}\n",
      "Updating PLABpc_rm with {'enz': 'YML059C'}\n",
      "Updating PLABpe_en with {'enz': 'YMR006C'}\n",
      "Updating PLABpi_en with {'enz': 'YOL011W'}\n",
      "Updating PLABps_en with {'enz': 'YOL011W'}\n",
      "Updating PLACipc_r with {'enz': 'YER019W'}\n",
      "Updating PLACmipc_r with {'enz': 'YER019W'}\n",
      "Updating PLACmip2c_r with {'enz': 'YER019W'}\n",
      "Updating PLACipc_m with {'enz': 'YER019W'}\n",
      "Updating PLACmipc_m with {'enz': 'YER019W'}\n",
      "Updating PLACmip2c_m with {'enz': 'YER019W'}\n",
      "Updating PLACpail45p_c with {'enz': 'YPL268W'}\n",
      "Updating PLACpail45p_n with {'enz': 'YPL268W'}\n",
      "Updating PLACpg_mm with {'enz': 'YPL206C'}\n",
      "Updating PLADpc_en with {'enz': 'YKR031C'}\n",
      "Updating PI3PP_en with {'enz': 'YOR109W'}\n",
      "Updating PI3PP_c with {'enz': 'YJR110W'}\n",
      "Updating PI3PP_rm with {'enz': 'YKL212W'}\n",
      "Updating PI3PP_gm with {'enz': 'YKL212W'}\n",
      "Updating PI4PP_en with {'enz': 'YOR109W'}\n",
      "Updating PI4PP_c with {'enz': 'YOR109W'}\n",
      "Updating PI4PP_rm with {'enz': 'YKL212W'}\n",
      "Updating PI4PP_gm with {'enz': 'YKL212W'}\n",
      "Updating LPIAT_rm with {'enz': 'YBR042C'}\n",
      "Updating PAIL4K_en with {'enz': 'YLR305C'}\n",
      "Updating PAIL4K_gm with {'enz': 'YDR373WYNL267W'}\n",
      "Updating PAIL4K_n with {'enz': 'YNL267W'}\n",
      "Updating PAIL3K_vm with {'enz': 'YBR097WYLR240W'}\n",
      "Updating PI4P5K_n with {'enz': 'YDR208W'}\n",
      "Updating PI4P5K_en with {'enz': 'YDR208W'}\n",
      "Updating PI45P5P_c with {'enz': 'YIL002C'}\n",
      "Updating PI45P5P_en with {'enz': 'YIL002C'}\n",
      "Updating PI45P5P_rm with {'enz': 'YOL065C'}\n",
      "Updating PI3P5K_vm with {'enz': 'YFR019WYFR021WYLR386WYNL054WYNL325C'}\n",
      "Updating PI35P5P_en with {'enz': 'YOR109W'}\n",
      "Updating PI35P5P_c with {'enz': 'YOR109W'}\n",
      "Updating PI35P5P_vm with {'enz': 'YLR386WYNL325C'}\n",
      "Updating IPCS_g with {'enz': 'YDR367WYKL004W'}\n",
      "Updating MIPCS_g with {'enz': 'YBR036CYPL057C'}\n",
      "Updating IPT_g with {'enz': 'YDR072C'}\n",
      "Updating compCER_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compCER_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compCER_g with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compACYLCOA_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compACYLCOA_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compACYLCOA_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compACYLCOA_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compACYLCOA_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compACYLCOA_x with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compFALPD_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compFALPD_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compFALPD_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compFALPD_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating compFALPD_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAt_gm_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAt_mm_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAt_rm_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAt_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DAGt_c_l with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DAGt_c_mm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DAGt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DAGt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DAGt_gm_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TAGt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating TAGt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 1AGPEt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating 1AGPCt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PSt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PSt_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PSt_mm_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PSt_gm_rm with {'enz': 'YIL048W'}\n",
      "Updating PSt_rm_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAILt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAILt_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAILt_rm_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAILt_gm_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAILt_n_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating CERt_g_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PCt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PCt_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PCt_mm_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PCt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PEt_l_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PEt_gm_rm with {'enz': 'YAL026C'}\n",
      "Updating PEt_c_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PEt_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PEt_mm_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PEt_rm_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAIL4Pt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAIL4Pt_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAIL45Pt_c_n with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAIL45Pt_c_en with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAIL45Pt_en_rm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAIL3Pt_gm_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAIL3Pt_rm_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAIL35Pt_c_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating PAIL35Pt_en_vm with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating IPCt_g_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating MIPCt_g_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating MIP2Ct_g_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating BIOMASS_AERO_SC_RBA with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating AKGDH_m with {'enz': 'YDR148CYFL018CYIL125W'}\n",
      "Updating ETFOXRq6_m with {'enz': 'YGR207CYOR356WYPR004C'}\n",
      "Updating FADt_c_m with {'enz': 'YIL134W'}\n",
      "Updating PYDX5PS_c with {'enz': 'YFL059WYMR095C'}\n",
      "Updating PRENT13_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT11_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT20_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT16_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT8_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT19_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT14_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT7_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT12_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT21_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating GGTT_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT6_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT10_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT18_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating FRTT_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT17_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT9_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PPTT_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating PRENT15_rm with {'enz': 'YBR002CYDL193W'}\n",
      "Updating SERD_L_m with {'enz': 'YCL064C'}\n",
      "Updating HCITS_c with {'enz': 'YDL182W'}\n",
      "Updating HCITt_c_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating THFGLUS_m with {'enz': 'YKL132C'}\n",
      "Updating PEPSM1_c with {'enz': 'YBL091C'}\n",
      "Updating FLPT_c with {'enz': 'YDR538W'}\n",
      "Updating FMETTRS_c with {'enz': 'YBL013W'}\n",
      "Updating LDHDf_c with {'enz': 'YEL071W'}\n",
      "Updating compCERANAERO_r with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating BIOMASS_BATCHANAERO_SC_hvd with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating BIOMASS_CHEMOANAERO_SC_hvd with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating BIOMASS_STARVEAERO_SC_hvd with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating BIOMASS_STARVEANAERO_SC_hvd with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DICAT1_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DICAT2_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DICAT3_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DICAT4_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DICAT5_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DICAT6_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DICAT7_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating DICAT8_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_cobalt2_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating COBALT2t_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating EX_ni2_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating NI2t_c_e with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HEME1_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HEME2_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HEME3_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating HEME4_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating LIPOATE_m with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating MONCAT1_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating MONCAT2_c with {'gene_reaction_rule': 'SPONT'}\n",
      "Updating MONCAT3_c with {'gene_reaction_rule': 'SPONT'}\n"
     ]
    }
   ],
   "source": [
    "# check if all rxns in model are accounted for in df_enz_rxns, and if they have been automatically updated\n",
    "# make a list of all rxns in df_enz_rxns\n",
    "enz_rxns = []\n",
    "for index, row in df_enz_rxns.iterrows():\n",
    "\tenz_rxns += row['rxns'].split(',')\n",
    "enz_rxns = sorted(list(dict.fromkeys(enz_rxns)))\n",
    "# check if all rxns in model are in enz_rxns\n",
    "missing_rxns = []\n",
    "for rxn in model.reactions:\n",
    "\tif rxn.id not in rxns_not_needing_enzymes and rxn.id not in enz_rxns and rxn.id not in whole_biomass_rxns and rxn.gene_reaction_rule not in spont_rxn_dict.keys():\n",
    "\t\tmissing_rxns.append(rxn.id)\n",
    "if len(missing_rxns) > 0:\n",
    "\tprint(f'Warning: the following reactions are not accounted for in df_enz_rxns:')\n",
    "\tfor rxn in missing_rxns:\n",
    "\t\tprint(rxn)\n",
    "# remove rxns from rxn_updates if their value is {}\n",
    "for rxn in rxn_updates.copy():\n",
    "\tif not rxn_updates[rxn]:\n",
    "\t\tdel rxn_updates[rxn]\n",
    "\telse:\n",
    "\t\tprint(f'Updating {rxn} with {rxn_updates[rxn]}')\n",
    "# if rxn_updates isn't empty, print a warning\n",
    "# if len(rxn_updates) > 0:\n",
    "# \traise ValueError(f'Warning: the following reactions have been automatically updated:\\n{rxn_updates}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert df_enz_rxns to old ENZYME_stoich_curation file format from 2024 version of scRBA, for compatibility\n",
    "df_rxn_enz_pairs = pd.DataFrame(columns=['id','rxn_src','enz','protein_stoich'])\n",
    "catalyst_rxn_dict = {**spont_rxn_dict, **enz_rxn_dict}\n",
    "for k, v in catalyst_rxn_dict.items():\n",
    "\tif k in spont_rxn_dict:\n",
    "\t\tcatalyst_rxn_dict[k]['rxns'].extend(spont_rxn_dict[k]['rxns']) # adds rxns that were overrided by merger with enz_rxn_dict\n",
    "\t\tcatalyst_rxn_dict[k]['rxns'] = list(dict.fromkeys(catalyst_rxn_dict[k]['rxns']))\n",
    "for enz in catalyst_rxn_dict:\n",
    "\tenz_id = enz if enz != '' else 'SPONT'\n",
    "\t# don't seem to need MW of enzyme in prior RBA code\n",
    "\tprot_stoich_str = 'zeroCost' if enz in spont_GPRs + [''] else ','.join([prot+':'+str(catalyst_rxn_dict[enz]['protein_stoich'][prot]) for prot in catalyst_rxn_dict[enz]['protein_stoich']])\n",
    "\tfor rxn in catalyst_rxn_dict[enz]['rxns']:\n",
    "\t\tdirs = []\n",
    "\t\t# add FWD and REV to rxn_src if those directions are possible\n",
    "\t\tif model.reactions.get_by_id(rxn).lower_bound < 0:\n",
    "\t\t\tdirs.append('REV')\n",
    "\t\tif model.reactions.get_by_id(rxn).upper_bound > 0:\n",
    "\t\t\tdirs.append('FWD')\n",
    "\t\tfor dir in dirs:\n",
    "\t\t\t# print(enz,rxn)\n",
    "\t\t\t# df_rxn_enz_pairs = df_rxn_enz_pairs.append({'id':f'RXN-{rxn}_{dir}-{enz}','enz':enz,'gpr':'','protein_stoich':','.join([prot+':'+str(enz_rxn_dict[enz]['protein_stoich'][prot]) for prot in enz_rxn_dict[enz]['protein_stoich']]),'status':'automatically added'}, ignore_index=True)\n",
    "\t\t\tdf_rxn_enz_pairs.loc[len(df_rxn_enz_pairs)] = [f'RXN-{rxn}_{dir}-{enz_id}',rxn,enz,prot_stoich_str]\n",
    "\t\t\t# df_rxn_enz_pairs = pd.concat([df_rxn_enz_pairs,pd.DataFrame({'id':f'RXN-{rxn}_{dir}-{enz}','enz':enz,'protein_stoich':','.join([prot+':'+str(enz_rxn_dict[enz]['protein_stoich'][prot]) for prot in enz_rxn_dict[enz]['protein_stoich']])})], ignore_index=True)\n",
    "\n",
    "# sort by ID\n",
    "df_rxn_enz_pairs = df_rxn_enz_pairs.sort_values(by='id')\t\t\t\n",
    "df_rxn_enz_pairs.to_csv('./model/ENZYME_rxn_pairs.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_eqn = pd.DataFrame(columns=['id', 'type', 'coupling_type', 'coupling_species', 'reaction','tag','FBA_name','dir','enz_id'])\n",
    "c = -1\n",
    "medium = []\n",
    "\n",
    "### Metabolic network reaction\n",
    "# Exchange reactions\n",
    "for rxn in model.reactions:\n",
    "\t# check if rxn is in df_enz\n",
    "\t# if rxn.id not in df_enz.rxn_src.to_list():\n",
    "\tif rxn.id[:3] == 'EX_':\n",
    "\t\tenz_id = 'SPONT'\n",
    "\t\tmake_reversible = True\n",
    "# else:\n",
    "# \tgpr = rxn.gene_reaction_rule if rxn.gene_reaction_rule != \"\" else 'UNKNOWN'\n",
    "\n",
    "\t\tprotein_stoich = 'zeroCost' if enz_id in spont_GPRs else ''\n",
    "\t\t\n",
    "\t\tmet = [i for i in rxn.metabolites.keys()][0]\n",
    "\t\tc += 1\n",
    "\t\ttag = 'RXN'\n",
    "\t\tdir = 'FWD'\n",
    "\t\tnew_id = f'{tag}-{rxn.id}_{dir}-{enz_id}'\n",
    "\t\tdf_eqn.loc[c, 'id'] = new_id\n",
    "\t\tdf_eqn.loc[c, 'type'] = 'metabolic'\n",
    "\t\tdf_eqn.loc[c, 'reaction'] = 'MET-' + met.id + ' -->'\n",
    "\t\t# df_eqn.loc[c, ['FBA_name','dir']] = [rxn.id,'FWD']\n",
    "\t\tdf_eqn.loc[c, ['tag','FBA_name','dir','enz_id']] = [tag,rxn.id,dir,enz_id]\n",
    "\t\t\n",
    "\t\tc += 1\n",
    "\t\tdir = 'REV'\n",
    "\t\tnew_id = f'{tag}-{rxn.id}_{dir}-{enz_id}'\n",
    "\t\tdf_eqn.loc[c, 'id'] = new_id\n",
    "\t\tdf_eqn.loc[c, 'type'] = 'metabolic'\n",
    "\t\tdf_eqn.loc[c, 'reaction'] = '-->' + 'MET-' + met.id\n",
    "\t\t# df_eqn.loc[c, ['FBA_name','dir']] = [rxn.id,'REV']\n",
    "\t\tdf_eqn.loc[c, ['tag','FBA_name','dir','enz_id']] = [tag,rxn.id,dir,enz_id]\n",
    "\n",
    "# Reactions that are not exchange reactions\n",
    "for i in df_enz.index:\n",
    "\trxn_id = df_enz.id[i]\n",
    "\ttag,rxn_base_id,rxn_dir,enz_id = extract_details_from_rxnid(rxn_id)\n",
    "\t\n",
    "\tif rxn_base_id[:3] == 'EX_':\n",
    "\t\tcontinue\n",
    "\t\n",
    "\tc += 1\n",
    "\trxn_base = model.reactions.get_by_id(rxn_base_id)\n",
    "\t\n",
    "\tmet_dict = metabolites_dict_from_reaction_equation_RBA(rxn_base.reaction)\n",
    "\tmet_dict = {k:v for k,v in met_dict.items() if k != ''}\n",
    "\tmet_dict = {'MET-' + k:v for k,v in met_dict.items()}\n",
    "\tif rxn_dir == 'REV':\n",
    "\t\tmet_dict = {k:-v for k,v in met_dict.items()}\n",
    "\telif rxn_dir == 'FWD':\n",
    "\t\tNone\n",
    "\telse:\n",
    "\t\tprint(\"Unknown ID that indicate reaction direction, only accepting 'FWD' and 'REV'\")\n",
    "\t\n",
    "\tif enz_id not in spont_GPRs:\n",
    "\t\tdf_eqn.loc[c, 'coupling_type'] = 'rxn_enz'\n",
    "\t\tdf_eqn.loc[c, 'coupling_species'] = enz_id\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = rxn_id\n",
    "\tdf_eqn.loc[c, 'type'] = 'metabolic'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(met_dict, arrow='-->')\n",
    "\tdf_eqn.loc[c, ['tag','FBA_name','dir','enz_id']] = [tag,rxn_base_id,rxn_dir,enz_id]\n",
    "\t\n",
    "### Enzyme synthesis network reaction\n",
    "enz_stoich = OrderedDict()\n",
    "for i in df_enz.index:\n",
    "\tenz_stoich[df_enz.enz[i]] = df_enz.protein_stoich[i]\n",
    "\n",
    "c = df_eqn.shape[0] - 1\n",
    "for enz_id,prot_str in enz_stoich.items():\n",
    "\tif prot_str == 'zeroCost':\n",
    "\t\tcontinue\n",
    "\t\n",
    "\tc += 1\n",
    "\tprot_str = prot_str.split(',')\n",
    "\tcoeffs = OrderedDict({'PTMPRO-' + i.split(':')[0]:-int(i.split(':')[1]) for i in prot_str})\n",
    "\tcoeffs['ENZ-' + enz_id] = 1\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = 'ENZSYN-' + enz_id\n",
    "\tdf_eqn.loc[c, 'type'] = 'enzyme'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(coeffs, arrow='-->')\n",
    "\n",
    "#these ENZLOAD equations also need to have their MW listed in the enz_mw_g_per_mmol file\n",
    "#initialize an enzload sting\n",
    "enzload_str = \"\"\n",
    "for i in df_enz.index:\n",
    "\tif df_enz.protein_stoich[i] == 'zeroCost':\n",
    "\t\tcontinue\n",
    "\t\t\n",
    "\tc += 1\n",
    "\tcoeffs = OrderedDict()\n",
    "\tcoeffs['ENZ-' + df_enz.enz[i]] = -1\n",
    "\n",
    "\tenzload_id = 'ENZLOAD-' + df_enz.id[i][4:]\n",
    "\n",
    "\t#for debugging, comment out if unused\n",
    "\t#print(\"i: \",i,\"\\tenzload_id: \", enzload_id)\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = enzload_id\n",
    "\tdf_eqn.loc[c, 'type'] = 'enzymeRxnLoad'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(coeffs, arrow='-->')\n",
    "\n",
    "\t#get the associated mw\n",
    "\tenz_mw = getattr(df_enz,\"MW (g/mmol)\")[i]\n",
    "\n",
    "\tnew_line = enzload_id + \"\\t\" + str(enz_mw) + \"\\n\"\n",
    "\tenzload_str = enzload_str + new_line\n",
    "\n",
    "# write the file resulting from enzload_str\n",
    "with open('./model/enz_mw_g_per_mmol.txt', 'w') as f:\n",
    "\tf.write(enzload_str)\n",
    "with open('../input/enz_mw_g_per_mmol.txt', 'w') as f:\n",
    "\tf.write(enzload_str)\n",
    "\n",
    "### Ribosome\n",
    "ribo_subunits = {}\n",
    "c = df_eqn.shape[0] - 1\n",
    "for rna in df_rnas.index:\n",
    "\tc += 1\n",
    "\trna_stoich = OrderedDict({i:0 for i in ['MET-'+rna, 'MET-atp_c', 'MET-ctp_c',\n",
    "\t\t\t\t\t\t\t\t\t\t\t'MET-gtp_c', 'MET-utp_c', 'MET-ppi_c']})\n",
    "\trna_stoich['RIBO-'+rna] = 1\n",
    "\trna_stoich['MET-atp_c'] = -int(df_rnas.A[rna])\n",
    "\trna_stoich['MET-ctp_c'] = -int(df_rnas.C[rna])\n",
    "\trna_stoich['MET-gtp_c'] = -int(df_rnas.G[rna])\n",
    "\trna_stoich['MET-utp_c'] = -int(df_rnas.U[rna])\n",
    "\trna_stoich['MET-ppi_c'] = int(df_rnas.loc[rna, ['A','C','G','U']].sum())\n",
    "\trna_stoich['BIO-rrna'] = df_rnas.loc[rna, 'MW (g/mmol)']\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = 'RIBOSYN-' + rna\n",
    "\tdf_eqn.loc[c, 'type'] = 'ribosome'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(rna_stoich, arrow='-->')\n",
    "\tgene_expression_dict.setdefault(rna, []).append(df_eqn.loc[c, 'id'])\n",
    "with open('./model/prosyn_ribo_subunits.txt', 'w') as f, open('./model/pro_ribo_subunits.txt', 'w') as f2, open('./model/kribo.txt', 'w') as f4, open('./model/j_ribo.txt', 'w') as f5, open('./model/ribosomes.txt', 'w') as f6:\n",
    "\tfor file in [f,f2,f4,f5,f6]: file.write('/')\n",
    "\tfor loc, ribo in ribo_dict.items():\n",
    "\t\t## proteins into ribosome subunits (to account for paralogs)\n",
    "\t\tribo_stoich = OrderedDict()\n",
    "\t\t# if only 1 ribosome is present, all rRNAs are used\n",
    "\t\trnas = [item for item in df_rnas.index if item.endswith(loc) or len(ribo_dict.keys()) == 1]\n",
    "\t\t# assume 1 of each rRNA and protein is used per ribosome\n",
    "\t\tfor i in ribo.index:\n",
    "\t\t\tif ribo.id[i] in ribo_subunits:\n",
    "\t\t\t\tribo_subunits[ribo.id[i]].append(loc[1:])\n",
    "\t\t\telse:\n",
    "\t\t\t\tribo_subunits[ribo.id[i]] = [loc[1:]]\n",
    "\t\t\tif ribo.id[i] in rnas:\n",
    "\t\t\t\tribo_stoich['RIBO-' + ribo.id[i]] = -1\n",
    "\t\t\telse:\n",
    "\t\t\t\tf.write(\"\\n'\"+loc[1:]+\"'.'PROSYN-\"+ribo.id[i]+\"'\")\n",
    "\t\t\t\tf2.write(\"\\n'\"+loc[1:]+\"'.'\"+ribo.id[i]+\"'\")\n",
    "\t\t\t\t# if it's not a paralog of another protein, all paralogs between it and the next non-paralog are treated as paralogs\n",
    "\t\t\t\tif pd.isnull(ribo.paralog[i]):\n",
    "\t\t\t\t\tribosub_name = 'RIBOSUB-' + ribo.id[i]\n",
    "\t\t\t\t# add rxn converting the protein into the ribosome subunit (RIBOSUB prefix)\n",
    "\t\t\t\tc += 1\n",
    "\t\t\t\tribosub_stoich = OrderedDict()\n",
    "\t\t\t\tribosub_stoich['PTMPRO-' + ribo.id[i]] = -1\n",
    "\t\t\t\tribosub_stoich[ribosub_name] = 1\n",
    "\t\t\t\tdf_eqn.loc[c, 'id'] = 'RIBOSUBSYN-' + ribosub_name + '-FROM-' + ribo.id[i]\n",
    "\t\t\t\tdf_eqn.loc[c, 'type'] = 'ribosome-subunit'\n",
    "\t\t\t\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(ribosub_stoich, arrow='-->')\n",
    "\t\t\t\t# add the ribosome subunit to the ribosome\n",
    "\t\t\t\tribo_stoich[ribosub_name] = -1\n",
    "\t\t# making the ribosome itself from the subunits\n",
    "\t\tc += 1\n",
    "\t\tdf_eqn.loc[c, 'id'] = 'RIBOSYN-ribonuc' if loc == '_c' else 'RIBOSYN-ribomito' if loc == '_m' else 'RIBOSYN-ribo'+loc[1:]\n",
    "\t\t# add to ribosome files\n",
    "\t\tribo_id = \"'\"+loc[1:]+\"'\"\n",
    "\t\tf6.write('\\n'+ribo_id); f5.write('\\n'+ribo_id+\".'\"+df_eqn.loc[c, 'id']+\"'\"); f4.write('\\n'+ribo_id+' '+str(kribo_default))\n",
    "\t\tdf_eqn.loc[c, 'type'] = 'ribosome'\n",
    "\t\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(ribo_stoich, arrow='-->')\n",
    "\tfor file in [f,f2,f4,f5,f6]: file.write('\\n/')\n",
    "# write the ribosome subunit stoichiometry to a json file\n",
    "with open('./model/ribo_subunits.json', 'w') as f:\n",
    "\tjson.dump(ribo_subunits, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Build files related to MW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = getattr(df_pro,'MW (g/mmol)').to_dict()\n",
    "ids = getattr(df_pro,'id').to_list()\n",
    "\n",
    "#build the string to write\n",
    "prot_mw_str = \"\"\n",
    "\n",
    "#for each id\n",
    "for id in ids:\n",
    "\tnew_line = id + \"\\t\" + str(data[id]) + \"\\n\"\n",
    "\tprot_mw_str = prot_mw_str + new_line\n",
    "# add dummy proteins to the string\n",
    "for prot in dummy_list:\n",
    "\tnew_line = prot[\"name\"].replace('PROSYN-','') + \"\\t\" + str(prot[\"MW\"]) + \"\\n\"\n",
    "\tprot_mw_str = prot_mw_str + new_line\n",
    "\n",
    "#write to output\n",
    "with open('./model/pro_mw_g_per_mmol.txt', 'w') as f:\n",
    "\tf.write(prot_mw_str)\n",
    "with open('../input/pro_mw_g_per_mmol.txt', 'w') as f:\n",
    "\tf.write(prot_mw_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TEST: this whole block. get MW list\n",
    "mw_list = getattr(df_enz,\"MW (g/mmol)\").to_list()\n",
    "\n",
    "#get enz list\n",
    "enz_list = getattr(df_enz,\"enz\").to_list()\n",
    "\n",
    "#create string for writing output file\n",
    "enz_mw_str = \"enz_id\\tMW (g/mmol)\\n\"\n",
    "\n",
    "#add to the string for enzyme entries\n",
    "#need to initialize a counter\n",
    "c = 0\n",
    "\n",
    "#list to keep track of existing entries\n",
    "#predefined values prevent adding unknown\n",
    "existing_pro = ['nan','unknown']\n",
    "\n",
    "for enz in enz_list:\n",
    "\n",
    "\t#only add a new line if unique entry\n",
    "\tif not(str(enz) in existing_pro):\n",
    "\n",
    "\t\t#for debugging, comment out if unused\n",
    "\t\t#print(\"enz: |\",enz,\"|, type: \",type(enz))\n",
    "\t\n",
    "\t\tnew_line = str(enz) + \"\\t\" + str(mw_list[c]) + \"\\n\"\n",
    "\n",
    "\t\t#for debugging, comment out if unused\n",
    "\t\t#print(new_line)\n",
    "\n",
    "\t\tenz_mw_str = enz_mw_str + new_line\n",
    "\n",
    "\t\texisting_pro.append(enz)\n",
    "\n",
    "\tc = c + 1\n",
    "\n",
    "#write to output\n",
    "with open('../input/enz_mw_g_per_mmol_norxnmapped.txt', 'w') as f:\n",
    "\tf.write(enz_mw_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Protein file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Protein\n",
    "\n",
    "pro_lengths = dict()\n",
    "\n",
    "c = df_eqn.shape[0] - 1\n",
    "translation_loci = set()\n",
    "unknown_ribo_prosyn = set()\n",
    "protein_functions = dict()\n",
    "\n",
    "def edit_stoich(stoich:OrderedDict,edits):\n",
    "\tif not pd.isna(edits):\n",
    "\t\tcofs_st = edits.split(',')\n",
    "\t\tcofs_st = OrderedDict({c.split(':')[0]:c.split(':')[1] for c in cofs_st})\n",
    "\t\tfor k,v in cofs_st.items():\n",
    "\t\t\tstoich['MET-'+k] = -int(v)\n",
    "\treturn stoich\n",
    "\n",
    "for i in df_pro.index: \n",
    "\tc += 1\n",
    "\tpro_id = df_pro.loc[i, 'id'] if 'id' in df_pro.columns else i\n",
    "\tprepro = 'PRO-' + str(pro_id) # preprotein (from translation)\n",
    "\tprot = 'PTMPRO-' + str(pro_id) # protein after post-translational modifications (PTMs)\n",
    "\n",
    "\ttrans_st = OrderedDict()\n",
    "\tPTM_st = OrderedDict({prepro:-1,prot:1})\n",
    "\tfor met in ['MET-atp_c', 'MET-h2o_c',\n",
    "\t\t\t\t'MET-adp_c', 'MET-pi_c', 'MET-h_c', 'MET-gtp_c',\n",
    "\t\t\t\t'MET-gdp_c']:\n",
    "\t\ttrans_st[met] = 0\n",
    "\tseq = df_pro.sequence[i].replace(\"*\",\"\")\n",
    "\n",
    "\ttrans_st = edit_stoich(trans_st,df_pro.trans_stoich_edits[i])\n",
    "\tPTM_st = edit_stoich(PTM_st,df_pro.PTM_stoich_edits[i])\n",
    "\n",
    "\tpro_lengths[i] = len(seq)\n",
    "\tfor aa in df_aamap.index:\n",
    "\t\ttrans_st[df_aamap.tRNA_in[aa]] = -seq.count(aa)\n",
    "\t\ttrans_st[df_aamap.tRNA_out[aa]] = seq.count(aa)\n",
    "\t\t\t\t\n",
    "\ttrans_st[prepro] = 1\n",
    "\tdf_eqn.loc[c, 'coupling_type'] = 'prot_ribo'\n",
    "\tdf_eqn.loc[c, 'coupling_species'] = 'ribo'\n",
    "\t\n",
    "\t# protein will occupy cellular space in its specific compartment\n",
    "\tif df_pro.subloc_assigned[i] in mito_compartments:\n",
    "\t\ttrans_st['BIO-protmito'] = df_pro.loc[i, 'MW (g/mmol)']\n",
    "\telse:\n",
    "\t\ttrans_st['BIO-protcyt'] = df_pro.loc[i, 'MW (g/mmol)']\n",
    "\t\n",
    "\t# Cost: Initiation: 1 ATP + 2 GTP (initiate and bind Methionine)\n",
    "\tfor met in ['MET-atp_c', 'MET-h2o_c']:\n",
    "\t\ttrans_st[met] -= 1\n",
    "\tfor met in ['MET-adp_c', 'MET-pi_c', 'MET-h_c']:\n",
    "\t\ttrans_st[met] += 1\n",
    "\t\t\t\t\n",
    "\t# Elongation: 2 GTP / cycle (2(n-1) cycles for n amino acids + 2 GTP from initiation = 2n GTP consumed so far)\n",
    "\t# (elongation process excludes Methionine since it is already bound in initiation process)\n",
    "\t# Termination and recycling: 1 GTP + 1 ATP: https://doi.org/10.1534/genetics.115.186221\n",
    "\t# \tHowever, we exclude these costs since it's not always required (e.g., via \"leaky scanning\"): https://doi.org/10.1093/nar/gkp765\n",
    "\tfor met in ['MET-gtp_c', 'MET-h2o_c']:\n",
    "\t\ttrans_st[met] -= 2*len(seq)\n",
    "\tfor met in ['MET-gdp_c', 'MET-pi_c', 'MET-h_c']:\n",
    "\t\ttrans_st[met] += 2*len(seq)\n",
    "\t\n",
    "\tdf_eqn.loc[c, 'id'] = 'PROSYN-' + df_pro.id[i]\n",
    "\n",
    "\t# add translation rxn to gene_expression_dict list, or create a new list if it doesn't exist\n",
    "\tgene_expression_dict.setdefault(df_pro.gene_src[i], []).append(df_eqn.loc[c, 'id'])\n",
    "\n",
    "\t# add to protein functions dictionary under its gene_src\n",
    "\tprotein_functions.setdefault(df_pro.gene_src[i], {})[df_pro.id[i]] = {}\n",
    "\tdf_eqn.loc[c, 'type'] = 'protein'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(trans_st, arrow='-->')\n",
    "\t# Set ribosome translating the protein\n",
    "\tif df_pro.loc[i, 'translation_loc'] in ['unknown','any']:\n",
    "\t\tunknown_ribo_prosyn.add(\"'\"+df_eqn.loc[c, 'id']+\"'\")\n",
    "\telse:\n",
    "\t\ttranslation_loci.add(\"'\"+df_pro.loc[i, 'translation_loc']+\"'.'\"+df_eqn.loc[c, 'id']+\"'\")\n",
    "\t# add PTM reaction\n",
    "\tc += 1\n",
    "\tdf_eqn.loc[c, 'id'] = 'PTM-' + df_pro.id[i]\n",
    "\tdf_eqn.loc[c, 'type'] = 'proteinPTM'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(PTM_st, arrow='-->')\n",
    "# write the protein length file\n",
    "len_str = \"/\\n\"\n",
    "for prot in pro_lengths:\n",
    "\tlen_str = len_str + \"'PROSYN-\" + prot + \"' \" + str(pro_lengths[prot]) + \"\\n\"\n",
    "for prot in dummy_list:\n",
    "\tlen_str += \"'\" + prot[\"name\"] + \"' \" + str(prot[\"medianL\"]) + \"\\n\"\n",
    "len_str = len_str + \"/\"\n",
    "with open('./model/RBA_proteinLength.txt', 'w') as f:\n",
    "\tf.write(len_str)\n",
    "\n",
    "### Dummy protein\n",
    "for prot in dummy_list:\n",
    "\tdummy_MW = prot[\"MW\"]\n",
    "\ttrans_st = make_dummy_protein_stoich(aa_standards_df=df_aamap, prot_df=prot[\"df\"], length=prot[\"medianL\"], rxn_name=prot[\"name\"], mw=dummy_MW)\n",
    "\n",
    "\tc += 1\n",
    "\t# prot_st[\"BIO-protdummy\"] = prot[\"MW\"]\n",
    "\tdf_eqn.loc[c, 'id'] = prot[\"name\"]\n",
    "\tdf_eqn.loc[c, 'coupling_type'] = 'prot_ribo'\n",
    "\tdf_eqn.loc[c, 'coupling_species'] = 'ribo' #TEST: Update for ribonuc, ribomito, and other ribo types\n",
    "\tdf_eqn.loc[c, 'type'] = 'protein'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(trans_st,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\tarrow='-->', floatdecimal=6)\n",
    "\tif prot[\"translation_loc\"] in ['unknown','any']:\n",
    "\t\tunknown_ribo_prosyn.add(\"'\"+df_eqn.loc[c, 'id']+\"'\")\n",
    "\telse:\n",
    "\t\ttranslation_loci.add(\"'\"+df_eqn.loc[c, 'id']+\"'.'\"+df_eqn.loc[c, 'id']+\"'\")\n",
    "\n",
    "# Protein waste reactions\n",
    "c = df_eqn.shape[0] - 1\n",
    "with open('./model/pro_syn_waste.txt', 'w') as f:\n",
    "\tf.write('/')\n",
    "\tfor i in df_pro.index: \n",
    "\t\tfor abbv,tag in {'':'PROSYN','PTM':'PTM'}.items():\n",
    "\t\t\tc += 1\n",
    "\t\t\tpro=abbv+'PRO-'+df_pro.id[i]\n",
    "\t\t\tnew_st = OrderedDict()\n",
    "\t\t\tnew_st[pro] = -1\n",
    "\t\t\t# add the protein's MW as the coefficient for BIO-protwasted\n",
    "\t\t\tnew_st['BIO-protwasted'] = df_pro.loc[i, 'MW (g/mmol)']\n",
    "\t\t\t\n",
    "\t\t\tpw_id = f\"{abbv}PROWASTE-{df_pro.id[i]}\"\n",
    "\t\t\tf.write(f\"\\n'{tag}-{df_pro.id[i]}'.'{pw_id}'\")\n",
    "\n",
    "\t\t\tdf_eqn.loc[c, 'id'] = pw_id\n",
    "\t\t\tdf_eqn.loc[c, 'type'] = 'proteinWaste'\n",
    "\t\t\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA(new_st, arrow='-->')\n",
    "\n",
    "\tf.write('\\n/')\n",
    "# Pseudoreaction representing mass of waste protein (for kapps): PROWASTE-TOTALPROTEIN\n",
    "c += 1\n",
    "df_eqn.loc[c, 'id'] = 'PROWASTE-TOTALPROTEIN'\n",
    "df_eqn.loc[c, 'type'] = 'proteinWaste'\n",
    "df_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA({'BIO-protwasted':-1}, arrow='-->')\n",
    "for loc in ['modeled','dummy','cyt','mito']: # add PROWASTE rxns for biomass-related byproduct of PROSYN rxns, to support simulations at 0 growth\n",
    "\tc += 1\n",
    "\tdf_eqn.loc[c, 'id'] = 'PROWASTE-PROT'+loc.upper()\n",
    "\tdf_eqn.loc[c, 'type'] = 'proteinWaste'\n",
    "\tdf_eqn.loc[c, 'reaction'] = build_reaction_equation_from_metabolites_dict_RBA({'BIO-prot'+loc:-1}, arrow='-->')\n",
    "\n",
    "### Write translation localization lists\n",
    "with open('./model/translation.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(list(translation_loci)) + ['/']))\n",
    "with open('./model/translated_by_any_ribo.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(list(unknown_ribo_prosyn)) + ['/']))\n",
    "\n",
    "### Biomass\n",
    "for i in df_biom.index:\n",
    "\tc += 1\n",
    "\tif not pd.isna(df_biom.rxn_id[i]):\n",
    "\t\tdf_eqn.loc[c, 'id'] = df_biom.rxn_id[i]\n",
    "\t\tdf_eqn.loc[c, 'type'] = 'biomass'\n",
    "\t\tdf_eqn.loc[c, 'reaction'] = df_biom.rxn_equation[i]\n",
    "# make into file RBA_rxns_whole_biomass.txt\n",
    "with open('./model/RBA_rxns_whole_biomass.txt', 'w') as f:\n",
    "\t# find whole_biomass_rxns in df_eqn and write them to file\n",
    "\tf.write('\\n'.join(['/'] + [i for i in whole_biomass_rxns if i in df_eqn.id.to_list()] + ['/']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k,v in gene_expression_dict.items():\n",
    "\t# check for entries in v starting with \"'RIBO\"\n",
    "\tif any([i.startswith(\"'RIBO\") for i in v]):\n",
    "\t\tprint(f\"Warning: {k} has entries starting with 'RIBO' in gene_expression_dict\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save gene list to file\n",
    "with open('./model/genes.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + [f\"'{i}'\" for i in genes] + ['/']))\n",
    "for gene in genes:\n",
    "\tif gene not in gene_expression_dict:\n",
    "\t\tgene_expression_dict[gene] = []\n",
    "# remove duplicates from gene_expression_dict, preserve order\n",
    "gene_expression_dict = {k: list(dict.fromkeys([f\"'{i}'\" for i in v])) for k, v in gene_expression_dict.items()}\n",
    "# save gene-protein mapping to file\n",
    "with open('./model/genes_expression.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + [f\"'{gene}'.({','.join(gene_expression_dict[gene])})\" for gene in genes if gene_expression_dict[gene]] + ['/']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make extra kapp calculation files\n",
    "directory = '../GAMS/parameterization/enz_from_proteome'\n",
    "if not os.path.exists(directory):\n",
    "\tos.makedirs(directory)\n",
    "\n",
    "with open(directory + '/pro_and_enz.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(list(set([\"'ENZ-\" + i + \"'\" for i in enz_list] + [\"'PTMPRO-\" + i + \"'\" for i in df_pro.index]))) + ['/']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save excel files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_eqn.to_csv('./model/RBA_stoichiometry.tsv', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build GAMS Sij and flux bounds (originally A02_build_GAMS_Sij_and_fluxBounds.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load excel file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_eqn = read_spreadsheet('./model/RBA_stoichiometry.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assemble list of species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "met_list_raw = []\n",
    "for i in df_eqn.index:\n",
    "\tmet_dict = metabolites_dict_from_reaction_equation_RBA(df_eqn.reaction[i])\n",
    "\tmet_list_raw += met_dict.keys()\n",
    "\n",
    "met_list = sorted(list(set(met_list_raw)))\n",
    "met_list = [i for i in met_list if i != '']\n",
    "with open('./model/RBA_species.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + [\"'\" + i + \"'\" for i in met_list] + ['/']))\n",
    "# add list of demand rxns for each metabolite (for testing purposes)\n",
    "demand_list = [\"DM-\" + i for i in met_list]\n",
    "# WIP: add to df_eqn\n",
    "with open('./model/RBA_rxns_demand.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + [\"'DM-\" + i + \"'\" for i in met_list] + ['/']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assemble list of reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = df_eqn.id.to_list()\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of uptake reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag,rxn_base_id,rxn_dir,enz_id = extract_details_from_rxnid(i)\n",
    "\tif tag == 'RXN' and rxn_base_id[:3] == 'EX_' and rxn_dir == 'REV':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_EXREV.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of secretion reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag,rxn_base_id,rxn_dir,enz_id = extract_details_from_rxnid(i)\n",
    "\tif tag == 'RXN' and rxn_base_id[:3] == 'EX_' and rxn_dir == 'FWD':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_EXFWD.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of protein waste reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag = i.split('-')[0]\n",
    "\tif tag in ['PROWASTE','PTMPROWASTE']:\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_prowaste.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of enzyme syn reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag = i.split('-')[0]\n",
    "\tif tag == 'ENZSYN':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_enzsyn.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of enzyme load reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag = i.split('-')[0]\n",
    "\tif tag == 'ENZLOAD':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "# list created so GAMS knows which ENZLOAD corresponds to which rxn\n",
    "rxn_enzload_coupling_list = [i + '.' + i.replace('ENZLOAD-','RXN-') + ' 1' for i in rxn_list]\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_enzload.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))\n",
    "with open('./model/RBA_rxn_enzload_coupling.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + rxn_enzload_coupling_list + ['/']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of metabolic network reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_list = []\n",
    "for i in df_eqn.id:\n",
    "\ttag = i.split('-')[0]\n",
    "\tif tag == 'RXN':\n",
    "\t\trxn_list.append(i)\n",
    "\n",
    "rxn_list = [\"'\" + i + \"'\" for i in rxn_list if i != '']\n",
    "rxn_list = ['/'] + rxn_list + ['/']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./model/RBA_rxns_rxnmetabolicnetwork.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rxn_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make k_app (enzyme turnover rate) file draft using placeholder values\n",
    "kapp_list = []\n",
    "kapp_list = [i + \" 360000\" for i in rxn_list if i != '/']\n",
    "kapp_list = ['/'] + kapp_list + ['/']\n",
    "with open('./model/kapp_placeholders.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(kapp_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assemble stoichiometry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make extra kapp calculation files\n",
    "directory = '../GAMS/parameterization/enz_from_proteome/'\n",
    "if not os.path.exists(directory):\n",
    "\tos.makedirs(directory)\n",
    "\n",
    "sij = []\n",
    "rxns_enz = []\n",
    "rxns_proenz = []\n",
    "sij_proenz = []\n",
    "\n",
    "for i in df_eqn.index:\n",
    "\tx = metabolites_dict_from_reaction_equation_RBA(df_eqn.reaction[i])\n",
    "\tmet_dict = dict()\n",
    "\tfor k,v in x.items():\n",
    "\t\tif k == '':\n",
    "\t\t\tcontinue\n",
    "\t\t# note when a protein is used in a reaction other than its PROWASTE and production reactions\n",
    "\t\tmet = k.split('-',1)[1]\n",
    "\t\tif k.startswith('PTMPRO-') and df_eqn.id[i] not in ['PTMPROWASTE-' + met,'PTM-' + met]:\n",
    "\t\t\tfor gene,prots in protein_functions.items():\n",
    "\t\t\t\tif met in prots:\n",
    "\t\t\t\t\tprotein_functions[gene][met].update({df_eqn.id[i]:v})\n",
    "\t\t\t\t\tbreak\n",
    "\t\tif v.is_integer():\n",
    "\t\t\tmet_dict[k] = int(v)\n",
    "\t\telse:\n",
    "\t\t\tmet_dict[k] = v\n",
    "\t\t\t\n",
    "\tif df_eqn.id[i].startswith(\"ENZ\") or df_eqn.id[i].startswith(\"PROSYN\"):\n",
    "\t\trxns_proenz.append(\"'\" + df_eqn.id[i].replace('PROSYN-','PROIN-') + \"'\")\n",
    "\t\tif df_eqn.id[i].startswith(\"ENZ\"):\n",
    "\t\t\trxns_enz.append(\"'\" + df_eqn.id[i] + \"'\")\n",
    "\t\t# sij.append(\"'\" + df_eqn.id[i] + \"'.'\" + df_eqn.id[i] + \"' 1\")\n",
    "\t\tfor k,v in met_dict.items():\n",
    "\t\t\tsij.append(\"'\" + k + \"'.'\" + df_eqn.id[i] + \"' \" + str(v))\n",
    "\t\t\tif k.startswith('PTMPRO-') or k.startswith('ENZ-'):\n",
    "\t\t\t\tsij_proenz.append(\"'\" + k + \"'.'\" + df_eqn.id[i].replace('PROSYN-','PROIN-') + \"' \" + str(v))\n",
    "\telse:\n",
    "\t\tfor k,v in met_dict.items():\n",
    "\t\t\tsij.append(\"'\" + k + \"'.'\" + df_eqn.id[i] + \"' \" + str(v))\n",
    "\t\n",
    "sij = ['/'] + sij + ['/']\n",
    "with open('./model/RBA_sij.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(sij))\n",
    "# write protein functions to file\n",
    "with open('./protein_functions.json', 'w') as f:\n",
    "\tjson.dump(protein_functions, f)\n",
    "# make set of all proteins in protein_functions with no modeled functions\n",
    "proteins_without_modeled_functions = set()\n",
    "# identify them by checking if all values within values are empty\n",
    "for gene,prots in protein_functions.items():\n",
    "\tfor prot,funcs in prots.items():\n",
    "\t\t# if all values are empty, add to set\n",
    "\t\tif not any(funcs.values()):\n",
    "\t\t\tproteins_without_modeled_functions.add(prot)\n",
    "with open('./proteins_without_modeled_functions.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(sorted(list(proteins_without_modeled_functions))))\n",
    "with open(directory + 'rxns_enz.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(rxns_enz) + ['/']))\n",
    "with open(directory + 'rxns_pro_and_enz.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(rxns_proenz) + ['/']))\n",
    "with open(directory + 'sij_pro_and_enz.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sij_proenz + ['/']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build GAMS RBA constraints (originally A03_build_GAMS_RBA_constraints.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### LOAD INPUTS AND PARAMETERS\n",
    "copy_to_gams = True # if True, copy the model to the gams folder\n",
    "\n",
    "# Stoichiometry\n",
    "df_stoich = read_spreadsheet('./model/RBA_stoichiometry.tsv')\n",
    "df_stoich.index = df_stoich.id.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Write prosyn reaction\n",
    "idx = [i for i in df_stoich.index if i[:7] == 'PROSYN-']\n",
    "prosyn = [\"'\" + i + \"'\" for i in idx]\n",
    "prosyn = ['/'] + prosyn + ['/']\n",
    "with open('./model/RBA_rxns_prosyn.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(prosyn))\n",
    "## write ribosyn set\n",
    "idx = [i for i in df_stoich.index if i[:8] == 'RIBOSYN-']\n",
    "ribosyn = [\"'\" + i + \"'\" for i in idx]\n",
    "ribosyn = ['/'] + ribosyn + ['/']\n",
    "with open('./model/RBA_rxns_ribosyn.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(ribosyn))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make fwd and reversible rxn list for metabolic network and all rxns (not needed but helpful for testing)\n",
    "# Also, make flux coupling analysis file in case you use that for testing\n",
    "\n",
    "# Using Patrick's FCA code\n",
    "rxn_types = {'irrev': 0, 'reversible-fwd-half': 1, 'reversible-rev-half': 2, 'pseudoreaction': 3, 'exchange-fwd-half': 4, 'exchange-rev-half': 5}\n",
    "\n",
    "rev_rxn_list = []; rev_list = []\n",
    "fwd_rxn_list = []; fwd_list = []\n",
    "fca_list = []\n",
    "\n",
    "for i in df_stoich.id:\n",
    "\ttag,rxn_base_id,rxn_dir,enz_id = extract_details_from_rxnid(i)\n",
    "\t#print(\"'\"+i+\"' \"+str(rxn_types['irrev']))\n",
    "\t#print(rxn_base_id)\n",
    "\t#print(i,'\\t',tag,'\\t',rxn_dir)\n",
    "\tif rxn_dir == 'FWD':\n",
    "\t\tfwd_list.append(i)\n",
    "\t\tif tag == 'RXN':\n",
    "\t\t\tfwd_rxn_list.append(i)\n",
    "\t\tif rxn_base_id[:3] == 'EX_':\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['exchange-fwd-half']))\n",
    "\t\t# Add as irreversible if no counterpart found in other direction\n",
    "\t\telif i.replace(\"_FWD-\",\"_REV-\") not in df_stoich.id:\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['irrev'])) \n",
    "\t\telse:\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['reversible-fwd-half']))\n",
    "\telif rxn_dir == 'REV':\n",
    "\t\trev_list.append(i)\n",
    "\t\tif tag == 'RXN':\n",
    "\t\t\trev_rxn_list.append(i)\n",
    "\t\tif rxn_base_id[:3] == 'EX_':\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['exchange-rev-half']))\n",
    "\t\t# Add as irreversible if no counterpart found in other direction\n",
    "\t\telif i.replace(\"_REV-\",\"_FWD-\") not in df_stoich.id:\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['irrev'])) \n",
    "\t\telse:\n",
    "\t\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['reversible-rev-half']))\n",
    "\telse:\n",
    "\t\tfca_list.append(\"'\"+i+\"' \"+str(rxn_types['irrev'])) \n",
    "\n",
    "fwd_rxn_list = [\"'\" + i + \"'\" for i in fwd_rxn_list if i != '/']\n",
    "fwd_rxn_list = ['/'] + fwd_rxn_list + ['/']\n",
    "with open('./model/RBA_rxns_rxnmetabolicnetworkFWD.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(fwd_rxn_list))\n",
    "rev_rxn_list = [\"'\" + i + \"'\" for i in rev_rxn_list if i != '/']\n",
    "rev_rxn_list = ['/'] + rev_rxn_list + ['/']\n",
    "with open('./model/RBA_rxns_rxnmetabolicnetworkREV.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rev_rxn_list))\n",
    "\n",
    "fwd_list = [\"'\" + i + \"'\" for i in fwd_list if i != '/']\n",
    "fwd_list = ['/'] + fwd_list + ['/']\n",
    "with open('./model/RBA_rxns_FWD.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(fwd_list))\n",
    "rev_list = [\"'\" + i + \"'\" for i in rev_list if i != '/']\n",
    "rev_list = ['/'] + rev_list + ['/']\n",
    "with open('./model/RBA_rxns_REV.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(rev_list))\n",
    "\n",
    "fca_list = ['/'] + fca_list + ['/']\n",
    "with open('./model/rxntype.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(fca_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating set pairing SM values w/ RBA ones, for use in converting flux constraints from SM format to RBA\n",
    "\n",
    "# filter out rxns with no FBA_name\n",
    "df_sm = df_stoich[df_stoich.FBA_name.notnull()]\n",
    "# get SM rxn IDs from FBA_name column\n",
    "sm_rxn_ids = df_sm.FBA_name.to_list()\n",
    "# get RBA rxn IDs from index\n",
    "rba_rxn_ids = df_sm.index.to_list()\n",
    "rba_rxn_dirs = [extract_details_from_rxnid(i)[2] for i in rba_rxn_ids]\n",
    "# convert dir to -1 if rev, 1 otherwise\n",
    "rba_rxn_dirs = [-1 if i == 'REV' else 1 for i in rba_rxn_dirs]\n",
    "# create list of strings pairing SM rxn IDs with RBA rxn IDs\n",
    "sm_rba_rxn_pairs_dict = {rba_rxn_ids[i]:sm_rxn_ids[i] for i in range(0, len(sm_rxn_ids))}\n",
    "# save as JSON file\n",
    "with open('./model/SM_RBA_rxn_pairs.json', 'w') as f:\n",
    "\tjson.dump(sm_rba_rxn_pairs_dict, f)\n",
    "sm_rba_rxn_pairs = [\"'\"+sm_rxn_ids[i]+\"'.'\"+rba_rxn_ids[i]+\"'\"+' '+str(int(rba_rxn_dirs[i])) for i in range(0, len(sm_rxn_ids))]\n",
    "# convert SM rxn bounds to RBA format; for each sm rxn, get the corresponding rba rxn and its direction\n",
    "with open('./model/SM_rxn_bounds.txt', 'w') as f:\n",
    "\tdelim = ' '\n",
    "\tsm_rxns_checked = []\n",
    "\t# find longest rxn ID for formatting purposes\n",
    "\tmax_sm_rxn_id_len = max([len(i)+2 for i in sm_rxn_ids])+1\n",
    "\tmax_lb_len = max([len(str(rxn.lower_bound)) for rxn in model.reactions if rxn.id in sm_rxn_ids])+1\n",
    "\tmax_ub_len = max([len(str(rxn.upper_bound)) for rxn in model.reactions if rxn.id in sm_rxn_ids])\n",
    "\t# write headers to file, with ljust\n",
    "\tf.write(f\"{'':>{max_sm_rxn_id_len}}{'lo':>{max_lb_len}}{' up':<{max_ub_len}}\")\n",
    "\tfor rba_rxn,sm_rxn in sm_rba_rxn_pairs_dict.items():\n",
    "\t\t# find bounds of sm rxn in model\n",
    "\t\trxn = model.reactions.get_by_id(sm_rxn)\n",
    "\t\tub = rxn.upper_bound\n",
    "\t\tlb = rxn.lower_bound\n",
    "\n",
    "\t\tif sm_rxn not in sm_rxns_checked:\n",
    "\t\t\tsm_rxns_checked.append(sm_rxn)\n",
    "\t\t\tname=f\"'{sm_rxn}'\"\n",
    "\t\t\t# make string with SM rxn ID, lb, and ub, using max lengths for formatting\n",
    "\t\t\tf.write(f\"\\n{name:>{max_sm_rxn_id_len}}{str(lb):>{max_lb_len}} {str(ub):<{max_ub_len}}\")\n",
    "# write to file\n",
    "with open('./model/SM_RBA_rxn_pairs.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted(sm_rba_rxn_pairs) + ['/']))\n",
    "# make file w/ all FBA IDs\n",
    "with open('./model/SM_rxn_ids.txt', 'w') as f:\n",
    "\tf.write('\\n'.join(['/'] + sorted([\"'\"+i+\"'\" for i in set(sm_rxn_ids)]) + ['/']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Applications/anaconda3/envs/rba/lib/python3.9/site-packages/cobra/core/metabolite.py:191: UserWarning: The element 'R' does not appear in the periodic table\n",
      "  warn(f\"The element {e} does not appear in the periodic table\")\n",
      "/Applications/anaconda3/envs/rba/lib/python3.9/site-packages/cobra/core/metabolite.py:191: UserWarning: The element 'Tr' does not appear in the periodic table\n",
      "  warn(f\"The element {e} does not appear in the periodic table\")\n",
      "/Applications/anaconda3/envs/rba/lib/python3.9/site-packages/cobra/core/metabolite.py:191: UserWarning: The element 'X' does not appear in the periodic table\n",
      "  warn(f\"The element {e} does not appear in the periodic table\")\n",
      "/Applications/anaconda3/envs/rba/lib/python3.9/site-packages/cobra/core/metabolite.py:191: UserWarning: The element 'Fd' does not appear in the periodic table\n",
      "  warn(f\"The element {e} does not appear in the periodic table\")\n"
     ]
    }
   ],
   "source": [
    "# make file for all formula weights (some may be inaccurate due to R groups and other non-standard groups)\n",
    "with open('./model/formula_weights.txt','w') as f:\n",
    "\tf.write('/\\n')\n",
    "\tfor met in model.metabolites:\n",
    "\t\t# write met name, formula, and formula weight\n",
    "\t\ttry:\n",
    "\t\t\tformula = str(met.formula)\n",
    "\t\t\tmw = str(met.formula_weight)\n",
    "\t\texcept:\n",
    "\t\t\tmw = 'unknown'\n",
    "\t\t\tformula = 'unknown'\n",
    "\t\tf.write(str(met.id) + '\\t' + formula + '\\t' + mw + '\\n')\n",
    "\tf.write('/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for AA in df_aamap, find all mets in met_list that match its name (without the location)\n",
    "with open('./model/aa.txt', 'w') as f, open('./model/tRNAs.txt', 'w') as f2:\n",
    "\tfis = [f,f2]\n",
    "\tfor fi in fis: fi.write('/')\n",
    "\tfor aa in df_aamap.index:\n",
    "\t\taa_mets = [f\"'{met}'\" for met in met_list if met.rsplit('_',1)[0] == df_aamap.tRNA_in[aa].rsplit('_',1)[0]]\n",
    "\t\tf.write(f\"\\n'{aa}'\")\n",
    "\t\tif aa_mets:\n",
    "\t\t\tf2.write(f\"\\n'{aa}'.({','.join(aa_mets)})\")\n",
    "\tfor fi in fis: fi.write('\\n/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write combinations of gene_src and protein IDs in form 'gene_src'.('protein1','protein2',...)\n",
    "with open('./model/proteins_and_locations.txt', 'w') as f, open('./model/proteins.txt', 'w') as f2:\n",
    "\tfis = [f,f2]\n",
    "\tfor fi in fis: fi.write('/')\n",
    "\tfor gene,prots in protein_functions.items():\n",
    "\t\tprots_str = ','.join([f\"'PROSYN-{prot}'\" for prot in prots])\n",
    "\t\tf.write(f\"\\n'{gene}'.({prots_str})\")\n",
    "\t\tf2.write(f\"\\n'{gene}'\")\n",
    "\tfor fi in fis: fi.write('\\n/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "if copy_to_gams: # automatically copies the model folder to the GAMS folder (removing the previous version), if you choose to\n",
    "\timport shutil\n",
    "\t# copy \"model\" folder to GAMS folder\n",
    "\t# Source folder path\n",
    "\tsource_folder = './model'\n",
    "\n",
    "\t# Destination folder path\n",
    "\tdestination_folder = '../GAMS/model'\n",
    "\n",
    "\t# Remove the existing destination folder\n",
    "\tshutil.rmtree(destination_folder)\n",
    "\n",
    "\t# Copy the folder\n",
    "\tshutil.copytree(source_folder, destination_folder)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rba",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
